// @prose-reader/cfi/dist/index.js
function H(e) {
  const n = [e];
  let t = e;
  for (; t.parentNode; )
    n.push(t.parentNode), t = t.parentNode;
  return n;
}
function U(e, n) {
  if (e === n) return e;
  const t = H(e), i = new Set(t);
  let r = n;
  for (; r; ) {
    if (i.has(r))
      return r;
    r = r.parentNode;
  }
  return null;
}
const G = /[[\]^,();]/g;
function p(e) {
  return e.replace(G, "^$&");
}
const J = /^epubcfi\((.*)\)$/, v = (e) => e.nodeType === Node.ELEMENT_NODE, P = (e) => typeof e == "object" && e !== null && "nodeType" in e && (e.nodeType === Node.ELEMENT_NODE || e.nodeType === Node.TEXT_NODE), K = (e) => e.nodeType === Node.TEXT_NODE;
function Q(e) {
  if (k(e))
    return !1;
  const n = e[e.length - 1];
  return e.length > 1 && (n === void 0 || n.length === 0);
}
function k(e) {
  return e !== null && typeof e == "object" && "parent" in e && "start" in e && "end" in e;
}
function Z(e) {
  const n = e.match(J);
  return n && n[1] || e;
}
function ee(e) {
  const n = [];
  let t = null, i = !1, r = "";
  const s = (f) => {
    n.push(f), t = null, r = "";
  }, l = (f) => {
    r += f, i = !1;
  }, a = Z(e).trim(), c = Array.from(a).concat("");
  for (let f = 0; f < c.length; f++) {
    const o = c[f];
    if (!o) {
      t === "/" || t === ":" ? s([t, parseInt(r, 10)]) : t === "~" ? s(["~", parseFloat(r)]) : t === "@" ? s(["@", parseFloat(r)]) : t === "[" ? s(["[", r]) : t === ";" || t?.startsWith(";") ? s([t, r]) : t === "!" && s(["!", 0]);
      break;
    }
    if (o === "^" && !i) {
      i = !0;
      continue;
    }
    if (t === "!")
      s(["!", 0]);
    else if (t === ",")
      s([",", 0]);
    else if (t === "/" || t === ":") {
      if (/^\d$/.test(o)) {
        l(o);
        continue;
      }
      s([t, parseInt(r, 10)]);
    } else if (t === "~") {
      if (/^\d$/.test(o) || o === ".") {
        l(o);
        continue;
      }
      s(["~", parseFloat(r)]);
    } else if (t === "@") {
      if (o === ":") {
        s(["@", parseFloat(r)]), t = "@";
        continue;
      }
      if (/^\d$/.test(o) || o === ".") {
        l(o);
        continue;
      }
      s(["@", parseFloat(r)]);
    } else if (t === "[")
      if (o === ";" && !i)
        s(["[", r]), t = ";";
      else if (o === "]" && !i)
        s(["[", r]);
      else {
        l(o);
        continue;
      }
    else if (t === ";")
      if (o === "=" && !i)
        t = `;${r}`, r = "";
      else if (o === ";" && !i)
        s([t, r]), t = ";";
      else if (o === "]" && !i)
        s([t, r]);
      else {
        l(o);
        continue;
      }
    else if (t?.startsWith(";"))
      if (o === ";" && !i)
        s([t, r]), t = ";";
      else if (o === "]" && !i)
        s([t, r]);
      else {
        l(o);
        continue;
      }
    else t === null && o === ";" && (t = ";");
    (o === "/" || o === ":" || o === "~" || o === "@" || o === "[" || o === "!" || o === ",") && (t = o);
  }
  return n;
}
function z(e, n) {
  return e ? e.map((t, i) => t[0] === n ? i : null).filter((t) => t !== null) : [];
}
function Y(e, n) {
  const t = [];
  let i = 0;
  for (const r of n)
    t.push(e.slice(i, r)), i = r;
  return t.push(e.slice(i)), t;
}
function te(e) {
  const n = [], t = {};
  let i = -1;
  for (let r = 0; r < e.length; r++) {
    const s = e[r];
    if (!s) continue;
    const [l, a] = s;
    l === "/" ? (i++, n[i] = { index: a }, t[i] = []) : i >= 0 && t[i]?.push(s);
  }
  for (let r = 0; r < n.length; r++) {
    const s = n[r];
    if (!s) continue;
    const l = t[r] || [];
    for (let a = 0; a < l.length; a++) {
      const c = l[a];
      if (!c) continue;
      const [f, o] = c;
      if (f === ":")
        s.offset = o;
      else if (f === "~")
        s.temporal = o;
      else if (f === "@")
        s.spatial = (s.spatial || []).concat(o);
      else if (f === ";s")
        s.side = o;
      else if (f.startsWith(";") && f !== ";s") {
        const u = f.substring(1);
        s.extensions || (s.extensions = {}), s.extensions[u] = o;
      } else if (f === "[") {
        const u = typeof o == "string" && !o.includes(" ") && o.length < 50;
        if (a === 0 && u && !s.id)
          s.id = o;
        else if (o !== "") {
          const d = o.split(",").map((h) => h.trim());
          s.text = d;
        }
      }
    }
  }
  return n;
}
function A(e) {
  const n = z(e, "!");
  return Y(e, n).map(te);
}
function w(e) {
  if (!e)
    throw new Error("CFI string cannot be empty");
  const n = ee(e);
  if (!n || n.length === 0)
    throw new Error("Failed to tokenize CFI string");
  const t = z(n, ",");
  if (t.length === 0)
    return A(n);
  const [i, r, s] = Y(n, t);
  function l(o, u) {
    if (!u || u.length === 0) return [[]];
    const d = u.filter((h) => h[0] !== ",");
    if (d.length === 1 && d[0] && d[0][0] === ":") {
      const h = o.length > 0 ? o[o.length - 1] : void 0, g = d[0];
      if (h && h.length > 0 && g) {
        const x = h.map((y) => ({ ...y })), E = x[x.length - 1];
        return E && (E.offset = g[1]), [x];
      }
      return [[]];
    }
    return A(u);
  }
  const a = A(i || []), c = l(a, r || []), f = l(a, s || []);
  return {
    parent: a,
    start: c,
    end: f
  };
}
function ce(e, n, t = {}) {
  try {
    const i = typeof e != "string" ? e : w(e);
    return ne(i, n, t);
  } catch (i) {
    if (t.throwOnError)
      throw i;
    return { node: null, isRange: !1 };
  }
}
function ne(e, n, t = { asRange: !1 }) {
  if (k(e))
    return re(e, n);
  if (Q(e))
    return N(null);
  const i = e.at(-1);
  if (i)
    return b(i, n, t);
  throw new Error("Invalid CFI structure");
}
function re(e, n) {
  const t = e.parent, i = e.start[0] || [], r = e.end[0] || [];
  let s = n.documentElement;
  if (t.length > 0)
    if (t.length > 1) {
      const d = t[0];
      if (d) {
        const h = b(d, n);
        P(h.node) && (s = h.node);
      }
      if (t.length > 1 && s) {
        const h = t[1];
        if (h) {
          const g = b(h, n);
          P(g.node) && (s = g.node);
        }
      }
    } else {
      const d = t[0];
      if (d) {
        const h = b(d, n);
        P(h.node) && (s = h.node);
      }
    }
  if (!s)
    throw new Error("Failed to resolve parent node in CFI range");
  const l = s.nodeType === Node.TEXT_NODE;
  let a, c, f = 0, o = 0;
  if (l) {
    a = s, c = s;
    const d = i[i.length - 1], h = r[r.length - 1];
    f = (Array.isArray(d?.offset) ? d.offset[0] : d?.offset) ?? 0, o = (Array.isArray(h?.offset) ? h.offset[0] : h?.offset) ?? 0;
  } else {
    const d = i.length === 0 || i.length === 1 && typeof i[0]?.offset == "number", h = r.length === 0 || r.length === 1 && typeof r[0]?.offset == "number";
    if (d) {
      if (!s)
        throw new Error("Failed to resolve parent node in CFI range (start)");
      a = s, f = i[0]?.offset ?? 0;
    } else {
      const g = O(s, i, 0, !0);
      if (!g)
        throw new Error("Failed to resolve start node in CFI range");
      a = g;
      const x = i[i.length - 1];
      f = (Array.isArray(x?.offset) ? x.offset[0] : x?.offset) ?? 0;
    }
    if (h) {
      if (!s)
        throw new Error("Failed to resolve parent node in CFI range (end)");
      c = s, o = r[0]?.offset ?? 0;
    } else {
      const g = O(s, r, 0, !0);
      if (!g) throw new Error("Failed to resolve end node in CFI range");
      c = g;
      const x = r[r.length - 1];
      o = (Array.isArray(x?.offset) ? x.offset[0] : x?.offset) ?? 0;
    }
  }
  const u = n.createRange();
  return u.setStart(a, f), u.setEnd(c, o), {
    ...F(i[i.length - 1]),
    node: u,
    isRange: !0
  };
}
function ie(e) {
  if (e?.side) return e.side;
  if (e?.text && e.text.length > 0) {
    const n = e.text[0];
    if (n) {
      const t = n.match(/^([ab])$/);
      if (t)
        return t[1];
    }
  }
}
function C(e) {
  return e.index % 2 !== 0;
}
function de(e) {
  return (k(e) ? e.end : e).at(-1)?.at(-1)?.extensions;
}
function F(e) {
  const n = ie(e), t = e?.extensions;
  return {
    offset: e?.offset,
    temporal: e?.temporal,
    spatial: e?.spatial,
    side: n,
    extensions: t
  };
}
function N(e, n) {
  return {
    node: e,
    isRange: !1,
    ...F(n)
  };
}
function I(e, n) {
  return {
    node: e,
    isRange: !0,
    ...F(n)
  };
}
function D(e, n, t) {
  const i = e.createRange();
  if (i.selectNodeContents(n), t !== void 0) {
    const r = Array.isArray(t) ? t[0] : t;
    K(n) && i.setStart(n, r || 0);
  }
  return i;
}
function O(e, n, t, i) {
  let r = e;
  for (let s = t; s < n.length; s++) {
    const l = n[s];
    if (!r || !l) break;
    if (C(l)) {
      const a = l.index - 1;
      if (a >= 0 && a < r.childNodes.length)
        r = r.childNodes[a];
      else {
        if (i)
          throw new Error(`Invalid text node index: ${l.index}`);
        r = null;
        break;
      }
    } else {
      const a = Array.from(r.childNodes).filter(
        (f) => f.nodeType === Node.ELEMENT_NODE
      ), c = Math.floor(l.index / 2) - 1;
      if (c >= 0 && c < a.length) {
        const f = a[c];
        f && (r = f);
      } else {
        if (i)
          throw new Error(`Invalid element step index: ${l.index}`);
        r = null;
        break;
      }
    }
  }
  return r;
}
function b(e, n, t = {}) {
  const { throwOnError: i = !1, asRange: r = !1 } = t;
  if (!n) {
    if (i)
      throw new Error("Document is not available");
    return N(null);
  }
  const { node: s, remainingPathIndex: l } = se(n, e);
  if (s && l >= e.length) {
    const o = e.at(-1);
    if (o && C(o)) {
      const u = o.index - 1;
      if (u >= 0 && u < s.childNodes.length) {
        const d = s.childNodes[u];
        return N(d, o);
      }
    }
    if (r) {
      const u = D(n, s, o?.offset);
      return I(u, o);
    }
    return N(s, o);
  }
  let a = s || n.documentElement;
  const c = s ? l : 0;
  if (r && e.length > 0) {
    const o = e[e.length - 1];
    if (o && !C(o)) {
      if (o.index === 0 && a) {
        const d = n.createRange();
        return d.setStart(a, 0), d.setEnd(a, 0), I(d, o);
      }
      const u = O(
        a,
        e.slice(0, -1),
        c,
        i
      );
      if (u) {
        const d = Array.from(u.childNodes).filter(
          (g) => g.nodeType === Node.ELEMENT_NODE
        );
        if (Math.floor(o.index / 2) - 1 === d.length) {
          const g = n.createRange();
          return g.selectNodeContents(u), g.collapse(!1), I(g, o);
        }
      }
    }
  }
  if (a = O(a, e, c, i), !a) {
    if (i)
      throw new Error("Failed to resolve CFI path");
    return N(null);
  }
  const f = e.at(-1);
  if (r) {
    const o = D(n, a, f?.offset);
    return I(o, f);
  }
  return N(a, f);
}
function se(e, n) {
  for (let t = n.length - 1; t >= 0; t--) {
    const i = n[t];
    if (i?.id) {
      const r = e.getElementById(i.id);
      if (r) return { node: r, remainingPathIndex: t + 1 };
    }
  }
  return { node: null, remainingPathIndex: 0 };
}
function V(e, n, t = {}) {
  if (!e.textContent || e.textContent.trim() === "")
    return null;
  const i = e.textContent, r = t.textAssertionLength || 10;
  if (n !== void 0 && n <= i.length) {
    const s = Math.floor(r / 2), l = Math.max(0, n - s), a = Math.min(i.length, n + s);
    return i.substring(l, a);
  }
  return i.substring(0, Math.min(i.length, r));
}
function oe(e) {
  const [n, t] = e, i = Math.max(0, Math.min(100, n)), r = Math.max(0, Math.min(100, t));
  return `@${i}:${r}`;
}
function S(e, n, t) {
  let i = e;
  return n !== void 0 && (i += `~${n}`), t !== void 0 && (i += oe(t)), i;
}
function q(e, n, t) {
  let i = e;
  if (n && (i += `[${p(n)}]`), t) {
    const r = t === "before" ? "b" : "a";
    n ? i = `${i.substring(0, i.length - 1)};s=${r}]` : i += `[;s=${r}]`;
  }
  return i;
}
function R(e, n, t = {}, i) {
  let r = "", s = e, l = null;
  if (e?.nodeType === Node.TEXT_NODE) {
    l = e;
    const c = e.parentNode;
    if (!c)
      throw new Error("Text node doesn't have a parent");
    const o = Array.from(c.childNodes).indexOf(e);
    if (o === -1)
      throw new Error("Node not found in parent's children");
    if (r = `/${o + 1}`, v(c) && c.id) {
      const u = c.id, d = Array.from(c.parentNode?.childNodes || []);
      r = `/${d.slice(0, d.indexOf(c) + 1).filter((x) => x.nodeType === Node.ELEMENT_NODE).length * 2}[${p(u)}]${r}`, s = c.parentNode;
    } else
      s = c;
  }
  let a = null;
  for (l && t.includeTextAssertions && (a = V(l, n, t)); s?.parentNode; ) {
    if (!(l && s === l.parentNode && r.includes(`[${v(s) ? s.id : ""}]`))) {
      const c = s.parentNode, f = Array.from(c.childNodes), o = f.indexOf(s);
      if (o === -1)
        throw new Error("Node not found in parent's children");
      const u = f.slice(0, o + 1).filter((g) => g.nodeType === Node.ELEMENT_NODE);
      let d;
      s.nodeType, Node.ELEMENT_NODE, d = u.length;
      const h = d * 2;
      v(s) && s.id ? r = `/${h}[${p(s.id)}]${r}` : r = `/${h}${r}`;
    }
    if (s.parentNode.nodeName.toLowerCase() === "html")
      break;
    s = s.parentNode;
  }
  return n !== void 0 && (r += `:${n}`), r = S(
    r,
    i?.temporal,
    i?.spatial || t.spatialOffset
  ), r = q(r, a, t.includeSideBias), r = m(r, t.extensions), r;
}
function L(e, n, t, i = {}, r) {
  if (e === n) {
    let c = t !== void 0 ? `:${t}` : "";
    return c = S(c, r?.temporal, r?.spatial), c;
  }
  const s = [];
  let l = n;
  for (; l && l !== e; ) {
    const c = l.parentNode;
    if (!c) break;
    const f = Array.from(c.childNodes), o = f.indexOf(l);
    if (o === -1)
      throw new Error("Node not found in parent's children");
    let u;
    l.nodeType === Node.ELEMENT_NODE ? u = (f.filter(
      (g) => g.nodeType === Node.ELEMENT_NODE
    ).indexOf(l) + 1) * 2 : u = o + 1, v(l) && l.id ? s.unshift(`/${u}[${p(l.id)}]`) : s.unshift(`/${u}`), l = c;
  }
  let a = s.join("");
  if (t !== void 0 && (a += `:${t}`), a = S(a, r?.temporal, r?.spatial), i.includeTextAssertions && n.nodeType === Node.TEXT_NODE) {
    const c = V(n, t, i);
    a = q(
      a,
      c,
      i.includeSideBias
    );
  }
  return a = m(a, i.extensions), a;
}
const le = (e) => Object.entries(e).map(([n, t]) => {
  const i = p(t);
  return `${n}=${encodeURIComponent(i)}`;
}).join(";");
function m(e, n) {
  if (!n || Object.keys(n).length === 0)
    return e;
  const t = le(n);
  if (e.endsWith("]")) {
    const i = e.lastIndexOf("["), r = e.substring(i + 1, e.length - 1);
    return r.includes(t) ? e : `${e.substring(0, e.length - 1)}${r.includes(";"), ";"}${t}]`;
  }
  return e.endsWith("!") ? e : `${e}[;${t}]`;
}
function ae(e, n, t, i, r = {}, s, l) {
  const a = U(e, t);
  if (!a)
    throw new Error("No common ancestor found");
  const c = R(a, void 0, r), f = L(
    a,
    e,
    n,
    r,
    s
  ), o = L(
    a,
    t,
    i,
    r,
    l
  );
  if (r.extensions && Object.keys(r.extensions).length > 0) {
    const u = m(c, r.extensions), d = m(f, r.extensions), h = m(o, r.extensions);
    return `${u},${d},${h}`;
  }
  return `${c},${f},${o}`;
}
const B = (e, n, t = {}, i) => {
  let l = `/6/${(e + 1) * 2}`;
  return n && (l += `[${p(n)}]`), l = i ? m(l, t.extensions) : l, `${l}!`;
};
function ue(e, n = {}) {
  if (P(e))
    return `epubcfi(${R(e, void 0, n)})`;
  let t = "";
  if (!("start" in e))
    return e.spineIndex !== void 0 && (t = B(
      e.spineIndex,
      e.spineId,
      n,
      !e.node
    ), !e.node) ? `epubcfi(${t})` : (t += R(
      e.node ?? null,
      e.offset,
      n,
      e
    ), `epubcfi(${t})`);
  const { start: i, end: r } = e;
  return i.spineIndex !== void 0 && (t = B(
    i.spineIndex,
    i.spineId,
    n,
    !i.node
  )), i.node && r.node && (t += ae(
    i.node,
    i.offset ?? 0,
    r.node,
    r.offset ?? 0,
    n,
    i,
    r
  )), `epubcfi(${t})`;
}
function $(e, n = !1) {
  return typeof e == "string" ? $(w(e), n) : "parent" in e ? n ? e.parent.concat(e.end) : e.parent.concat(e.start) : e;
}
function W(e) {
  return e.offset !== void 0 ? 1 : e.index !== void 0 ? 2 : e.temporal !== void 0 || e.spatial !== void 0 ? 3 : 4;
}
function X(e, n) {
  const t = typeof e == "string" ? w(e) : e, i = typeof n == "string" ? w(n) : n;
  if ("parent" in t || "parent" in i)
    return X($(t), $(i)) || X($(t, !0), $(i, !0));
  for (let r = 0; r < Math.max(t.length, i.length); r++) {
    const s = t[r] || [], l = i[r] || [], a = Math.max(s.length, l.length) - 1;
    for (let c = 0; c <= a; c++) {
      const f = s[c], o = l[c];
      if (!f) return -1;
      if (!o) return 1;
      const u = W(f), d = W(o);
      if (u !== d)
        return u - d;
      if (f.index > o.index) return 1;
      if (f.index < o.index) return -1;
      const h = f.temporal !== void 0, g = o.temporal !== void 0;
      if (h && !g) return 1;
      if (!h && g) return -1;
      if (h && g) {
        if ((f.temporal ?? 0) > (o.temporal ?? 0)) return 1;
        if ((f.temporal ?? 0) < (o.temporal ?? 0)) return -1;
      }
      const x = f.spatial !== void 0, E = o.spatial !== void 0;
      if (x && !E) return 1;
      if (!x && E) return -1;
      if (x && E) {
        const y = f.spatial?.[1] ?? 0, M = o.spatial?.[1] ?? 0;
        if (y > M) return 1;
        if (y < M) return -1;
        const j = f.spatial?.[0] ?? 0, _ = o.spatial?.[0] ?? 0;
        if (j > _) return 1;
        if (j < _) return -1;
      }
      if (c === a) {
        if ((f.offset ?? 0) > (o.offset ?? 0)) return 1;
        if ((f.offset ?? 0) < (o.offset ?? 0)) return -1;
      }
    }
  }
  return 0;
}
function fe(e) {
  let n = `/${e.index}`;
  if (e.id) {
    if (n += `[${p(e.id)}`, e.extensions)
      for (const [i, r] of Object.entries(e.extensions))
        n += `;${i}=${p(r)}`;
    n += "]";
  }
  e.offset !== void 0 && (n += `:${e.offset}`), e.temporal !== void 0 && (n += `~${e.temporal}`), e.spatial && e.spatial.length > 0 && (n += `@${e.spatial.join(":")}`);
  const t = [];
  if (e.text && e.text.length > 0 && t.push(e.text.map(p).join(",")), (e.side || e.extensions && !e.id) && (e.side && t.push(`;s=${e.side}`), e.extensions && !e.id))
    for (const [i, r] of Object.entries(e.extensions))
      t.push(`;${i}=${p(r)}`);
  return t.length > 0 && (n += `[${t.join("")}]`), n;
}
function T(e) {
  return e.map((n) => fe(n)).join("");
}
function he(e) {
  if (Array.isArray(e))
    return `epubcfi(${e.map(T).join("!")})`;
  const n = e.parent.map(T).join("!"), t = e.start.map(T).join("!"), i = e.end.map(T).join("!");
  return `epubcfi(${n},${t},${i})`;
}
export {
  X as compare,
  ue as generate,
  Q as isIndirectionOnly,
  k as isParsedCfiRange,
  w as parse,
  ce as resolve,
  de as resolveExtensions,
  he as serialize
};
