// @prose-reader/shared/dist/index.js
const w = (e, n) => e === n ? !0 : e.length !== n.length ? !1 : e.every((t, r) => t === n[r]), f = (e) => e.split(/[#?]/)[0]?.split(".").pop()?.trim() || "";
function g(e) {
  const n = [];
  if (e.length === 0)
    return "";
  if (typeof e[0] != "string")
    throw new TypeError(`Url must be a string. Received ${e[0]}`);
  e[0].match(/^[^/:]+:\/*$/) && e.length > 1 && (e[0] = e.shift() + e[0]), e[0].match(/^file:\/\/\//) ? e[0] = e[0].replace(/^([^/:]+):\/*/, "$1:///") : e[0] = e[0].replace(/^([^/:]+):\/*/, "$1://");
  for (let l = 0; l < e.length; l++) {
    let o = e[l];
    if (typeof o != "string")
      throw new TypeError(`Url must be a string. Received ${o}`);
    o !== "" && (l > 0 && (o = o.replace(/^[/]+/, "")), l < e.length - 1 ? o = o.replace(/[/]+$/, "") : o = o.replace(/[/]+$/, "/"), n.push(o));
  }
  let t = n.join("/");
  t = t.replace(/\/(\?|&|#[^!])/g, "$1");
  const r = t.split("?");
  return t = r.shift() + (r.length > 0 ? "?" : "") + r.join("&"), t;
}
function y(...e) {
  const n = Array.from(Array.isArray(e[0]) ? e[0] : e);
  return g(n);
}
function b(e) {
  const n = e.split("/");
  return n.pop(), n.join("/");
}
const d = (e) => {
  switch (f(e)) {
    case "png":
      return "image/png";
    case "jpg":
      return "image/jpg";
    case "jpeg":
      return "image/jpeg";
    case "txt":
      return "text/plain";
    case "webp":
      return "image/webp";
    case "xhtml":
      return "application/xhtml+xml";
  }
}, m = ({
  mimeType: e,
  uri: n
}) => (e ?? d(n ?? ""))?.startsWith("application/xhtml+xml"), R = (e) => e.length ? e.indexOf(";") ? e.substring(0, e.indexOf(";")) : e : void 0, a = Object.prototype.hasOwnProperty, E = (e, n) => e === n ? e !== 0 || n !== 0 || 1 / e === 1 / n : !1, O = (e, n, t) => {
  if (e === n)
    return !0;
  if (typeof e != "object" || e === null || typeof n != "object" || n === null)
    return !1;
  const r = Object.keys(e), l = Object.keys(n);
  if (r.length !== l.length)
    return !1;
  const o = t && typeof t.customEqual == "function" ? t.customEqual : E;
  for (let c = 0; c < r.length; c++) {
    const u = r[c] || "";
    if (!a.call(n, u) || !o(e[u], n[u]))
      return !1;
  }
  return !0;
}, _ = (e, n) => e.reduce(
  (t, r) => {
    const l = n(r);
    return t[l] || (t[l] = []), t[l].push(r), t;
  },
  {}
);
function x(e, n) {
  const t = { ...e };
  for (const r in n)
    if (Object.hasOwn(n, r)) {
      const l = n[r];
      (l !== void 0 || !(r in e)) && (t[r] = l);
    }
  return t;
}
const h = () => {
  if (!(typeof window > "u"))
    return window;
};
function $() {
  const e = h()?.__PROSE_READER_DEBUG;
  return e === !0 || e === "true";
}
const s = (e, n = $(), t) => {
  let r = n;
  const l = t?.color ? `color: ${t.color}` : void 0, o = {
    enable: (i) => {
      u(i);
    },
    namespace: (i, p) => s(`${e} [${i}]`, p, t),
    isEnabled: () => r,
    debug: () => {
    },
    info: () => {
    },
    log: () => {
    },
    groupCollapsed: () => {
    },
    groupEnd: () => {
    },
    getGroupArgs: (i) => l ? [`%c${e ? `${e} ${i}` : i}`, l] : [e ? `${e} ${i}` : i],
    warn: () => {
    },
    error: () => {
    }
  }, c = (i) => {
    if (!i) {
      o.debug = () => {
      }, o.info = () => {
      }, o.log = () => {
      }, o.groupCollapsed = () => {
      }, o.groupEnd = () => {
      }, o.warn = () => {
      }, o.error = () => {
      };
      return;
    }
    o.debug = e ? Function.prototype.bind.call(
      console.debug,
      console,
      e,
      `%c${e}`,
      l
    ) : Function.prototype.bind.call(
      console.debug,
      console,
      `%c${e}`,
      l
    ), o.info = Function.prototype.bind.call(
      console.info,
      console,
      `%c${e}`,
      l
    ), o.log = e ? Function.prototype.bind.call(
      console.log,
      console,
      `%c${e}`,
      l
    ) : Function.prototype.bind.call(console.log, console), o.groupCollapsed = Function.prototype.bind.call(
      console.groupCollapsed,
      console
    ), o.groupEnd = Function.prototype.bind.call(console.groupEnd, console), o.warn = Function.prototype.bind.call(
      console.warn,
      console,
      `%c${e}`,
      l
    ), o.error = Function.prototype.bind.call(
      console.error,
      console,
      `%c${e}`,
      l
    );
  }, u = (i) => {
    r !== i && (r = i, c(r));
  };
  return c(r), o;
}, F = {
  ...s(),
  namespace: (e, n, t) => s(`[${e}]`, n, t)
}, T = "prose-reader-resource-error";
export {
  T as PROSE_READER_RESOURCE_ERROR_INJECTED_META_NAME,
  F as Report,
  w as arrayEqual,
  d as detectMimeTypeFromName,
  b as getParentPath,
  f as getUrlExtension,
  _ as groupBy,
  E as is,
  O as isShallowEqual,
  m as isXmlBasedMimeType,
  R as parseContentType,
  x as shallowMergeIfDefined,
  y as urlJoin
};
