// @prose-reader/core/dist/index.js
import { switchMap as P, of as v, fromEvent as G, take as Dt, map as y, from as H, takeUntil as T, defer as He, Observable as se, Subject as R, combineLatest as O, merge as w, EMPTY as X, NEVER as Se, tap as x, share as A, BehaviorSubject as Y, distinctUntilChanged as D, filter as M, mergeMap as fe, first as j, endWith as Ie, catchError as Te, finalize as ye, lastValueFrom as pn, scheduled as gn, animationFrameScheduler as be, exhaustMap as hn, debounceTime as De, withLatestFrom as z, throttleTime as mn, startWith as ge, switchScan as fn, shareReplay as Z, defaultIfEmpty as Sn, auditTime as yn, forkJoin as In, delay as Fe, timer as Me, skip as Rt, ReplaySubject as lt, identity as Ce, concatMap as bn } from "rxjs";
import { Report as vn, isShallowEqual as V, shallowMergeIfDefined as Le, getParentPath as wn, parseContentType as Pn, detectMimeTypeFromName as et, arrayEqual as xn } from "@prose-reader/shared";
import { isShallowEqual as As } from "@prose-reader/shared";
import { switchMap as B, first as $n, map as oe, distinctUntilChanged as Pe, startWith as Tn, shareReplay as tt, tap as k, pairwise as Fn, take as Cn, takeUntil as Q, filter as me, debounceTime as Ln, skip as Ot, mergeMap as Mn, catchError as pt, share as En } from "rxjs/operators";
import { generate as Be, parse as Vt, isIndirectionOnly as Nn, isParsedCfiRange as An, resolve as Dn } from "@prose-reader/cfi";
const Rn = "@prose-reader/core", E = vn.namespace(Rn, void 0, {
  color: "#98cde7"
}), On = [
  "pointercancel",
  "pointerdown",
  "pointerenter",
  "pointerleave",
  "pointermove",
  "pointerout",
  "pointerover",
  "pointerup"
  // `touchstart` as const,
  // `touchend` as const,
];
function Vn(t, e, n) {
  if ("caretPositionFromPoint" in t)
    return t.caretPositionFromPoint(e, n);
  if ("caretRangeFromPoint" in t && // @ts-expect-error limited availability
  typeof t.caretRangeFromPoint < "u")
    return t.caretRangeFromPoint(e, n);
}
const kn = (t, e) => {
  const n = "createRange" in t ? t : t.ownerDocument;
  if (!n) return;
  const i = "body" in t ? Xe(t.body, e) : Xe(t, e);
  if (i) {
    let o, r = 0;
    const s = n.createRange();
    return Array.from(i.childNodes).some((a) => {
      s.selectNodeContents(a);
      const c = s.getClientRects(), d = jn(c, e);
      if (d) {
        o = s.cloneRange();
        const u = Vn(
          n,
          Math.ceil(d.left),
          Math.ceil(d.top)
        );
        return u && "startContainer" in u && u.startContainer === o.startContainer && (r = u.startOffset), u && "offsetNode" in u && u.offsetNode === o.startContainer && (r = u.offset), !0;
      }
      return !1;
    }), o ? {
      node: o.startContainer,
      offset: r
    } : { node: i, offset: 0 };
  }
}, Xe = (t, e) => {
  const n = t.getBoundingClientRect(), i = kt(
    n,
    e
  );
  let o;
  i !== "before" && i !== "after" && (o = t);
  for (const r of t.children) {
    const s = Xe(r, e);
    if (s)
      return s;
  }
  return o;
};
function kt(t, { left: e, right: n }) {
  return t.left <= e && t.right <= e ? "before" : t.left <= e && t.right > e && t.right <= n ? "partially-before" : t.left <= n && t.right > n ? "partially-after" : t.left > n ? "after" : "within";
}
function jn(t, e) {
  return Array.from(t).find((n) => {
    const i = kt(n, e);
    return i !== "before" && i !== "after";
  });
}
const Un = (t, e) => {
  if (t.nodeType !== Node.CDATA_SECTION_NODE && t.nodeType !== Node.DOCUMENT_TYPE_NODE) {
    const n = t.ownerDocument?.createRange();
    n?.selectNodeContents(t);
    try {
      e <= (n?.endOffset || 0) && n?.setStart(t, e || 0);
    } catch (i) {
      E.error(i);
    }
    return n;
  }
}, nt = (t) => {
  if (t?.target && t?.target?.ownerDocument?.defaultView) {
    const e = t?.target?.ownerDocument?.defaultView;
    if (e.PointerEvent && t instanceof e.PointerEvent)
      return !0;
  }
  if (t?.view?.window) {
    const e = t?.view;
    if (e.PointerEvent && t instanceof e.PointerEvent)
      return !0;
  }
  return !!On.includes(t.type);
}, zn = (t) => {
  if (nt(t)) return !1;
  if (t?.target && t?.target?.ownerDocument?.defaultView) {
    const e = t?.target?.ownerDocument?.defaultView;
    if (e.MouseEvent)
      return t instanceof e.MouseEvent;
  }
  if (t?.view?.window) {
    const e = t?.view;
    if (e.MouseEvent)
      return t instanceof e.MouseEvent;
  }
  return !1;
}, Wn = (t) => {
  if (t?.target && t?.target?.ownerDocument?.defaultView) {
    const e = t?.target?.ownerDocument?.defaultView;
    if (e.TouchEvent)
      return t instanceof e.TouchEvent;
  }
  if (t?.view?.window) {
    const e = t?.view;
    if (e.TouchEvent)
      return t instanceof e.TouchEvent;
  }
  return !1;
}, _n = () => document.createElement("div"), jt = (t) => {
  const e = [
    "img",
    // Images
    "video",
    // Video files
    "audio",
    // Audio files
    "source",
    // Source elements within video/audio
    "link",
    // Stylesheets and other linked resources
    "script"
    // JavaScript files
  ].join(",");
  return Array.from(t?.querySelectorAll(e) || []);
}, Hn = (t) => {
  if (jt(t).forEach((n) => {
    const i = n.getAttribute("src") || n.getAttribute("href");
    i?.startsWith("blob:") && t?.defaultView?.URL.revokeObjectURL(i);
  }), t) {
    const n = Array.from(t.styleSheets || []);
    for (const i of n) {
      const r = (() => {
        try {
          return Array.from(i.cssRules || []);
        } catch (s) {
          return E.warn(`Error getting rules for sheet: ${i.href}`, s), [];
        }
      })();
      for (const s of r)
        if (t.defaultView && s instanceof t.defaultView.CSSFontFaceRule) {
          const c = s.style.getPropertyValue("src").match(/blob:[^,\s'")]+/g);
          c && c.forEach((d) => {
            t?.defaultView?.URL.revokeObjectURL(d);
          });
        }
    }
  }
};
function Ut(t, e, n = []) {
  if (typeof t != "object" || t === null) return !1;
  for (const i of e)
    if (!(i in t)) return !1;
  for (const i of n)
    if (i in t && typeof t[i] != "function")
      return !1;
  return !0;
}
function zt(t) {
  return Ut(
    t,
    ["nodeType"],
    []
    // biome-ignore lint/suspicious/noExplicitAny: TODO
  ) && t.nodeType === Node.ELEMENT_NODE;
}
function Bn(t, e) {
  return zt(t) && t.tagName.toLowerCase() === e.toLowerCase();
}
function Xn(t) {
  return Ut(
    t,
    [
      "startContainer",
      "endContainer",
      "startOffset",
      "endOffset",
      "collapsed",
      "commonAncestorContainer"
    ],
    ["setStart", "setEnd", "selectNodeContents"]
  );
}
const it = (t, e, n, i) => {
  const o = t.createElement("style");
  return o.id = e, o.innerHTML = n, i ? t.head.prepend(o) : t.head.appendChild(o), () => {
    ot(t, e);
  };
}, ot = (t, e) => {
  if (t?.head) {
    const n = t.getElementById(e);
    n && n.remove();
  }
}, Ye = (t, e) => {
  const n = new RegExp(`${e}\\s*=\\s*([^,\\s]+)`, "i");
  return t.match(n)?.[1]?.trim().toLowerCase();
}, gt = (t, e) => {
  const n = Ye(t, e);
  if (!n) return 0;
  const i = Number.parseFloat(n);
  return Number.isFinite(i) ? i : 0;
}, Yn = (t, e, n, i) => {
  t?.contentDocument?.head && it(t.contentDocument, e, n, i);
}, bs = (t, e) => {
  t?.contentDocument && ot(t.contentDocument, e);
}, re = (t, e, n, i) => {
  if (!t) return;
  const o = t?.contentDocument?.getElementById(
    e
  );
  if (o) {
    o.innerHTML = n;
    return;
  }
  Yn(t, e, n, i);
}, rt = (t) => {
  if (t?.contentDocument) {
    const n = t.contentDocument.querySelector("meta[name='viewport']");
    if (n) {
      const i = n.getAttribute("content");
      if (i) {
        const o = Ye(i, "width"), r = Ye(
          i,
          "height"
        ), s = gt(i, "width"), a = gt(i, "height");
        return s > 0 && a > 0 ? {
          hasViewport: !0,
          widthToken: o,
          heightToken: r,
          width: s,
          height: a
        } : { hasViewport: !0, widthToken: o, heightToken: r };
      }
    }
  }
  return { hasViewport: !1 };
}, qn = (t) => t.pipe(
  P((e) => e.src === "about:blank" && e.contentDocument?.readyState === "complete" && e.contentDocument.body ? v(e) : G(e, "load").pipe(
    Dt(1),
    y(() => e)
  ))
), Gn = (t) => t.pipe(
  P((e) => {
    const n = e?.contentDocument?.fonts.ready;
    return n ? H(n).pipe(y(() => e)) : v(e);
  })
), Zn = (t) => (e) => {
  const n = t(e), i = new IntersectionObserver((o) => {
    o.forEach((r) => {
      r.isIntersecting ? r.target.removeAttribute("tab-index") : r.target.setAttribute("tab-index", "-1");
    });
  }, {});
  return n.hookManager.register(
    "item.onDocumentLoad",
    ({ itemId: o, destroy: r }) => {
      const s = n.spineItemsManager.get(o);
      if (!s) return;
      const a = s.renderer.getDocumentFrame();
      if (!a) return;
      re(
        a,
        "prose-reader-accessibility",
        `
              :focus-visible {
                
                outline: -webkit-focus-ring-color auto 1px;
              }
            `
      );
      const c = a.contentDocument?.body.querySelectorAll("a");
      c?.forEach((d) => {
        i.observe(d);
      }), r(() => {
        c?.forEach((d) => {
          i.unobserve(d);
        });
      });
    }
  ), {
    ...n
  };
}, Jn = (t) => (e) => {
  const n = t(e);
  return n.context.watch("rootElement").pipe(T(n.context.destroy$)).subscribe((i) => {
    if (!i) return;
    const o = () => {
      n.settings.values.computedPageTurnMode === "controlled" && i.scrollTo(0, 0);
    };
    i.addEventListener("scroll", o);
  }), n.hookManager.register("item.onDocumentLoad", ({ itemId: i }) => {
    const r = n.spineItemsManager.get(i)?.renderer.getDocumentFrame();
    r && r.contentDocument?.body.setAttribute("tabindex", "-1");
  }), n;
}, Ve = ({
  position: t,
  frameElement: e
}) => {
  const n = e.getBoundingClientRect(), i = n.width / e.offsetWidth, o = n.height / e.offsetHeight, { left: r = 0, top: s = 0 } = n, a = t.clientX * i + r, c = t.clientY * o + s;
  return {
    clientX: a,
    clientY: c
  };
}, Kn = (t, e, n, i) => {
  const o = e?.view?.frameElement;
  if (!e || !o) return t;
  const r = n.getSpineItemFromIframe(o), s = o, { height: a, width: c } = i.pageSize;
  if (!r || !(s instanceof HTMLIFrameElement)) return t;
  if (nt(t)) {
    const { clientX: d, clientY: u } = Ve({
      position: t,
      frameElement: s
    }), l = new PointerEvent(t.type, {
      ...t,
      pointerId: t.pointerId,
      clientX: d,
      clientY: u
    });
    return Object.defineProperty(l, "target", {
      value: e.target,
      enumerable: !0
    }), l;
  }
  if (zn(t)) {
    const { clientX: d, clientY: u } = Ve({
      position: t,
      frameElement: s
    }), l = new MouseEvent(t.type, {
      ...t,
      clientX: d,
      clientY: u
    });
    return Object.defineProperty(l, "target", {
      value: e.target,
      enumerable: !0
    }), l;
  }
  if (Wn(t)) {
    const d = Array.from(t.touches).map((l) => {
      const { clientX: g, clientY: p } = Ve({
        position: l,
        frameElement: s
      });
      return new Touch({
        identifier: l.identifier,
        target: l.target,
        clientX: g,
        clientY: p
      });
    }), u = new TouchEvent(t.type, {
      touches: d,
      changedTouches: d,
      targetTouches: d
    });
    return Object.defineProperty(u, "target", {
      value: e.target,
      enumerable: !0
    }), u;
  }
  return t;
}, Qn = [
  "pointercancel",
  "pointerdown",
  "pointerenter",
  "pointerleave",
  "pointermove",
  "pointerout",
  "pointerover",
  "pointerup"
], ei = [
  ...Qn
  /*, ...mouseEvents*/
], ti = (t) => (e) => {
  const n = t(e);
  return n.hookManager.register(
    "item.onDocumentLoad",
    ({ destroy: i, itemId: o }) => {
      const r = n.spineItemsManager.get(o), s = r?.renderer.getDocumentFrame();
      if (!s || !r) return;
      const a = ei.map((c) => {
        const d = (u) => {
          let l = u;
          if (nt(u) && (l = new PointerEvent(u.type, u)), l !== u) {
            const g = Kn(
              l,
              u,
              n.spine.locator,
              n.viewport
            );
            n.context.value.rootElement?.dispatchEvent(g);
          }
        };
        return s.contentDocument?.addEventListener(c, d), () => {
          s.contentDocument?.removeEventListener(c, d);
        };
      });
      i(() => {
        a.forEach((c) => {
          c();
        });
      });
    }
  ), n;
}, vs = (t) => {
  const e = new FileReader();
  return new Promise((n) => {
    e.addEventListener(
      "load",
      () => {
        n(e.result);
      },
      !1
    ), e.readAsDataURL(t);
  });
}, ni = (t, e) => Object.entries(t).reduce(
  (n, [i, o]) => e.includes(i) ? {
    ...n,
    [i]: o
  } : n,
  {}
), Wt = (t) => oe((e) => ni(e, t)), st = (t) => (e) => e.pipe(Wt(t), Pe(V));
function at(t) {
  return new se((e) => {
    const n = new ResizeObserver((i) => {
      e.next(i);
    });
    return n.observe(t), () => {
      n.disconnect();
    };
  });
}
const Re = (t) => (e) => e.pipe(
  B(
    (n) => t.pipe(
      $n(),
      oe(() => n)
    )
  )
), ii = (t) => {
  let e;
  const n = t.subscribe((i) => {
    e = { result: i };
  });
  return () => (n.unsubscribe(), e ? v(e.result) : t);
};
function _t() {
  return new se((t) => {
    if (window.requestIdleCallback) {
      const n = window.requestIdleCallback(() => {
        t.next(), t.complete();
      });
      return () => cancelIdleCallback(n);
    }
    const e = setTimeout(() => {
      t.next(), t.complete();
    }, 1);
    return () => clearTimeout(e);
  });
}
function ht(t) {
  return He(() => _t().pipe(B(t)));
}
const mt = (t, e) => new se((n) => {
  const i = new MutationObserver((o) => {
    n.next(o);
  });
  return i.observe(t, e), () => i.disconnect();
});
function ws(t, e) {
  return new se((n) => {
    const i = new IntersectionObserver((o) => {
      n.next(o);
    }, e);
    return i.observe(t), () => {
      i.disconnect();
    };
  });
}
class Ht {
  constructor(e, n) {
    this.settingsManager = n;
    const i = Le(
      this.getDefaultSettings(),
      e
    );
    this.outputSettings = this.computeOutputSettings(i), this.inputSettings = Le(
      this.getDefaultSettings(),
      e
    ), this.outputSettingsUpdateSubject = new R(), this.values$ = O([
      this.settingsManager.values$,
      this.outputSettingsUpdateSubject.pipe(Tn(this.outputSettings))
    ]).pipe(
      oe(([o, r]) => ({ ...o, ...r })),
      tt(1)
    ), this.values$.subscribe();
  }
  _prepareUpdate(e) {
    const n = this.getCleanedParentInputSettings(e), i = this.settingsManager._prepareUpdate(n), o = Le(this.inputSettings, e), r = this.computeOutputSettings(o), s = this.hasSettingsChanged(r);
    return {
      hasChanged: s || i.hasChanged,
      commit: () => (this.inputSettings = o, this.outputSettings = r, !i.hasChanged && s && this.outputSettingsUpdateSubject.next(r), {
        ...i.commit(),
        ...r
      })
    };
  }
  update(e) {
    const { commit: n } = this._prepareUpdate(e);
    n();
  }
  get values() {
    return {
      ...this.settingsManager.values,
      ...this.outputSettings
    };
  }
  watch(e) {
    return Array.isArray(e) ? this.values$.pipe(st(e)) : this.values$.pipe(
      oe((n) => n[e]),
      Pe(V)
    );
  }
  destroy() {
    this.outputSettingsUpdateSubject.complete();
  }
}
let oi = class extends Ht {
  computeOutputSettings(e) {
    return e;
  }
  hasSettingsChanged(e) {
    return !V(this.outputSettings, e);
  }
  getCleanedParentInputSettings(e) {
    const {
      fontJustification: n,
      fontScale: i,
      fontWeight: o,
      lineHeight: r,
      ...s
    } = e;
    return s;
  }
  getDefaultSettings() {
    return {
      fontScale: 1,
      fontWeight: "publisher",
      lineHeight: "publisher",
      fontJustification: "publisher"
    };
  }
};
const ri = (t) => (e) => {
  const { fontScale: n, lineHeight: i, fontWeight: o, fontJustification: r } = e, s = new R(), a = t(e), c = new oi(
    {
      fontScale: n,
      lineHeight: i,
      fontWeight: o,
      fontJustification: r
    },
    a.settings
  ), d = () => `
    
    body {
      ${c.values.fontScale !== 1 ? `font-size: ${c.values.fontScale}em !important;` : ""}
      ${c.values.lineHeight !== "publisher" ? `line-height: ${c.values.lineHeight} !important;` : ""}
      ${c.values.fontWeight !== "publisher" ? `font-weight: ${c.values.fontWeight} !important;` : ""}
      ${c.values.fontJustification !== "publisher" ? `text-align: ${c.values.fontJustification} !important;` : ""}
    }
  `, u = (g) => {
    a.spineItemsManager.items.forEach((p) => {
      if (p.renditionLayout !== "pre-paginated") {
        const h = p.renderer.getDocumentFrame();
        h && re(h, "prose-reader-fonts", d());
      }
    }), g && a.layout();
  };
  a.hookManager.register("item.onDocumentLoad", ({ itemId: g }) => {
    const p = a.spineItemsManager.get(g);
    if (p?.renditionLayout !== "pre-paginated") {
      const h = p?.renderer.getDocumentFrame();
      h && re(h, "prose-reader-fonts", d());
    }
  });
  const l = (g) => g.pipe(
    Fn(),
    y(([p, h]) => h.fontScale !== p.fontScale || h.lineHeight !== p.lineHeight)
  );
  return c.values$.pipe(
    l,
    k(u),
    T(a.$.destroy$)
  ).subscribe(), {
    ...a,
    destroy: () => {
      s.complete(), c.destroy(), a.destroy();
    },
    settings: c
  };
}, si = (t) => (e) => {
  const n = t(e), i = (o) => G(o, "keyup").pipe(
    y((r) => {
      const { pageTurnDirection: s, computedPageTurnMode: a } = n.settings.values;
      if (a === "scrollable")
        return r;
      if (s === "horizontal") {
        if (r.key === "ArrowRight")
          return n.navigation.turnRight();
        if (r.key === "ArrowLeft")
          return n.navigation.turnLeft();
      }
      if (s === "vertical") {
        if (r.key === "ArrowDown")
          return n.navigation.turnRight();
        if (r.key === "ArrowUp")
          return n.navigation.turnLeft();
      }
      return r;
    })
  );
  return i(document).pipe(T(n.$.destroy$)).subscribe(), n.spineItemsManager.items$.pipe(
    P(
      (o) => w(
        ...o.map(
          (r) => r.watch("isLoaded").pipe(
            P(() => {
              const s = r.renderer.getDocumentFrame();
              return s instanceof HTMLIFrameElement && s?.contentDocument ? i(s.contentDocument) : X;
            })
          )
        )
      )
    ),
    T(n.$.destroy$)
  ).subscribe(), n;
}, ai = (t) => t.spine.spineItemsManager.items$.pipe(
  P(
    (e) => w(
      ...e.map((n) => n.watch("isLoaded").pipe(
        P(() => {
          const i = n.renderer.getDocumentFrame();
          if (!i || !i?.contentDocument) return Se;
          const r = Array.from(
            i.contentDocument.querySelectorAll("a")
          ).map(
            (s) => G(s, "click")
          );
          return w(...r);
        })
      ))
    )
  ),
  x((e) => {
    e.preventDefault();
  }),
  A()
);
class ee extends se {
  constructor(e) {
    super((n) => this.stateSubject.pipe(T(this._destroy$)).subscribe(n)), this._destroy$ = new R(), this.destroy$ = this._destroy$.asObservable(), this.stateSubject = new Y(e);
  }
  next(e) {
    this.stateSubject.next(e);
  }
  /**
   * Default shallow compare.
   */
  mergeCompare(e) {
    const n = { ...this.value, ...e };
    V(this.value, n) || this.stateSubject.next(n);
  }
  watch(e) {
    return Array.isArray(e) ? this.stateSubject.pipe(st(e)) : this.stateSubject.pipe(
      y((n) => n[e]),
      D(V)
    );
  }
  get value() {
    return this.stateSubject.value;
  }
  destroy() {
    this.stateSubject.complete(), this._destroy$.complete();
  }
}
const Ae = class Ae extends ee {
  constructor(e) {
    super({
      state: "idle",
      error: void 0
    }), this.triggerSubject = new R(), this.context = e.context, this.settings = e.settings, this.hookManager = e.hookManager, this.item = e.item, this.containerElement = e.containerElement, this.resourcesHandler = e.resourcesHandler, this.viewport = e.viewport;
    const n = this.triggerSubject.pipe(
      M((o) => o.type === "unload")
    ), i = this.triggerSubject.pipe(
      M((o) => o.type === "load")
    );
    this.loaded$ = i.pipe(
      fe(() => this.value.state === "loaded" || this.value.state === "loading" ? X : (this.next({ state: "loading", error: void 0 }), this.onCreateDocument().pipe(j()).pipe(
        fe((s) => (this.hookManager.execute("item.onDocumentCreated", this.item.id, {
          itemId: this.item.id,
          documentContainer: s
        }), this.onLoadDocument().pipe(
          Ie(null),
          j()
        ).pipe(
          Re(this.context.bridgeEvent.viewportFree$),
          P(() => {
            const c = this.hookManager.execute("item.onDocumentLoad", this.item.id, {
              itemId: this.item.id,
              documentContainer: s
            }).filter(
              (d) => d instanceof se
            );
            return O([v(null), ...c]).pipe(j());
          })
        ))),
        y(() => {
          this.next({ state: "loaded", error: void 0 });
        }),
        T(n)
      ))),
      A()
    ), this.unloaded$ = n.pipe(
      fe(() => this.value.state === "unloading" || this.value.state === "idle" ? X : (this.next({ state: "unloading", error: void 0 }), this.context.bridgeEvent.viewportFree$.pipe(
        j(),
        P(() => (this.hookManager.destroy("item.onDocumentLoad", this.item.id), He(() => this.onUnload()).pipe(
          Ie(null),
          j(),
          Te((s) => (E.error("Error unloading document", s), v(null)))
        ))),
        y(() => {
          this.next({ state: "idle", error: void 0 });
        }),
        T(i)
      ))),
      A()
    ), w(this.loaded$, this.unloaded$).pipe(
      Te((o) => (this.next({ state: "error", error: o }), X)),
      T(this.destroy$)
    ).subscribe();
  }
  setDocumentContainer(e) {
    this._documentContainer = e, this._documentContainer.classList.add(
      Ae.DOCUMENT_CONTAINER_CLASS_NAME
    );
  }
  attach() {
    this.documentContainer && this.containerElement.appendChild(this.documentContainer);
  }
  detach() {
    this._documentContainer?.remove(), this._documentContainer = void 0;
  }
  get state$() {
    return this.stateSubject;
  }
  get isLoaded$() {
    return this.state$.pipe(y((e) => e.state === "loaded"));
  }
  load() {
    this.triggerSubject.next({ type: "load" });
  }
  unload() {
    this.triggerSubject.next({ type: "unload" });
  }
  /**
   * Automatically release on complete or error.
   */
  renderHeadless() {
    const e = new R();
    return He(() => this.onRenderHeadless({ release: e })).pipe(
      Ie(void 0),
      j(),
      y((n) => {
        if (n)
          return {
            doc: n,
            release: () => {
              e.next(void 0);
            }
          };
      }),
      ye(() => {
        e.complete();
      }),
      Te((n) => (E.error(n), v(void 0)))
    );
  }
  layout(e) {
    return this.onLayout(e);
  }
  destroy() {
    this.unload(), this.stateSubject.complete(), super.destroy();
  }
  get documentContainer() {
    return this._documentContainer;
  }
  get writingMode() {
  }
  get readingDirection() {
  }
  get renditionLayout() {
    const e = this.item.renditionLayout;
    if (e) return e;
    const n = this.getDocumentFrame();
    if (n) {
      const { hasViewport: i } = rt(n);
      if (i) return "pre-paginated";
    }
    return this.context.manifest?.renditionLayout ?? "reflowable";
  }
};
Ae.DOCUMENT_CONTAINER_CLASS_NAME = "prose-reader-document-container";
let Ee = Ae;
const ci = (t) => new URL(t.href);
class ct {
  constructor(e, n) {
    this.item = e, this.settings = n;
  }
  async getResource() {
    return await pn(
      this.settings.values.getResource?.(this.item) ?? v(void 0)
    ) ?? ci(this.item);
  }
  async fetchResource() {
    const e = await this.getResource();
    return e instanceof Response ? e : e instanceof URL ? fetch(e) : e;
  }
}
const Bt = (t, e) => {
  const n = t.startsWith("file://"), i = n ? t.replace("file://", "http://") : t, o = new URL(e, i).toString();
  return n ? o.replace("http://", "file://") : o;
}, ui = async (t, e, n, i, o) => {
  if (!t || !t.defaultView) return;
  const r = e.sheet;
  if (r)
    try {
      const s = Array.from(r.cssRules || []);
      for (let a = 0; a < s.length; a++) {
        const c = s[a];
        if (t.defaultView && c instanceof t.defaultView.CSSFontFaceRule) {
          const d = c.style.getPropertyValue("src");
          if (d.match(/url\(['"]?([^'"]+)['"]?\)/g)) {
            const l = d.split(",").map((h) => h.trim()), g = await Promise.all(
              l.map(async (h) => {
                if (h.startsWith("local("))
                  return h;
                const m = h.match(/url\(['"]?([^'"]+)['"]?\)/);
                if (!m) return h;
                const f = m[1] ?? "", S = i.manifest?.items.find(({ href: I }) => `${Bt(n, f).toLowerCase()}`.endsWith(
                  `${I.toLowerCase()}`
                ));
                if (S) {
                  const I = new ct(S, o);
                  try {
                    const b = await I.getResource();
                    if (b instanceof Response) {
                      const F = await b.blob(), $ = t.defaultView?.URL.createObjectURL(F);
                      return h.replace(
                        m[0],
                        `url("${$}")`
                      );
                    }
                  } catch (b) {
                    console.error("Error loading font:", b);
                  }
                }
                return h;
              })
            ), p = c.cssText.replace(
              /src:\s*[^;]+;/,
              `src: ${g.join(", ")};`
            );
            r.deleteRule(a), r.insertRule(p, a);
          }
        }
      }
    } catch (s) {
      console.error("Could not access stylesheet rules:", s);
    }
}, di = (t, e, n, i, o) => {
  const r = e.getAttribute("src") || e.getAttribute("href");
  if (!r) return v(null);
  const s = i.manifest?.items.find(({ href: c }) => `${Bt(n, r).toLowerCase()}`.endsWith(
    `${c.toLowerCase()}`
  ));
  if (!s) return v(null);
  const a = new ct(s, o);
  return H(a.getResource()).pipe(
    fe(
      (c) => c instanceof Response ? H(c.blob()) : v(void 0)
    ),
    fe((c) => {
      if (!c)
        return v(null);
      const d = t?.defaultView?.URL.createObjectURL(c) ?? "";
      if (e.hasAttribute("src"))
        e.setAttribute("src", d);
      else if (e.hasAttribute("href") && (e.setAttribute("href", d), t?.defaultView && e instanceof t.defaultView.HTMLLinkElement))
        return new se((u) => {
          e.onload = async () => {
            try {
              e.sheet && await ui(
                t,
                e,
                n,
                i,
                o
              ), u.next(), u.complete();
            } catch (l) {
              u.error(l);
            }
          }, e.onerror = u.error;
        });
      return v(null);
    })
  );
}, li = ({
  settings: t,
  item: e,
  context: n
}) => (i) => i.pipe(
  P((o) => {
    const r = jt(
      o.contentDocument
    ), s = wn(e.href), a = r.map(
      (c) => di(
        o.contentDocument,
        c,
        s,
        n,
        t
      )
    );
    return a.length === 0 ? v(o) : O(a).pipe(y(() => o));
  })
), pi = (t) => {
  Hn(t?.contentDocument);
}, gi = async (t, e) => {
  if (typeof t == "string")
    return new Blob([t], { type: "text/html" });
  const n = Pn(t.headers.get("Content-Type") || "") || et(e.href);
  if (["image/jpg", "image/jpeg", "image/png", "image/webp"].some(
    (i) => i === n
  )) {
    const i = await t.blob(), o = URL.createObjectURL(i), r = await createImageBitmap(i), { width: s, height: a } = { width: r.width, height: r.height };
    r.close();
    const c = Math.random().toString(36).substring(2, 15).toString();
    return new Blob(
      [
        `
        <html>
          <head>
            ${e.renditionLayout === "pre-paginated" ? `<meta name="viewport" content="width=${s}, height=${a}">` : ""}
          </head>
          <body style="margin: 0px;" tab-index="-1;">
             <img
               id="${c}"
               src="${o}"
               style="max-width:100%;height:100%;object-fit:contain;display:block;"
             >
             <script>
               const imageElement = document.getElementById("${c}")
               if (imageElement instanceof HTMLImageElement) {
                 imageElement.addEventListener(
                   "load",
                   () => {
                     URL.revokeObjectURL("${o}")
                   },
                   { once: true },
                 )
               }
             <\/script>
          </body>
        </html>
          `
      ],
      { type: "text/html" }
    );
  }
  if (["text/plain"].some((i) => i === n)) {
    const i = await t.text();
    return new Blob(
      [
        `
        <!DOCTYPE html>
        <html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops" xml:lang="en" lang="en">
          <head>
            <style>
              pre {
                white-space: pre;
                white-space: pre-wrap;
                word-wrap: break-word;
              }
            </style>
          </head>
          <body>
            <pre>${i}</pre>
          </body>
        </html>
      `
      ],
      { type: "text/html" }
    );
  }
  return await t.blob();
}, hi = [
  ".txt",
  ".jpg",
  ".jpeg",
  ".png",
  ".webp"
], mi = [
  "text/plain",
  "image/jpg",
  "image/jpeg",
  "image/png",
  "image/webp"
], qe = (t) => {
  const e = t?.getAttribute("src");
  e?.startsWith("blob:") && URL.revokeObjectURL(e);
}, fi = ({
  item: t,
  resourcesHandler: e
}) => {
  const n = (i, o) => {
    qe(i);
    const r = URL.createObjectURL(o);
    i.setAttribute("src", r);
  };
  return (i) => i.pipe(
    P((o) => H(e.getResource()).pipe(
      P((r) => {
        const s = hi.some(
          (c) => t.href.endsWith(c)
        ) || t.mediaType && mi.includes(t.mediaType);
        return r instanceof URL && t.href.startsWith(window.location.origin) && !s ? (qe(o), o?.setAttribute("src", r.href), v(o)) : (r instanceof URL ? H(e.fetchResource()) : v(r)).pipe(
          P((c) => {
            if (!(c instanceof Response))
              throw new Error("Invalid resource");
            return gi(c, t);
          }),
          x((c) => {
            n(o, c);
          }),
          y(() => o),
          Te((c) => (E.error(
            `Error while trying to fetch or load resource for item ${t.id}`,
            r
          ), E.error(c), v(o)))
        );
      })
    ))
  );
}, Si = () => {
  const t = document.createElement("iframe");
  return t.frameBorder = "no", t.tabIndex = 0, t.setAttribute(
    "sandbox",
    `
  allow-same-origin 
  allow-scripts 
  allow-top-navigation-to-custom-protocols
`
  ), t.style.cssText = `
  overflow: hidden;
  background-color: transparent;
  border: 0px none transparent;
  padding: 0px;
`, t.setAttribute("role", "main"), t;
}, yi = "body{margin:0}html,body{height:100%;width:100%}html[data-prose-reader-html-renderer-has-viewport-dimensions=false] body{display:flex}html[data-prose-reader-html-renderer-has-viewport-dimensions=false][data-prose-reader-html-renderer-spread-position=left] body{justify-content:flex-end}html[data-prose-reader-html-renderer-has-viewport-dimensions=false][data-prose-reader-html-renderer-spread-position=right] body{justify-content:flex-start}html[data-prose-reader-html-renderer-has-viewport-dimensions=false][data-prose-reader-html-renderer-spread-position=none] body{justify-content:center}:not(html.prose-reader-html-renderer-touch-enabled),:not(html.prose-reader-html-renderer-touch-enabled) body{touch-action:none}img{-webkit-user-select:none;user-select:none;-webkit-user-drag:none;-khtml-user-drag:none;-moz-user-drag:none;-o-user-drag:none;display:flex}html[data-prose-reader-html-renderer-has-viewport-dimensions=false] img{max-width:100%;height:100%;object-fit:contain}", Xt = ({
  pageHeight: t,
  pageWidth: e,
  frameElement: n
}) => {
  const i = rt(n);
  let o = i.width, r = i.height;
  !o && i.widthToken === "device-width" && (o = e), !r && i.heightToken === "device-height" && (r = t), !r && (o || i.widthToken === "device-width") && (r = t), !o && (r || i.heightToken === "device-height") && (o = e);
  const s = i.hasViewport && o && r ? {
    ...i,
    width: o,
    height: r
  } : i;
  if (n?.contentDocument && n.contentWindow && s) {
    const a = e / (s.width ?? 1);
    return {
      computedScale: Math.min(
        a,
        t / (s.height ?? 1)
      ),
      computedWidthScale: a,
      viewportDimensions: s
    };
  }
}, ft = /* @__PURE__ */ new WeakMap(), J = (t, e, n) => {
  t.style.getPropertyValue(e) !== n && t.style.setProperty(e, n);
}, he = (t, e) => {
  t.style.getPropertyValue(e) !== "" && t.style.removeProperty(e);
}, St = (t, e) => {
  const n = `${e.width}px`, i = `${e.height}px`;
  t.style.width !== n && (t.style.width = n), t.style.height !== i && (t.style.height = i);
}, Ii = (t, e, n) => {
  const i = t.contentDocument?.documentElement;
  i && i.getAttribute(e) !== n && i.setAttribute(e, n);
}, bi = ({
  hasViewportDimensions: t,
  viewportWidth: e,
  viewportHeight: n,
  computedScale: i,
  spreadPosition: o,
  blankPagePosition: r,
  isRTL: s,
  pageWidth: a,
  pageHeight: c
}) => [
  t ? "1" : "0",
  e ?? "",
  n ?? "",
  i,
  o,
  r,
  s ? "1" : "0",
  a,
  c
].join("|"), vi = ({
  frameElement: t,
  viewportDimensions: e,
  computedScale: n,
  spreadPosition: i,
  blankPagePosition: o,
  isRTL: r,
  pageWidth: s
}) => {
  if (e) {
    J(t, "position", "absolute"), J(t, "top", "50%"), i === "left" ? (J(t, "right", "0"), he(t, "left")) : o === "before" && r ? (J(t, "right", "50%"), he(t, "left")) : i === "right" ? (J(t, "left", "0"), he(t, "right")) : (J(
      t,
      "left",
      o === "before" ? r ? "25%" : "75%" : o === "after" ? r ? "75%" : "25%" : "50%"
    ), he(t, "right"));
    const a = i !== "none" ? "0" : "-50%", c = i === "right" && o !== "before" ? "left" : i === "left" || o === "before" && r ? "right" : "center";
    J(
      t,
      "transform",
      `translate(${a}, -50%) scale(${n})`
    ), J(
      t,
      "transform-origin",
      `${c} center`
    );
    return;
  }
  o === "before" ? r ? J(t, "margin-right", `${s}px`) : J(t, "margin-left", `${s}px`) : (he(t, "margin-left"), he(t, "margin-right"));
}, wi = ({
  minPageSpread: t,
  blankPagePosition: e,
  spreadPosition: n,
  pageHeight: i,
  pageWidth: o,
  frameElement: r,
  isRTL: s
}) => {
  const a = t * o;
  if (r?.contentDocument && r?.contentWindow) {
    const { viewportDimensions: c, computedScale: d = 1 } = Xt({ frameElement: r, pageHeight: i, pageWidth: o }) ?? {}, u = !!c, l = o, g = i, p = bi({
      hasViewportDimensions: u,
      viewportWidth: c?.width,
      viewportHeight: c?.height,
      computedScale: d,
      spreadPosition: n,
      blankPagePosition: e,
      isRTL: s,
      pageWidth: o,
      pageHeight: i
    });
    return ft.get(r) === p ? { width: a, height: g } : (ft.set(r, p), Ii(
      r,
      "data-prose-reader-html-renderer-has-viewport-dimensions",
      u.toString()
    ), c ? St(r, {
      width: c.width ?? 1,
      height: c.height ?? 1
    }) : St(r, {
      width: l,
      height: g
    }), vi({
      frameElement: r,
      viewportDimensions: c,
      computedScale: d,
      spreadPosition: n,
      blankPagePosition: e,
      isRTL: s,
      pageWidth: o
    }), { width: a, height: g });
  }
  return { width: a, height: i };
}, Pi = () => `
      body {
        margin: 0;
      }
      html {
        width: 100%;
        height: 100%;
      }
      body {
        width: 100%;
        height: 100%;
        margin: 0;
      }
      
      html, body {
        touch-action: none;
      }
    `, xi = ({
  isScrollable: t,
  enableTouch: e
}) => `
      
      html, body {
        width: 100%;
        margin: 0;
        padding: 0;
        ${e ? "" : `
          touch-action: none
        `}
      }
      ${t ? `
        img {
          height: auto !important;
          margin: 0;
          padding: 0;
          box-sizing: border-box;
          
          width: 100%;
          
          display: block;
        }
      ` : ""}
    `, $i = ({
  width: t,
  columnHeight: e,
  columnWidth: n
}) => `
      parsererror {
        display: none !important;
      }
      
      html, body {
        margin: 0;
        padding: 0 !important;
        -max-width: ${n}px !important;
      }
      
      body {
        padding: 0 !important;
        width: ${t}px !important;
        height: ${e}px !important;
        overflow-y: hidden;
        column-gap: 0px !important;
        column-width: ${n}px !important;
        column-fill: auto !important;
        word-wrap: break-word;
        box-sizing: border-box;
      }
      body {
        margin: 0;
      }
      body:focus-visible {
        
        outline: none;
      }
      
      html, body {
        touch-action: none;
      }
      
      * {
        -max-width: ${n}px !important;
      }
      
      img, video, audio, object, svg {
        max-width: 100%;
        -max-width: ${n}px !important;
        -max-height: ${e}px !important;
        -pointer-events: none;
        -webkit-column-break-inside: avoid;
        page-break-inside: avoid;
        break-inside: avoid;
      }
      figure {
        d-max-width: ${n}px !important;
      }
      img {
        object-fit: contain;
        break-inside: avoid;
        box-sizing: border-box;
        d-max-width: ${n}px !important;
      }
      
      table {
        max-width: ${n}px !important;
        table-layout: fixed;
      }
      td {
        max-width: ${n}px;
      }
    `, Ti = ({
  isUsingVerticalWriting: t,
  minimumWidth: e,
  pageHeight: n,
  pageWidth: i
}) => {
  let s = i - 0;
  const a = n - 0;
  let c = i - 0;
  return t && (c = e - 0, s = a), {
    columnHeight: a,
    columnWidth: s,
    width: c
  };
}, xe = (t, e) => {
  t.style.width = `${e.width}px`, t.style.height = `${e.height}px`;
}, Fi = ({
  pageHeight: t,
  pageWidth: e,
  frameElement: n,
  manifest: i,
  minPageSpread: o,
  isRTL: r,
  blankPagePosition: s,
  isImageType: a,
  enableTouch: c,
  isUsingVerticalWriting: d
}) => {
  const u = o * e, l = i?.renditionLayout === "reflowable" && i?.renditionFlow === "scrolled-continuous", g = l ? Math.min(400, t) : t;
  n?.style.setProperty("width", `${e}px`), l ? n?.style.removeProperty("height") : n?.style.setProperty("height", `${g}px`);
  const { viewportDimensions: p, computedScale: h = 1 } = Xt({
    frameElement: n,
    pageHeight: g,
    pageWidth: e
  }) ?? {}, m = i?.renditionLayout === "pre-paginated";
  if (n?.contentDocument && n?.contentWindow && n.contentDocument.body) {
    let f = e, S = g;
    if (p?.hasViewport)
      re(
        n,
        "prose-reader-html-renderer-framce-css",
        Pi()
      ), xe(n, {
        width: p.width ?? 1,
        height: p.height ?? 1
      }), n?.style.setProperty("position", "absolute"), n?.style.setProperty("top", "50%"), n?.style.setProperty(
        "left",
        s === "before" ? r ? "25%" : "75%" : s === "after" ? r ? "75%" : "25%" : "50%"
      ), n?.style.setProperty(
        "transform",
        `translate(-50%, -50%) scale(${h})`
      ), n?.style.setProperty("transform-origin", "center center");
    else {
      const b = a ? xi({
        isScrollable: i?.renditionFlow === "scrolled-continuous",
        enableTouch: c
      }) : $i(
        Ti({
          isUsingVerticalWriting: d,
          minimumWidth: u,
          pageHeight: g,
          pageWidth: e
        })
      );
      if (re(n, "prose-reader-css", b, !0), d)
        S = Math.ceil(
          n.contentDocument.documentElement.scrollHeight / g
        ) * g, xe(n, {
          width: u,
          height: S
        });
      else if (i?.renditionFlow === "scrolled-continuous")
        S = n.contentDocument.body.scrollHeight, xe(n, {
          width: u,
          height: S
        });
      else {
        const F = Math.ceil(
          n.contentDocument.documentElement.scrollWidth / e
        );
        m ? f = e : f = F * e, xe(n, {
          width: f,
          height: S
        });
      }
    }
    return f % u === 0 ? n?.style.setProperty("margin-left", "0px") : (f = f + e, r && !d && n?.style.setProperty("margin-left", `${e}px`)), { width: f, height: S };
  }
};
class Ci extends Ee {
  constructor() {
    super(...arguments), this.isPrePaginated = () => this.item.renditionLayout === "pre-paginated" || !this.item.renditionLayout && this.context.manifest?.renditionLayout === "pre-paginated", this.isImageType = () => !!(this.item.mediaType ?? et(this.item.href))?.startsWith("image/");
  }
  onCreateDocument() {
    const e = Si();
    return this.setDocumentContainer(e), v(e);
  }
  onLoadDocument() {
    const e = this.getFrameElement();
    if (!e) throw new Error("invalid frame");
    return v(e).pipe(
      fi({
        item: this.item,
        resourcesHandler: this.resourcesHandler,
        settings: this.settings
      }),
      /**
       * Preload iframes only when the viewport is stable to avoid competing with
       * navigation/transition work on the main thread.
       */
      Re(this.context.bridgeEvent.viewportFree$),
      x((n) => {
        this.prepareFrameForAttach(n), this.attach();
      }),
      qn,
      x((n) => {
        this.isPrePaginated() ? re(n, "prose-reader-css", yi) : this.isImageType();
      }),
      li({
        context: this.context,
        item: this.item,
        settings: this.settings
      }),
      Gn,
      x((n) => {
        this.revealFrameAfterReady(n);
      })
    );
  }
  onUnload() {
    const e = this.getFrameElement();
    return pi(e), qe(e), this.detach(), X;
  }
  onLayout({
    minPageSpread: e,
    blankPagePosition: n,
    spreadPosition: i
  }) {
    const { width: o, height: r } = this.viewport.pageSize, s = this.getFrameElement();
    if (!s) return v(void 0);
    const a = this.settings.values.computedPageTurnMode === "scrollable";
    if (s.contentDocument?.documentElement && (s.contentDocument?.documentElement.setAttribute(
      "data-prose-reader-html-renderer-spread-position",
      i
    ), a ? s.contentDocument?.documentElement.classList.add(
      "prose-reader-html-renderer-touch-enabled"
    ) : s.contentDocument?.documentElement.classList.remove(
      "prose-reader-html-renderer-touch-enabled"
    )), this.isPrePaginated()) {
      const l = wi({
        blankPagePosition: n,
        frameElement: s,
        isRTL: this.context.isRTL(),
        minPageSpread: e,
        pageHeight: r,
        pageWidth: o,
        spreadPosition: i
      });
      return v(l);
    }
    const c = !!this.writingMode?.startsWith("vertical"), d = this.isImageType(), u = Fi({
      pageHeight: r,
      pageWidth: o,
      frameElement: s,
      manifest: this.context.manifest,
      blankPagePosition: n,
      isUsingVerticalWriting: c,
      isRTL: this.context.isRTL(),
      minPageSpread: e,
      isImageType: d,
      /**
       * When we have scrollable content, we use "native" touch event from the frame instead of
       * our own gestures.
       * @todo move this into scroll navigator
       */
      enableTouch: this.settings.values.computedPageTurnMode === "scrollable"
    });
    return v(u);
  }
  onRenderHeadless() {
    return H(this.resourcesHandler.fetchResource()).pipe(
      P((e) => {
        if (e instanceof Response) {
          const n = e.headers.get("content-type") ?? "";
          if ([
            "application/xhtml+xml",
            "application/xml",
            "text/html",
            "text/xml"
          ].includes(n))
            return H(e.text()).pipe(
              y((o) => new DOMParser().parseFromString(
                o,
                n
              ))
            );
        }
        return v(void 0);
      })
    );
  }
  getFrameElement() {
    const e = this.documentContainer;
    if (e instanceof HTMLIFrameElement)
      return e;
  }
  prepareFrameForAttach(e) {
    e.style.setProperty("visibility", "hidden"), e.style.setProperty("pointer-events", "none"), e.style.setProperty("contain", "layout paint size"), e.style.setProperty("display", "block");
  }
  revealFrameAfterReady(e) {
    e.style.removeProperty("visibility"), e.style.removeProperty("pointer-events"), e.style.removeProperty("contain");
  }
  // @todo optimize
  getComputedStyleAfterLoad() {
    const e = this.getFrameElement(), n = e?.contentDocument?.body;
    if (n)
      return e?.contentWindow?.getComputedStyle(n);
  }
  get writingMode() {
    if (!this.isImageType())
      return this.getComputedStyleAfterLoad()?.writingMode;
  }
  get readingDirection() {
    if (this.writingMode === "vertical-rl")
      return "rtl";
    switch (this.getComputedStyleAfterLoad()?.direction) {
      case "ltr":
      case "inherit":
      case "initial":
        return "ltr";
      case "rtl":
        return "rtl";
      default:
        return;
    }
  }
  getDocumentFrame() {
    return this.getFrameElement();
  }
}
const Li = (t) => (e) => {
  const n = t({
    ...e,
    getRenderer(o) {
      return e.getRenderer?.(o) ?? ((s) => new Ci(s));
    }
  }), i = ai(n);
  return i.pipe(T(n.$.destroy$)).subscribe(), {
    ...n,
    links$: i
  };
};
function K(t) {
  return t != null;
}
class ut {
  constructor(e) {
    this.left = e.left, this.right = e.right, this.top = e.top, this.bottom = e.bottom, this.width = e.width, this.height = e.height, this.x = e.x, this.y = e.y;
  }
}
class N {
  constructor(e) {
    this.__symbol = /* @__PURE__ */ Symbol("SpineItemPosition"), this.x = e.x, this.y = e.y;
  }
}
class ke {
  constructor(e) {
    this.__symbol = /* @__PURE__ */ Symbol("SpineItemPagePosition"), this.x = e.x, this.y = e.y;
  }
}
class Mi extends ut {
  constructor() {
    super(...arguments), this.__symbol = "SpineItemPageLayout";
  }
}
class je extends ut {
  constructor() {
    super(...arguments), this.__symbol = "SpineItemSpineLayout";
  }
}
class Ei extends ut {
  constructor() {
    super(...arguments), this.__symbol = "SpineItemPageSpineLayout";
  }
}
class Yt {
  constructor(e) {
    this.x = e.x, this.y = e.y;
  }
}
class L extends Yt {
  constructor() {
    super(...arguments), this.__symbol = "SpinePosition";
  }
  static from(e) {
    return new L(e);
  }
}
class ce extends Yt {
  constructor() {
    super(...arguments), this.__symbol = "UnboundSpinePosition";
  }
  static from(e) {
    return new ce(e);
  }
}
const Ni = (t, e) => {
  const n = e.getBoundingClientRect(), { x: i, y: o } = t, { left: r, top: s } = n, a = n.width / e.offsetWidth, c = n.height / e.offsetHeight, d = i - r, u = o - s;
  return {
    x: d / a,
    y: u / c
  };
}, Ai = (t, e) => new ce(
  Ni(t, e)
), Di = (t) => ({
  getSpinePositionFromClientPosition: (e) => t.spine.element ? Ai(e, t.spine.element) : void 0
}), Ri = (t) => {
  let e;
  const n = t.context.watch("rootElement").pipe(
    B((c) => c ? new se(() => (e = c.ownerDocument.createElement("div"), e.style.cssText = `
        position: absolute;
        left: 0;
        top: 0;
        width: 100%;
        height: 100%;
        visibility: hidden;
      `, c.appendChild(e), () => {
      e?.remove(), e = void 0;
    })) : Se)
  ), i = (c) => gn(c, be).pipe(
    k(() => {
      e?.style.setProperty(
        "visibility",
        "hidden"
      );
    })
  ), o = t.context.bridgeEvent.viewportBusy$.pipe(
    k(() => {
      e?.style.setProperty(
        "visibility",
        "visible"
      );
    })
  ), r = i(t.viewportFree$).pipe(
    Cn(1)
  ), a = t.settings.values$.pipe(
    oe(() => t.settings.values.computedPageTurnMode),
    Pe()
  ).pipe(
    B(
      (c) => c === "controlled" ? o.pipe(B(() => r)) : i(v(void 0))
    ),
    Q(t.$.destroy$)
  );
  return w(n, a);
}, xs = "@prose-reader/core", $s = 0, Ts = 200, U = "prose-reader", Oi = `${U}-style`, Vi = `data-${U}-id`, ki = `${U}-viewport`, ji = `${U}-scroll-navigator`, Ui = `${U}-spine-item`, zi = `${U}-enhancer-loading`, qt = `${zi}-container`, Wi = ({
  container: t,
  item: e,
  viewport: n
}) => {
  const i = t.ownerDocument.createElement("div");
  i.classList.add(qt), i.style.cssText = `
      height: 100%;
      width: 100%;
      max-width: ${n.absoluteViewport.width}px;
      text-align: center;
      display: flex;
      justify-content: center;
      align-items: center;
      flex-direction: column;
      position: absolute;
      left: 0;
      top: 0;
      color: rgb(202, 202, 202);
      background-color: white;
      z-index: 1;
    `;
  const o = i.ownerDocument.createElement("div");
  o.innerText = "prose", o.style.cssText = `
      font-size: 4em;
    `;
  const r = i.ownerDocument.createElement("div");
  return r.setAttribute("data-details-element", "true"), r.innerText = `loading ${e.id}`, r.style.cssText = `
      font-size: 1.2em;
      text-overflow: ellipsis;
      white-space: nowrap;
      overflow: hidden;
      max-width: 300px;
      width: 80%;
    `, i.appendChild(o), i.appendChild(r), i;
}, _i = (t) => t.spineItemsManager.items$.pipe(
  B(
    (e) => w(
      ...e.map((n) => {
        n.containerElement.style.zIndex = "0";
        const i = n.containerElement.querySelector(
          `.${qt}`
        ), o = i instanceof HTMLElement ? i : Wi({
          container: n.containerElement,
          item: n.item,
          viewport: t.viewport
        });
        return n.containerElement.appendChild(o), w(
          n.pipe(
            k((r) => {
              if (o.style.setProperty(
                "visibility",
                r.isReady ? "hidden" : "visible"
              ), o.style.setProperty(
                "z-index",
                r.isReady ? "0" : "1"
              ), r.isError) {
                const s = o.querySelector(
                  "[data-details-element]"
                );
                s instanceof HTMLElement && (s.innerText = r.error?.toString() ?? "Unknown error");
              }
            })
          ),
          O([t.spine.layout$, t.theme.$.theme$]).pipe(
            k(([, r]) => {
              const s = t.viewport.absoluteViewport.width;
              o.style.setProperty(
                "max-width",
                `${s}px`
              ), o.style.setProperty(
                "color",
                r === "sepia" ? "#939393" : "rgb(202, 202, 202)"
              );
            })
          )
        );
      })
    )
  )
), Hi = (t) => {
  t.hookManager.register("item.onDocumentLoad", ({ itemId: e }) => {
    t.spineItemsManager.get(e)?.renderer.getDocumentFrame()?.setAttribute("scrolling", "no");
  });
}, Bi = (t) => {
  t.hookManager.register(
    "item.onAfterLayout",
    ({ item: e, blankPagePosition: n, minimumWidth: i }) => {
      const o = t.spineItemsManager.get(e.id), r = o?.renderer.getDocumentFrame();
      if (o?.renditionLayout !== "reflowable" || !(r instanceof HTMLIFrameElement))
        return;
      const { hasViewport: s } = rt(r), { width: a } = t.viewport.pageSize, c = o?.renderer.getDocumentFrame();
      if (s) {
        const d = a < i;
        n === "none" && d && c instanceof HTMLIFrameElement && c?.style.setProperty(
          "left",
          t.context.isRTL() ? "75%" : "25%"
        );
      }
    }
  );
}, Xi = (t) => t.spineItemsObserver.states$.pipe(
  x(({ item: e, isReady: n, isDirty: i }) => {
    e.containerElement.dataset.isDirty = i.toString(), e.containerElement.dataset.isReady = n.toString();
  })
);
let Yi = class extends Ht {
  computeOutputSettings(e) {
    return e;
  }
  hasSettingsChanged(e) {
    return !V(this.outputSettings, e);
  }
  getCleanedParentInputSettings(e) {
    const {
      layoutAutoResize: n,
      pageHorizontalMargin: i,
      pageVerticalMargin: o,
      layoutLayerTransition: r,
      ...s
    } = e;
    return s;
  }
  getDefaultSettings() {
    return {
      layoutAutoResize: "container",
      pageHorizontalMargin: 24,
      pageVerticalMargin: 24,
      layoutLayerTransition: !0
    };
  }
};
const qi = (t) => O([
  t.viewport.watch(["width", "height"]),
  t.context.watch("manifest")
]).pipe(
  x(([{ width: e, height: n }, i]) => {
    const o = e > n;
    return !o && i?.renditionSpread === "portrait" ? t.settings.update({ spreadMode: !0 }) : o && (i?.renditionSpread === void 0 || i?.renditionSpread === "auto" || i?.renditionSpread === "landscape" || i?.renditionSpread === "both") ? t.settings.update({ spreadMode: !0 }) : t.settings.update({ spreadMode: !1 });
  })
), Gi = (t) => (e) => {
  const {
    pageHorizontalMargin: n,
    pageVerticalMargin: i,
    layoutAutoResize: o,
    layoutLayerTransition: r
  } = e, s = t(e), a = new Yi(
    {
      pageHorizontalMargin: n,
      pageVerticalMargin: i,
      layoutAutoResize: o,
      layoutLayerTransition: r
    },
    s.settings
  );
  s.hookManager.register("onViewportOffsetAdjust", () => {
    for (const m of s.spineItemsManager.items) {
      const f = m.renderer.getDocumentFrame();
      if (f) {
        f.getBoundingClientRect().left;
        break;
      }
    }
  }), s.hookManager.register("item.onBeforeLayout", ({ item: m }) => {
    const f = s.spineItemsManager.get(m.id), I = !!(m.mediaType ?? et(m.href))?.startsWith("image/"), { pageHorizontalMargin: b = 0, pageVerticalMargin: F = 0 } = a.values, $ = s.viewport.pageSize;
    if (f?.renditionLayout === "reflowable" && !I) {
      let C = $.width - b * 2;
      const W = $.height - F * 2;
      let _ = $.width - b * 2, ae = b * 2;
      f.isUsingVerticalWriting() && (_ = $.width - b * 2, C = W, ae = F * 2);
      const ue = f?.renderer.getDocumentFrame();
      ue && re(
        ue,
        "prose-layout-enhancer-css",
        `
              body {
                width: ${_}px !important;
                margin: ${F}px ${b}px !important;
                column-gap: ${ae}px !important;
                column-width: ${C}px !important;
                height: ${W}px !important;
              }
              img, video, audio, object, svg {
                -max-width: ${C}px !important;
                -max-height: ${W}px !important;
              }
              table {
                max-width: ${C}px !important;
              }
              td {
                max-width: ${C}px;
              }
            `
      );
    }
  }), Bi(s), Hi(s), s.hookManager.register(
    "item.onDocumentCreated",
    ({ documentContainer: m }) => {
      m.style.opacity = "0", a.values.layoutLayerTransition && (m.style.transition = "opacity 800ms");
    }
  ), s.hookManager.register("item.onBeforeLayout", ({ item: m }) => {
    const S = s.spineItemsManager.get(m.id)?.renderer.documentContainer;
    s.settings.values.computedPageTurnMode !== "scrollable" && S?.setAttribute("tab-index", "0");
  });
  const c = s.spineItemsObserver.states$.pipe(
    me(({ isReady: m }) => m),
    k(({ item: m }) => {
      const f = m.renderer.documentContainer;
      f && (f.style.opacity = "1");
    })
  ), d = a.values$.pipe(
    me(({ layoutAutoResize: m }) => m === "container"),
    B(() => s.context.watch("rootElement")),
    me(K),
    B((m) => at(m)),
    Ln(100),
    me(K),
    k(() => {
      s?.layout();
    })
  ), u = Ri(s);
  a.watch(["pageHorizontalMargin", "pageVerticalMargin"]).pipe(
    Ot(1),
    k(() => {
      s.layout();
    }),
    Q(s.$.destroy$)
  ).subscribe();
  const l = s.spine.pages.pipe(
    tt({ refCount: !0, bufferSize: 1 })
  ), g = Xi(s), p = qi(s), h = _i(s);
  return w(
    c,
    u,
    d,
    l,
    g,
    p,
    h
  ).pipe(Q(s.$.destroy$)).subscribe(), {
    ...s,
    destroy: () => {
      a.destroy(), s.destroy();
    },
    settings: a,
    layout$: s.spine.layout$,
    layoutInfo$: l,
    coordinates: Di(s)
  };
}, Zi = (t) => (e) => {
  const n = t(e), i = new IntersectionObserver(
    (s) => {
      s.forEach((a) => {
        const c = a.target, d = Array.from(
          c.contentDocument?.body.getElementsByTagName("audio") || []
        );
        a.isIntersecting ? d.forEach((u) => {
          u.hasAttribute("autoplay") && u.paused && u.readyState >= 2 && u.play().catch(console.error);
        }) : d.forEach((u) => {
          u.pause(), u.currentTime = 0;
        });
      });
    },
    {
      threshold: 0.01
    }
  ), o = new IntersectionObserver(
    (s) => {
      s.forEach((a) => {
        if (a.target.tagName === "video") {
          const c = a.target;
          a.isIntersecting ? c.hasAttribute("autoplay") && c.paused && c.readyState >= 2 && c.play().catch(console.error) : (c.pause(), c.currentTime = 0);
        }
      });
    },
    {
      threshold: 0.5
    }
  );
  return n.hookManager.register(
    "item.onDocumentLoad",
    ({ destroy: s, itemId: a }) => {
      const c = n.spineItemsManager.get(a)?.renderer.getDocumentFrame();
      if (!c) return;
      i.observe(c);
      const d = c.contentDocument?.body.getElementsByTagName("video"), u = Array.from(d || []).map((l) => (o.observe(l), () => o.unobserve(l)));
      s(() => {
        i.unobserve(c), u.forEach((l) => {
          l();
        });
      });
    }
  ), {
    ...n,
    destroy: () => {
      i.disconnect(), o.disconnect(), n.destroy();
    }
  };
}, Ji = (t, e) => t.links$.pipe(
  x((n) => {
    if (!Bn(n.target, "a") || n.type !== "click") return;
    const i = new URL(n.target.href), o = `${i.origin}${i.pathname}`;
    t.context.manifest?.spineItems.some(
      (s) => s.href === o
    ) && e.goToUrl(i);
  })
), te = E.namespace("navigation"), Ki = ({
  position: t,
  spineItem: e,
  pageHeight: n,
  pageWidth: i,
  spineItemLocator: o
}) => {
  let r = new N({
    x: t.x - i,
    y: t.y
  });
  return e.isUsingVerticalWriting() && (r = new N({
    x: t.x,
    y: t.y + n
  })), o.getSpineItemClosestPositionFromUnsafePosition(
    r,
    e
  );
}, yt = ({
  position: t,
  navigationResolver: e,
  computedPageTurnDirection: n,
  spineItemsManager: i,
  spineLocator: o,
  viewport: r
}) => {
  const s = n, a = o.getSpineItemFromPosition(t) || i.get(0), c = t;
  if (!a)
    return c;
  const d = o.getSpineItemPositionFromSpinePosition(
    t,
    a
  ), u = Ki({
    position: d,
    spineItem: a,
    pageHeight: r.pageSize.height,
    pageWidth: r.pageSize.width,
    spineItemLocator: o.spineItemLocator
  });
  return e.arePositionsDifferent(
    u,
    d
  ) ? o.getSpinePositionFromSpineItemPosition({
    spineItemPosition: u,
    spineItem: a
  }) : e.fromUnboundSpinePosition(
    s === "horizontal" ? new L({
      x: t.x - r.pageSize.width,
      y: 0
    }) : new L({
      y: t.y - r.pageSize.height,
      x: 0
    })
  );
}, Qi = ({
  position: t,
  spineItem: e,
  context: n,
  navigationResolver: i,
  spineItemsManager: o,
  spineLocator: r,
  computedPageTurnDirection: s,
  viewport: a,
  settings: c
}) => {
  const d = yt({
    position: t,
    viewport: a,
    navigationResolver: i,
    computedPageTurnDirection: s,
    spineItemsManager: o,
    spineLocator: r
  });
  if (e?.isUsingVerticalWriting() && t.x === d.x)
    return i.getAdjustedPositionForSpread(d);
  if (c.values.computedSpreadMode) {
    if (e?.isUsingVerticalWriting() && t.x !== d.x)
      return i.getAdjustedPositionForSpread(
        i.fromUnboundSpinePosition(
          n.isRTL() ? { ...d, x: d.x + a.pageSize.width } : {
            ...d,
            x: d.x - a.pageSize.width
          }
        )
      );
    if (s === "vertical" && t.y !== d.y)
      return i.getAdjustedPositionForSpread(d);
    const u = yt({
      position: d,
      viewport: a,
      navigationResolver: i,
      computedPageTurnDirection: s,
      spineItemsManager: o,
      spineLocator: r
    });
    return i.getAdjustedPositionForSpread(u);
  }
  return i.getAdjustedPositionForSpread(d);
}, eo = ({
  position: t,
  spineItem: e,
  pageHeight: n,
  pageWidth: i,
  spineItemLocator: o
}) => {
  let r = new N({
    x: t.x + i,
    y: t.y
  });
  return e.isUsingVerticalWriting() && (r = new N({
    x: t.x,
    y: t.y - n
  })), o.getSpineItemClosestPositionFromUnsafePosition(
    r,
    e
  );
}, It = ({
  position: t,
  navigationResolver: e,
  computedPageTurnDirection: n,
  spineItemsManager: i,
  spineLocator: o,
  viewport: r
}) => {
  const s = n, a = o.getSpineItemFromPosition(t) || i.get(0), c = t;
  if (!a)
    return c;
  const d = o.getSpineItemPositionFromSpinePosition(
    t,
    a
  ), u = eo({
    position: d,
    spineItem: a,
    pageHeight: r.pageSize.height,
    pageWidth: r.pageSize.width,
    spineItemLocator: o.spineItemLocator
  });
  return e.arePositionsDifferent(
    u,
    d
  ) ? o.getSpinePositionFromSpineItemPosition({
    spineItemPosition: u,
    spineItem: a
  }) : e.fromUnboundSpinePosition(
    s === "horizontal" ? new ce({
      x: t.x + r.pageSize.width,
      y: 0
    }) : new ce({
      y: t.y + r.pageSize.height,
      x: 0
    })
  );
}, to = ({
  position: t,
  spineItem: e,
  context: n,
  navigationResolver: i,
  spineItemsManager: o,
  spineLocator: r,
  computedPageTurnDirection: s,
  viewport: a,
  settings: c
}) => {
  const d = It({
    position: t,
    viewport: a,
    navigationResolver: i,
    computedPageTurnDirection: s,
    spineItemsManager: o,
    spineLocator: r
  });
  if (e?.isUsingVerticalWriting() && t.x === d.x)
    return i.getAdjustedPositionForSpread(d);
  if (c.values.computedSpreadMode) {
    if (e?.isUsingVerticalWriting() && t.x !== d.x)
      return i.getAdjustedPositionForSpread(
        i.fromUnboundSpinePosition(
          n.isRTL() ? {
            ...d,
            x: d.x - a.pageSize.width
          } : {
            ...d,
            x: d.x + a.pageSize.width
          }
        )
      );
    if (s === "vertical" && t.y !== d.y)
      return i.getAdjustedPositionForSpread(d);
    const u = It({
      position: d,
      viewport: a,
      navigationResolver: i,
      computedPageTurnDirection: s,
      spineItemsManager: o,
      spineLocator: r
    });
    return i.getAdjustedPositionForSpread(u);
  }
  return i.getAdjustedPositionForSpread(d);
};
class no {
  constructor(e) {
    this.reader = e, this.movingLastDelta = { x: 0, y: 0 }, this.movingLastPosition = new L({ x: 0, y: 0 }), this.unlock = void 0;
  }
  turnRight() {
    return this.turnRightOrBottom();
  }
  turnLeft() {
    return this.turnLeftOrTop();
  }
  turnTop() {
    return this.turnLeftOrTop();
  }
  turnBottom() {
    return this.turnRightOrBottom();
  }
  turnRightOrBottom() {
    const e = this.reader.navigation.getNavigation(), n = this.reader.spineItemsManager.get(e.spineItem);
    if (!n) return;
    const i = to({
      context: this.reader.context,
      navigationResolver: this.reader.navigation.navigationResolver,
      position: e.position,
      computedPageTurnDirection: this.reader.settings.values.computedPageTurnDirection,
      spineItem: n,
      spineItemsManager: this.reader.spineItemsManager,
      spineLocator: this.reader.spine.locator,
      viewport: this.reader.viewport,
      settings: this.reader.settings
    });
    return this.reader.navigation.navigate({
      position: i
    });
  }
  turnLeftOrTop() {
    const e = this.reader.navigation.getNavigation(), n = this.reader.spineItemsManager.get(e.spineItem);
    if (!n) return;
    const i = Qi({
      context: this.reader.context,
      navigationResolver: this.reader.navigation.navigationResolver,
      position: e.position,
      computedPageTurnDirection: this.reader.settings.values.computedPageTurnDirection,
      spineItem: n,
      spineItemsManager: this.reader.spineItemsManager,
      spineLocator: this.reader.spine.locator,
      viewport: this.reader.viewport,
      settings: this.reader.settings
    });
    return this.reader.navigation.navigate({
      position: i
    });
  }
  goToCfi(e, n = { animate: !0 }) {
    return this.reader.navigation.navigate({
      animation: n.animate ? "turn" : !1,
      cfi: e
    });
  }
  goToSpineItem({
    indexOrId: e,
    ...n
  }) {
    const i = this.reader.spineItemsManager.get(e);
    if (i === void 0) {
      te.warn(
        "goToSpineItem",
        `Ignore navigation to ${e} since the item does not exist`
      );
      return;
    }
    this.reader.navigation.navigate({
      spineItem: i.index,
      ...n
    });
  }
  goToNextSpineItem() {
    const { endIndex: e = 0 } = this.reader.spine.locator.getVisibleSpineItemsFromPosition({
      position: this.reader.navigation.getNavigation().position,
      threshold: { type: "percentage", value: 0.5 }
    }) || {};
    this.goToSpineItem({ indexOrId: e + 1 });
  }
  goToPreviousSpineItem() {
    const { beginIndex: e = 0 } = this.reader.spine.locator.getVisibleSpineItemsFromPosition({
      position: this.reader.navigation.getNavigation().position,
      threshold: { type: "percentage", value: 0.5 }
    }) ?? {};
    this.goToSpineItem({ indexOrId: e - 1 });
  }
  goToUrl(e) {
    this.reader.navigation.navigate({
      url: e,
      animation: !1
    });
  }
  goToRightSpineItem() {
    if (this.reader.settings.values.computedPageTurnDirection === "vertical") {
      te.warn(
        "You cannot call this navigation method on vertical direction"
      );
      return;
    }
    return this.reader.context.isRTL() ? this.goToPreviousSpineItem() : this.goToNextSpineItem();
  }
  goToRightOrBottomSpineItem() {
    return this.reader.settings.values.computedPageTurnDirection === "vertical" ? this.goToBottomSpineItem() : this.goToRightSpineItem();
  }
  goToLeftOrTopSpineItem() {
    return this.reader.settings.values.computedPageTurnDirection === "vertical" ? this.goToTopSpineItem() : this.goToLeftSpineItem();
  }
  goToLeftSpineItem() {
    if (this.reader.settings.values.computedPageTurnDirection === "vertical") {
      te.warn(
        "You cannot call this navigation method on vertical direction"
      );
      return;
    }
    return this.reader.context.isRTL() ? this.goToNextSpineItem() : this.goToPreviousSpineItem();
  }
  goToTopSpineItem() {
    if (this.reader.settings.values.computedPageTurnDirection === "horizontal") {
      te.warn(
        "You cannot call this navigation method on horizontal direction"
      );
      return;
    }
    return this.goToPreviousSpineItem();
  }
  goToBottomSpineItem() {
    if (this.reader.settings.values.computedPageTurnDirection === "horizontal") {
      te.warn(
        "You cannot call this navigation method on horizontal direction"
      );
      return;
    }
    return this.goToNextSpineItem();
  }
  goToPageOfSpineItem({
    pageIndex: e,
    spineItemId: n,
    ...i
  }) {
    const o = this.reader.navigation.navigationResolver.getNavigationForSpineItemPage({
      pageIndex: e,
      spineItemId: n
    });
    te.debug(".goToPageOfSpineItem()", {
      pageIndex: e,
      spineItemId: n,
      ...i,
      position: o
    }), this.reader.navigation.navigate({
      position: o,
      ...i
    });
  }
  goToAbsolutePageIndex({
    absolutePageIndex: e,
    ...n
  }) {
    const i = this.reader.spine.pages.fromAbsolutePageIndex(e);
    if (te.debug(".goToAbsolutePageIndex()", {
      absolutePageIndex: e,
      ...n,
      foundInfo: i
    }), i) {
      const o = this.reader.navigation.navigationResolver.getNavigationForSpineItemPage(
        {
          pageIndex: i.pageIndex,
          spineItemId: i.itemIndex
        }
      );
      return this.reader.navigation.navigate({
        position: o,
        ...n
      });
    }
  }
}
class io extends ee {
  constructor(e) {
    super({
      lastDelta: { x: 0, y: 0 },
      lastPosition: new L({ x: 0, y: 0 }),
      lastStartPosition: new L({ x: 0, y: 0 }),
      isStarted: !1
    }), this.reader = e, this.unlock = void 0;
  }
  start(e) {
    this.unlock?.(), this.unlock = this.reader.navigation.lock(), this.mergeCompare({
      isStarted: !0,
      lastDelta: e,
      lastStartPosition: this.reader.navigation.controlledNavigationController.viewportPosition,
      lastPosition: this.reader.navigation.controlledNavigationController.viewportPosition
    });
  }
  stop(e) {
    this.panMoveTo(e), this.mergeCompare({
      isStarted: !1
    }), this.unlock?.();
  }
  panMoveTo(e) {
    if (this.reader.settings.values.computedPageTurnMode === "scrollable") {
      te.warn(
        "pan control is not available on free page turn mode"
      );
      return;
    }
    if (!this.value.isStarted) {
      te.error("Pan navigator is not started");
      return;
    }
    const n = this.reader.settings.values.computedPageTurnDirection;
    let i = this.reader.navigation.getNavigation().position;
    if (e) {
      const o = this.reader.viewport.absoluteViewport.width / this.reader.viewport.relativeViewport.width, r = Math.floor(e.x) - (this.value.lastDelta?.x || 0), s = Math.floor(e.y) - (this.value.lastDelta?.y || 0), a = Math.floor(
        n === "horizontal" ? this.value.lastPosition.x - r / o : 0
      ), c = Math.floor(
        n === "horizontal" ? 0 : this.value.lastPosition.y - s / o
      );
      i = new L({
        x: a,
        y: c
      }), this.mergeCompare({
        lastDelta: e
      });
    } else
      i = this.value.lastPosition;
    this.mergeCompare({
      lastPosition: i
    }), this.reader.navigation.navigate({
      position: i,
      animation: !1
    });
  }
}
class q {
  constructor() {
    this.isDestroyed = !1, this.destroySubject = new R(), this.destroy$ = this.destroySubject.asObservable();
  }
  destroy() {
    this.isDestroyed || (this.isDestroyed = !0, this.destroySubject.next(), this.destroySubject.complete());
  }
}
const oo = 500;
class ro extends q {
  constructor(e, n) {
    super(), this.scrollNavigationController = e, this.locker = n, this.navigationSubject = new R(), this.navigation$ = this.navigationSubject.asObservable();
    const i = this.scrollNavigationController.userScroll$.pipe(
      hn((o) => {
        const r = this.locker.lock();
        return w(
          this.scrollNavigationController.userScroll$,
          v(o)
        ).pipe(
          De(
            oo,
            be
          ),
          j(),
          x(() => {
            const s = this.scrollNavigationController.fromScrollPosition(
              this.scrollNavigationController.scrollPosition
            );
            this.navigationSubject.next({
              animation: !1,
              type: "scroll",
              position: s
            });
          }),
          ye(() => {
            r();
          })
        );
      })
    );
    w(i).pipe(T(this.destroy$)).subscribe();
  }
}
const so = (t) => t.pagination.state$.pipe(
  z(t.context.manifest$, t.settings.values$),
  y(
    ([
      e,
      { spineItems: n, readingDirection: i },
      { computedPageTurnDirection: o }
    ]) => {
      const r = n.length ?? 0, s = e.beginSpineItemIndex === 0, a = e.endSpineItemIndex === Math.max(r - 1, 0), c = e.endSpineItemIndex === Math.max(r - 1, 0), d = e.beginSpineItemIndex === 0, u = e.beginPageIndexInSpineItem === 0, l = e.endPageIndexInSpineItem === e.endNumberOfPagesInSpineItem - 1;
      return {
        canTurnLeft: o === "vertical" ? !1 : !u,
        canTurnRight: o === "vertical" ? !1 : !l,
        canGoTopSpineItem: o === "vertical" && !s,
        canGoBottomSpineItem: o === "vertical" && !a,
        canGoLeftSpineItem: o !== "vertical" && (i === "ltr" && !s || i === "rtl" && !c),
        canGoRightSpineItem: o !== "vertical" && (i === "ltr" && !a || i === "rtl" && !d)
      };
    }
  ),
  D(V)
), ao = ({ reader: t, duration: e }) => (n) => {
  let i;
  const o = () => {
    i?.(), i = void 0;
  };
  return n.pipe(
    x(() => {
      i || (i = t?.navigation.lock());
    }),
    mn(e, void 0, {
      trailing: !0,
      leading: !0
    }),
    x(o),
    ye(o)
  );
}, co = (t) => (e) => {
  const n = t(e), i = so(n), o = new no(n), r = new io(n), s = new ro(
    n.navigation.scrollNavigationController,
    n.navigation.locker
  ), a = Ji(n, o), c = s.navigation$.pipe(
    x((l) => {
      n.navigation.navigate(l);
    })
  );
  return w(a, c).pipe(T(n.$.destroy$)).subscribe(), {
    ...n,
    load: (l) => {
      const { cfi: g, ...p } = l;
      n.load(p), g && o.goToCfi(g, { animate: !1 });
    },
    destroy: () => {
      s.destroy(), n.destroy();
    },
    navigation: {
      ...n.navigation,
      state$: i,
      throttleLock: ({ duration: l, trigger: g }) => g.pipe(ao({ duration: l, reader: n })),
      panNavigator: r,
      turnBottom: o.turnBottom.bind(o),
      turnTop: o.turnTop.bind(o),
      turnLeftOrTop: o.turnLeftOrTop.bind(o),
      turnRightOrBottom: o.turnRightOrBottom.bind(o),
      turnLeft: o.turnLeft.bind(o),
      turnRight: o.turnRight.bind(o),
      goToCfi: o.goToCfi.bind(o),
      goToUrl: o.goToUrl.bind(o),
      goToSpineItem: o.goToSpineItem.bind(o),
      goToNextSpineItem: o.goToNextSpineItem.bind(o),
      goToPreviousSpineItem: o.goToPreviousSpineItem.bind(o),
      goToLeftOrTopSpineItem: o.goToLeftOrTopSpineItem.bind(o),
      goToRightOrBottomSpineItem: o.goToRightOrBottomSpineItem.bind(o),
      goToTopSpineItem: o.goToTopSpineItem.bind(o),
      goToBottomSpineItem: o.goToBottomSpineItem.bind(o),
      goToLeftSpineItem: o.goToLeftSpineItem.bind(o),
      goToRightSpineItem: o.goToRightSpineItem.bind(o),
      goToPageOfSpineItem: o.goToPageOfSpineItem.bind(o),
      goToAbsolutePageIndex: o.goToAbsolutePageIndex.bind(o)
    }
  };
}, le = (t) => Be({
  spineIndex: t.index,
  spineId: t.id
}), Gt = ({
  nodeOrRange: t,
  offset: e,
  item: n
}) => {
  const i = "ownerDocument" in t ? t.ownerDocument : t.startContainer.ownerDocument;
  return !i || !i?.documentElement || t === i ? le(n) : Xn(t) ? Be({
    start: {
      node: t.startContainer,
      offset: t.startOffset,
      spineIndex: n.index,
      spineId: n.id
    },
    end: {
      node: t.endContainer,
      offset: t.endOffset,
      spineIndex: n.index,
      spineId: n.id
    }
  }) : Be({
    node: t,
    offset: e,
    spineIndex: n.index,
    spineId: n.id
  });
}, Ge = ({
  spineItem: t,
  pageNode: e
}) => Gt({
  nodeOrRange: e.node,
  offset: e.offset,
  item: t
}).trim(), uo = (t, e) => Gt({
  nodeOrRange: t,
  item: e
}), Cs = (t) => t.index.toString(), bt = (t) => t[0]?.index === 6 && t.length > 1, Ze = (t) => {
  const e = Vt(t);
  return Nn(e);
}, lo = (t) => Array.isArray(t) ? t[0] && bt(t[0]) ? t[0] : void 0 : t.parent[0] && bt(t.parent[0]) ? t.parent[0] : void 0, dt = (t) => {
  const e = Vt(t), r = ((lo(e) ?? [])[1]?.index ?? 2) / 2 - 1, s = An(e), a = s ? e.end.at(-1)?.at(-1)?.offset : e.at(-1)?.at(-1)?.offset;
  return {
    isCfiRange: s,
    itemIndex: r,
    offset: a ?? 0
  };
}, Zt = ({
  cfi: t,
  spineItemsManager: e
}) => {
  const { itemIndex: n, ...i } = dt(t), o = e.get(n);
  if (!o)
    return {
      ...i,
      itemIndex: n,
      node: null
    };
  const r = o.renderer.getDocumentFrame();
  if (r instanceof HTMLIFrameElement) {
    const s = r.contentWindow?.document;
    if (s)
      try {
        const a = Dn(t, s, {
          throwOnError: !0
        });
        return {
          ...i,
          itemIndex: n,
          node: a.isRange ? a.node?.startContainer ?? null : a.node ?? null,
          isCfiRange: a.isRange,
          range: a.isRange ? a.node : void 0,
          offset: Array.isArray(a.offset) ? a.offset.at(-1) : a.offset,
          spineItem: o
        };
      } catch (a) {
        return E.warn(`Error resolving cfi: ${t}.`), E.warn(a), {
          ...i,
          spineItem: o,
          itemIndex: n,
          node: null
        };
      }
  }
  return {
    ...i,
    itemIndex: n,
    spineItem: o,
    node: null
  };
}, po = (t, e) => {
  if ("cfi" in e) {
    const { itemIndex: r, ...s } = t.cfi.parseCfi(e.cfi);
    return {
      ...e,
      ...s,
      itemIndex: r,
      node: null
    };
  }
  const n = t.spineItemsManager.get(e);
  if (!n)
    throw new Error("Spine item not found");
  const i = le(n.item);
  return {
    ...t.cfi.parseCfi(i),
    cfi: le(n.item),
    itemIndex: n.index,
    node: null
  };
}, go = (t, e) => {
  let n = t?.itemPageIndex;
  const { itemIndex: i, ...o } = e.cfi.parseCfi(t.cfi), r = e.spineItemsManager.get(i);
  return r ? _t().pipe(
    z(r.isReady$),
    y(([, s]) => {
      const {
        node: a = null,
        offset: c,
        range: d
      } = s ? e.cfi.resolveCfi({ cfi: t.cfi }) : {};
      r.renditionLayout !== "pre-paginated" && a && (n = e.spine.locator.spineItemLocator.getSpineItemPageIndexFromNode(
        a,
        c ?? 0,
        r
      ) ?? n);
      let l = t?.absolutePageIndex;
      return n === void 0 && (n = 0), l = e.spine.locator._getAbsolutePageIndexFromPageIndex({
        pageIndex: n,
        spineItemOrId: r
      }) ?? t?.absolutePageIndex, {
        ...t,
        ...o,
        range: d,
        itemIndex: i,
        node: a,
        startOffset: c,
        absolutePageIndex: l,
        itemPageIndex: n
      };
    })
  ) : v({ ...t, itemIndex: i, ...o });
};
class ho {
  constructor(e) {
    this.reader = e, this.locatorsByKey = /* @__PURE__ */ new Map(), this.deregisterMemoizedStream = (n) => {
      const i = this.locatorsByKey.get(n);
      i && (i.observerCount--, i.observerCount === 0 && this.locatorsByKey.delete(n));
    }, this.locate = (n, i) => {
      const o = {
        resource: n,
        meta: po(this.reader, n)
      };
      return ht(() => {
        const r = this.reader.spineItemsManager.get(
          o.meta.itemIndex ?? 0
        ), s = r?.renditionLayout === "reflowable", a = i.mode === "shallow" || !s || !r ? () => {
        } : this.reader.spine.spineItemsLoader.forceOpen([r.index]), c = "cfi" in n ? n.cfi : void 0, d = c ? this.locatorsByKey.get(c) : void 0, u = (g) => g.pipe(
          ye(() => {
            c && this.deregisterMemoizedStream(c), setTimeout(() => {
              a();
            }, 1e3);
          })
        );
        if (c && d)
          return d.observerCount++, u(
            d.consolidate$.pipe(
              y(({ resource: g, meta: p }) => ({
                resource: g,
                meta: p
              }))
            )
          );
        const l = this.reader.spine.layout$.pipe(
          De(10),
          ge(o),
          fn((g) => go(g.meta, this.reader).pipe(
            y((p) => ({
              ...g,
              meta: p
            }))
          ), o),
          Z({ refCount: !0, bufferSize: 1 })
        );
        return c && this.locatorsByKey.set(c, {
          observerCount: 1,
          consolidate$: l
        }), u(l);
      });
    };
  }
  locateResource(e, n) {
    return Array.isArray(e) ? ht(
      () => O(
        e.map((i) => this.locate(i, n ?? {}))
      ).pipe(Sn([]))
    ) : this.locate(e, n ?? {});
  }
}
const mo = E.namespace("spine"), fo = ({
  position: t,
  pageSize: e
}) => new Ei({
  ...t,
  left: t.x,
  top: t.y,
  width: e.width,
  height: e.height,
  bottom: t.y + e.height,
  right: t.x + e.width
}), So = 10, yo = (t) => {
  const { spineItem: e, ...n } = t;
  if (!t.spineItem.value.isLoaded)
    return {
      ...n,
      firstVisibleNode: void 0
    };
  const o = t.spineItem.renderer?.getDocumentFrame()?.contentWindow?.document;
  return !o || o.body === null ? {
    ...n,
    firstVisibleNode: void 0
  } : {
    ...n,
    firstVisibleNode: kn(
      o,
      t.layout
    )
  };
}, Io = (t) => new se((e) => {
  if (t.length === 0) {
    e.next([]), e.complete();
    return;
  }
  let n, i = 0;
  const o = new Array(t.length), r = () => {
    n !== void 0 && (cancelAnimationFrame(n), n = void 0);
  }, s = () => {
    n = void 0;
    let c = 0;
    for (; i < t.length && c < So; ) {
      const d = t[i];
      if (!d) {
        i = t.length;
        break;
      }
      o[i] = yo(d), i += 1, c += 1;
    }
    if (i >= t.length) {
      e.next(o), e.complete();
      return;
    }
    a();
  }, a = () => {
    n = requestAnimationFrame(s);
  };
  return a(), () => {
    r();
  };
}), ne = class ne extends ee {
  constructor(e, n, i, o, r, s) {
    super({ pages: [] }), this.spineLayout = e, this.spineItemsManager = n, this.spineItemLocator = i, this.context = o, this.locator = r, this.viewport = s, this.fromSpineItemPageIndex = (a, c) => ne.fromSpineItemPageIndex(this.value, a, c), this.fromNextPageWithinSameSpineItem = (a) => ne.fromNextPageWithinSameSpineItem(this.value, a), this.fromAbsolutePageIndex = (a) => ne.fromAbsolutePageIndex(this.value, a), this.observeFromAbsolutePageIndex = (a) => this.pipe(y(() => this.fromAbsolutePageIndex(a))), this.layout$ = e.layout$.pipe(
      z(s),
      P(([, { pageSize: a }]) => {
        const c = n.items.reduce(
          (d, u, l) => {
            for (let g = 0; g < u.numberOfPages; g += 1) {
              const p = i.getSpineItemPositionFromPageIndex({
                spineItem: u,
                pageIndex: g
              }), h = r.getSpinePositionFromSpineItemPosition({
                spineItem: u,
                spineItemPosition: p
              }), m = d.length;
              d.push({
                absoluteLayout: fo({
                  pageSize: a,
                  position: h
                }),
                layout: new Mi({
                  left: p.x,
                  right: p.x + a.width,
                  top: p.y,
                  bottom: p.y + a.height,
                  width: a.width,
                  height: a.height,
                  x: p.x,
                  y: p.y
                }),
                itemIndex: l,
                absolutePageIndex: m,
                spineItem: u,
                pageIndex: g
              });
            }
            return d;
          },
          []
        );
        return Io(c);
      }),
      y((a) => (mo.info("Pages layout", a), { pages: a })),
      A()
    ), this.layout$.pipe(T(this.destroy$)).subscribe(this.next.bind(this));
  }
};
ne.fromSpineItemPageIndex = (e, n, i) => {
  const o = typeof n == "number" ? n : n.index;
  return e.pages.find(
    (r) => r.itemIndex === o && r.pageIndex === i
  );
}, ne.fromNextPageWithinSameSpineItem = (e, n) => ne.fromSpineItemPageIndex(
  e,
  n.itemIndex,
  n.pageIndex + 1
), ne.fromAbsolutePageIndex = (e, n) => e.pages.find(
  (i) => i.absolutePageIndex === n
);
let pe = ne;
const ve = (t) => {
  const e = t.indexOf("#");
  return e >= 0 ? t.substring(0, e) : t;
}, vt = (t) => {
  if (!t) return;
  const e = t.indexOf("#");
  return e < 0 ? void 0 : t.substring(e + 1) || void 0;
}, bo = (t) => {
  try {
    return decodeURIComponent(t);
  } catch {
    return t;
  }
}, Jt = (t, e = []) => t.flatMap((n) => {
  const i = [...e, { title: n.title, path: n.path }], o = n.href || n.path;
  return [
    {
      chain: i,
      href: o,
      anchorId: vt(n.href) ?? vt(n.path)
    },
    ...Jt(n.contents, i)
  ];
}), vo = (t) => {
  const e = /* @__PURE__ */ new Map();
  return t.spineItems.forEach((n, i) => {
    e.has(n.href) || e.set(n.href, i);
  }), e;
}, wo = (t, e) => {
  const n = vo(e);
  return Jt(t).map((i) => {
    const o = ve(i.href), r = n.get(o) ?? -1;
    return {
      ...i,
      hrefWithoutAnchor: o,
      spineItemIndex: r
    };
  });
}, Po = ({
  manifest: t,
  tocIndex: e
}) => {
  const n = /* @__PURE__ */ new Map();
  e.forEach((r, s) => {
    const a = n.get(r.hrefWithoutAnchor);
    if (a) {
      a.push(s);
      return;
    }
    n.set(r.hrefWithoutAnchor, [s]);
  });
  const i = /* @__PURE__ */ new Map(), o = /* @__PURE__ */ new Map();
  for (const r of t.spineItems) {
    if (i.has(r.href)) continue;
    const s = ve(r.href);
    let a = o.get(s);
    if (!a) {
      const c = [];
      for (const [
        d,
        u
      ] of n)
        xo(
          s,
          d
        ) && c.push(...u);
      c.sort((d, u) => d - u), a = c.map(
        (d) => e[d]
      ), o.set(s, a);
    }
    i.set(r.href, a);
  }
  return i;
}, xo = (t, e) => t === e || t.endsWith(e) || e.endsWith(t), $o = (t, e) => {
  const n = t.endsWith(e), i = t.substring(
    0,
    t.lastIndexOf("/")
  ), o = e.substring(
    0,
    e.lastIndexOf("/")
  ), r = i !== "" && i.endsWith(o);
  return n || r;
}, Kt = (t) => {
  const [e, ...n] = t;
  if (!e) return;
  const i = {
    title: e.title,
    path: e.path
  };
  let o = i;
  for (const r of n) {
    const s = {
      title: r.title,
      path: r.path
    };
    o.subChapter = s, o = s;
  }
  return i;
}, To = ({
  doc: t,
  node: e,
  offset: n
}) => {
  const i = t.createRange();
  if (e.nodeType === Node.TEXT_NODE || e.nodeType === Node.CDATA_SECTION_NODE || e.nodeType === Node.COMMENT_NODE) {
    const o = e, r = Math.max(0, Math.min(n, o.length));
    i.setStart(o, r);
  } else {
    const o = e.childNodes.length, r = Math.max(0, Math.min(n, o));
    i.setStart(e, r);
  }
  return i.collapse(!0), i;
}, Fo = ({
  rangeToTest: t,
  upperBoundRange: e
}) => t.compareBoundaryPoints(Range.START_TO_START, e) < 0, Co = (t) => {
  const e = t.createRange(), n = t.body ?? t.documentElement;
  if (n)
    return e.selectNodeContents(n), e.collapse(!1), e;
}, Lo = ({
  anchorNode: t,
  nextPageNode: e
}) => t === e || t.contains(e), wt = ({
  node: t,
  offset: e,
  candidates: n,
  spineItem: i,
  nextPageEntry: o
}) => {
  if (!t || e === void 0 || !i || n.length === 0) return;
  const r = i.renderer.getDocumentFrame()?.contentDocument;
  if (!r || t.ownerDocument !== r) return;
  const s = o?.firstVisibleNode?.node, a = o?.firstVisibleNode?.offset, c = o ? s !== void 0 && a !== void 0 && s.ownerDocument === r ? To({
    doc: r,
    node: s,
    offset: a
  }) : void 0 : Co(r);
  if (!c) return;
  const d = o !== void 0 && s !== void 0 && s.ownerDocument === r;
  let u, l, g;
  for (const h of n) {
    if (!h.anchorId) {
      g = h.chain;
      continue;
    }
    const m = r.getElementById(bo(h.anchorId));
    if (!m || d && Lo({
      anchorNode: m,
      nextPageNode: s
    }))
      continue;
    const f = r.createRange();
    f.setStartBefore(m), f.collapse(!0), Fo({
      rangeToTest: f,
      upperBoundRange: c
    }) && (!l || f.compareBoundaryPoints(
      Range.START_TO_START,
      l
    ) > 0) && (u = h.chain, l = f);
  }
  const p = u ?? g;
  return p ? Kt(p) : void 0;
}, Mo = ({
  candidate: t,
  hrefHasAnchor: e
}) => {
  if (e || !t.anchorId) return !1;
  const n = t.chain[t.chain.length - 1], i = t.chain[t.chain.length - 2];
  return !n || !i ? !1 : ve(n.path) === ve(i.path);
}, Eo = ({
  href: t,
  tocIndex: e,
  manifest: n
}) => {
  const i = ve(t), o = t.includes("#"), r = n.spineItems.findIndex(
    (a) => a.href === i
  );
  let s;
  for (const a of e)
    $o(
      i,
      a.hrefWithoutAnchor
    ) && (r < a.spineItemIndex || Mo({ candidate: a, hrefHasAnchor: o }) || (s = a.chain));
  return s;
}, No = (t, e, n) => {
  const { href: i } = n, o = Eo({ href: i, tocIndex: e, manifest: t });
  return o ? Kt(o) : void 0;
}, Ao = (t, e) => t ? t.spineItems.reduce(
  (i, o) => (i[o.id] = No(t, e, o), i),
  {}
) : {}, Do = (t, e, n) => t + e * n, Ro = (t, e, n) => {
  const { height: i, width: o } = n.layoutInfo, { top: r, left: s } = t.spine.getSpineItemSpineLayoutInfo(n);
  return t.settings.values.computedPageTurnDirection === "vertical" ? Math.max(
    0,
    Math.min(
      1,
      (e.y - r + t.viewport.absoluteViewport.height) / i
    )
  ) : Math.max(
    0,
    Math.min(
      1,
      (e.x - s + t.viewport.absoluteViewport.width) / o
    )
  );
}, Oo = (t, e, n, i, o, r) => o.isReady$.pipe(
  j(),
  z(t.layoutInfo$),
  y(([s, a]) => {
    const c = r.renditionLayout === "pre-paginated", d = r.spineItems.length || 0, u = r.spineItems.slice(0, e).reduce((S, I) => S + (I.progressionWeight ?? 0), 0) || 0, l = t.spineItemsManager.getSpineItemIndex(o) ?? 0, g = r.spineItems.length ?? 0, p = a.pages.filter((S) => S.itemIndex === l).length ?? 0, h = r.spineItems[e]?.progressionWeight ?? // if no progressionWeight is defined we "assume" the document weight to be
    // relative to the total number of documents
    (l + 1) / g;
    let m = (n + 1) * (h / p);
    !c && o.renditionLayout === "reflowable" && !s && (m = 0);
    let f = u + m;
    return r.renditionFlow === "scrolled-continuous" && (s ? m = Ro(
      t,
      i,
      o
    ) : m = 0, f = Do(
      u,
      h,
      m
    )), e === d - 1 && n === p - 1 && f > 0.99 ? 1 : f;
  })
), Vo = {
  tocCandidatesBySpineHref: /* @__PURE__ */ new Map(),
  chaptersInfo: {}
}, ko = ({
  beginItem: t,
  endItem: e,
  paginationInfo: n,
  chaptersData: i,
  pagesState: o
}) => {
  const r = t && n.beginPageIndexInSpineItem !== void 0 ? pe.fromSpineItemPageIndex(
    o,
    t.index,
    n.beginPageIndexInSpineItem
  ) : void 0, s = r ? pe.fromNextPageWithinSameSpineItem(o, r) : void 0, a = e && n.endPageIndexInSpineItem !== void 0 ? pe.fromSpineItemPageIndex(
    o,
    e.index,
    n.endPageIndexInSpineItem
  ) : void 0, c = a ? pe.fromNextPageWithinSameSpineItem(o, a) : void 0, d = t ? wt({
    node: r?.firstVisibleNode?.node,
    offset: r?.firstVisibleNode?.offset,
    candidates: i.tocCandidatesBySpineHref.get(t.item.href) ?? [],
    spineItem: t,
    nextPageEntry: s
  }) : void 0, u = e ? wt({
    node: a?.firstVisibleNode?.node,
    offset: a?.firstVisibleNode?.offset,
    candidates: i.tocCandidatesBySpineHref.get(e.item.href) ?? [],
    spineItem: e,
    nextPageEntry: c
  }) : void 0;
  return {
    beginChapterInfo: d ?? (t ? i.chaptersInfo[t.item.id] : void 0),
    beginSpineItemReadingDirection: t?.readingDirection,
    beginAbsolutePageIndex: r?.absolutePageIndex,
    endChapterInfo: u ?? (e ? i.chaptersInfo[e.item.id] : void 0),
    endSpineItemReadingDirection: e?.readingDirection,
    endAbsolutePageIndex: a?.absolutePageIndex
  };
}, jo = ({
  items: t,
  pagesState: e
}) => {
  const n = t.map(() => 0);
  for (const i of e.pages)
    n[i.itemIndex] = (n[i.itemIndex] ?? 0) + 1;
  return {
    numberOfPagesPerItems: n,
    /**
     * This may be not accurate for reflowable due to dynamic load / unload.
     */
    numberOfTotalPages: e.pages.length
  };
}, Uo = (t, e) => t.numberOfTotalPages !== e.numberOfTotalPages || t.numberOfPagesPerItems.length !== e.numberOfPagesPerItems.length ? !1 : t.numberOfPagesPerItems.every(
  (n, i) => n === e.numberOfPagesPerItems[i]
), zo = ({
  reader: t,
  paginationInfo: e,
  navigationPosition: n,
  manifest: i
}) => {
  const o = t.spineItemsManager.get(e.endSpineItemIndex);
  return o ? Oo(
    t,
    e.endSpineItemIndex ?? 0,
    e.endPageIndexInSpineItem || 0,
    n,
    o,
    i
  ) : v(0);
}, Wo = (t) => t.context.manifest$.pipe(
  y((e) => {
    if (!e)
      return Vo;
    const n = wo(e.nav?.toc ?? [], e);
    return {
      tocCandidatesBySpineHref: Po({
        manifest: e,
        tocIndex: n
      }),
      chaptersInfo: Ao(e, n)
    };
  })
), _o = (t) => ({
  beginSpineItemIndex: t.beginSpineItemIndex,
  beginPageIndexInSpineItem: t.beginPageIndexInSpineItem,
  endSpineItemIndex: t.endSpineItemIndex,
  endPageIndexInSpineItem: t.endPageIndexInSpineItem
}), Ho = (t) => {
  const e = t.pagination.state$, n = t.spine.pages.layout$, i = Wo(t).pipe(
    Z({ bufferSize: 1, refCount: !0 })
  ), o = e.pipe(
    y(_o),
    D(V)
  ), r = O([
    o,
    i,
    n
  ]).pipe(
    y(([u, l, g]) => {
      const p = t.spineItemsManager.get(
        u.beginSpineItemIndex
      ), h = t.spineItemsManager.get(
        u.endSpineItemIndex
      );
      return ko({
        beginItem: p,
        endItem: h,
        paginationInfo: u,
        chaptersData: l,
        pagesState: g
      });
    }),
    Z({ bufferSize: 1, refCount: !0 })
  ), s = t.settings.watch(["computedSpreadMode"]).pipe(
    y((u) => u.computedSpreadMode ?? !1),
    D()
  ), a = O([
    n,
    t.spineItemsManager.items$
  ]).pipe(
    y(
      ([u, l]) => jo({ items: l, pagesState: u })
    ),
    D(Uo)
  ), c = O([
    e,
    s,
    r,
    a
  ]).pipe(
    y(([u, l, g, p]) => ({
      ...u,
      ...g,
      isUsingSpread: l,
      ...p
    }))
  ), d = O([
    e,
    t.layout$,
    t.navigation.navigation$,
    t.context.manifest$
  ]).pipe(
    P(
      ([u, l, g, p]) => zo({
        reader: t,
        paginationInfo: u,
        navigationPosition: g.position,
        manifest: p
      })
    ),
    y((u) => ({
      /**
       * This percentage is based of the weight (kb) of every items and the number of pages.
       * It is not accurate but gives a general good idea of the overall progress.
       * It is recommended to use this progress only for reflow books. For pre-paginated books
       * the number of pages and current index can be used instead since 1 page = 1 chapter.
       */
      percentageEstimateOfBook: u
    }))
  );
  return O([c, d]).pipe(
    y(([u, l]) => ({
      ...u,
      ...l
    })),
    D(V),
    yn(5)
  );
}, Bo = (t) => (e) => {
  const n = t(e), i = new Y({
    ...n.pagination.state,
    beginChapterInfo: void 0,
    beginCfi: void 0,
    beginPageIndexInSpineItem: void 0,
    isUsingSpread: !1,
    beginAbsolutePageIndex: 0,
    endAbsolutePageIndex: 0,
    numberOfTotalPages: 0,
    beginSpineItemReadingDirection: void 0,
    beginSpineItemIndex: void 0,
    endCfi: void 0,
    endChapterInfo: void 0,
    endSpineItemReadingDirection: void 0,
    percentageEstimateOfBook: 0
  }), o = new ho(n), r = Ho(n).pipe(x((s) => E.log("Pagination", s))).subscribe(i);
  return {
    ...n,
    locateResource: o.locateResource.bind(o),
    destroy: () => {
      r.unsubscribe(), n.destroy();
    },
    pagination: {
      ...n.pagination,
      get state() {
        return i.value;
      },
      get state$() {
        return i;
      }
    }
  };
}, Xo = (t) => ({
  put: (r, s) => new Promise((a, c) => {
    const d = t.transaction(["store"], "readwrite");
    d.onerror = (g) => {
      c(g);
    }, d.oncomplete = () => {
      a();
    };
    const l = d.objectStore("store").put(s, r);
    l.onsuccess = () => {
      a();
    }, l.onerror = (g) => {
      c(g);
    };
  }),
  keys: () => new Promise((r, s) => {
    const a = t.transaction(["store"], "readonly");
    a.onerror = (l) => {
      s(l);
    };
    const d = a.objectStore("store").openKeyCursor(), u = [];
    d.onsuccess = () => {
      const l = d.result;
      if (!l) {
        r(u);
        return;
      }
      u.push(l.key), l.continue();
    }, d.onerror = () => {
      s(d.error);
    };
  }),
  get: (r) => new Promise((s, a) => {
    const c = t.transaction(["store"], "readwrite"), u = c.objectStore("store").get(r);
    u.onsuccess = () => {
      let l = u.result;
      l === void 0 && (l = null), s(l);
    }, c.onerror = () => {
      a(u.error);
    };
  }),
  remove: (r) => new Promise((s, a) => {
    const c = t.transaction(["store"], "readwrite"), u = c.objectStore("store").delete(r);
    c.onerror = () => {
      a(u.error);
    }, c.oncomplete = () => {
      s();
    }, c.onabort = () => {
      const l = u.error ? u.error : u.transaction?.error;
      a(l);
    };
  })
}), Ue = async (t) => new Promise((e, n) => {
  const i = window.indexedDB.open(t);
  i.onerror = (o) => {
    n(o);
  }, i.onsuccess = () => {
    e(Xo(i.result));
  }, i.onupgradeneeded = () => {
    i.result.createObjectStore("store");
  };
}), Yo = (t) => {
  let e = Date.now().toString();
  const n = new R(), i = (c) => {
    if (typeof c == "string" || typeof c == "object") {
      const d = typeof c == "object" ? c.id : c;
      return t.manifest?.spineItems.find((u) => u.id === d);
    }
    return t.manifest?.spineItems[c];
  }, o = async (c, d) => {
    const u = i(c);
    if (!u) return new Response("Item not found", { status: 404 });
    const g = await (await Ue("prose-reader")).get(`${e}_${u.id}`);
    if (g)
      return new Response(g, { status: 200 });
    const p = d && await d(u) || await fetch(u.href);
    return r(u, p.clone()), p;
  }, r = (c, d) => {
    n.next({ id: c, data: d });
  };
  n.asObservable().pipe(
    Mn(({ id: c, data: d }) => {
      const u = i(c);
      return u ? H(
        In([Ue("prose-reader"), H(d.blob())])
      ).pipe(
        B(([l, g]) => H(l.put(`${e}_${u.id}`, g))),
        pt((l) => (E.error(l), X))
      ) : X;
    }),
    Q(t.destroy$)
  ).subscribe();
  const s = t.manifest$.pipe(
    k(() => {
      e = Date.now().toString();
    })
  );
  return w(s).pipe(
    B(() => (E.debug("Cleanup up old cache..."), H(Ue("prose-reader")).pipe(
      B(
        (c) => H(c.keys()).pipe(
          oe(
            (d) => d.filter((u) => !u.toString().startsWith(e))
          ),
          B((d) => {
            const u = d.map((l) => c.remove(l));
            return H(Promise.all(u));
          })
        )
      ),
      pt((c) => (E.error(c), X))
    ))),
    Q(t.destroy$)
  ).subscribe(), {
    get: o,
    destroy: () => {
      n.complete();
    }
  };
}, qo = (t) => (e) => {
  const n = t(e), i = Yo(n.context);
  return {
    ...n,
    // $: {
    //   ...reader.$,
    //   errors$: merge(reader.$.errors$, errorsSubject$.asObservable())
    // },
    destroy: () => {
      i.destroy(), n.destroy();
    }
    // load,
  };
}, Go = (t, e) => {
  const n = t.node.ownerDocument?.createRange(), i = t.node.compareDocumentPosition(e.node);
  if (n) {
    try {
      if (i & Node.DOCUMENT_POSITION_PRECEDING)
        n.setStart(e.node, e.offset || 0), n.setEnd(t.node, t.offset || 0);
      else if (i & Node.DOCUMENT_POSITION_FOLLOWING)
        n.setStart(t.node, t.offset || 0), n.setEnd(e.node, e.offset || 0);
      else {
        const o = Math.min(t.offset || 0, e.offset || 0), r = Math.max(t.offset || 0, e.offset || 0);
        n.setStart(t.node, o), n.setEnd(t.node, r);
      }
    } catch (o) {
      E.warn("Failed to create range from selection", o, {
        anchor: t,
        focus: e
      });
    }
    return n;
  }
}, Zo = ({
  selection: t,
  spineItem: e
}) => {
  const { anchorNode: n, anchorOffset: i, focusNode: o, focusOffset: r } = t;
  if (!(!n || !o))
    try {
      return Go(
        { node: n, offset: i },
        { node: o, offset: r }
      );
    } catch (s) {
      E.warn("Failed to create range from selection", s, {
        selection: t,
        spineItem: e
      });
      return;
    }
};
class Jo extends q {
  constructor(e) {
    super();
    const n = e.contentDocument || e.contentWindow?.document;
    if (!n)
      this.selectionChange$ = Se, this.selectionOver$ = Se;
    else {
      const i = mt(n.body, {
        childList: !0,
        subtree: !0
      }).pipe(
        M(
          (r) => !!r.find((s) => s.type === "childList" && s.removedNodes.length)
        )
      ), o = e.parentElement ? mt(e.parentElement, {
        childList: !0
      }).pipe(
        M(
          (r) => !!r.find(
            (s) => Array.from(s.removedNodes).includes(e)
          )
        )
      ) : v(null);
      this.selectionChange$ = w(
        G(n, "selectionchange"),
        i
      ).pipe(
        y(() => n.getSelection()),
        T(w(o, this.destroy$)),
        Ie(null)
      ), this.selectionOver$ = G(n, "pointerdown").pipe(
        P(
          () => w(
            G(n, "pointerup"),
            G(n, "pointercancel"),
            G(n, "contextmenu")
          ).pipe(
            j(),
            /**
             * The selection is still valid during the event even if it will
             * be discarded. The timeout make sure to detect this edge case.
             */
            Fe(0),
            y((r) => {
              const s = n.getSelection();
              return s && !s.isCollapsed ? [r, s] : void 0;
            }),
            M(K)
          )
        ),
        T(w(o, this.destroy$))
      );
    }
  }
}
const Ko = (t) => t.watch("isLoaded").pipe(
  P(() => {
    const e = t.renderer.getDocumentFrame(), n = e?.contentDocument || e?.contentWindow?.document;
    if (!e || !n) return Se;
    const i = new Jo(e);
    return w(
      i.selectionChange$.pipe(
        y((o) => {
          if (o?.toString())
            return {
              type: "change",
              selection: o
            };
        })
      ),
      i.selectionOver$.pipe(
        y(([o, r]) => ({
          type: "over",
          event: o,
          selection: r
        }))
      )
    ).pipe(
      T(t.unloaded$),
      Ie(void 0),
      ye(() => {
        i.destroy();
      })
    );
  }),
  D()
), Qo = (t) => (e) => {
  const n = t(e);
  let i;
  const o = n.spineItemsManager.items$.pipe(
    P((u) => {
      const l = u.map((g) => {
        const p = n.spineItemsManager.getSpineItemIndex(g) ?? 0;
        return Ko(g).pipe(
          y((h) => {
            if (h)
              return {
                ...h,
                itemIndex: p
              };
          })
        );
      });
      return w(...l);
    }),
    ge(void 0),
    D(),
    x((u) => {
      i = u;
    }),
    Z({ refCount: !0, bufferSize: 1 })
  ), r = o, s = o.pipe(
    y((u) => !!u),
    D(),
    M((u) => u),
    A()
  ), a = s.pipe(
    P(() => r),
    D(),
    M((u) => !u),
    A()
  ), c = o.pipe(
    M((u) => u?.type === "over"),
    A()
  ), d = n.context.watch("rootElement").pipe(
    M(K),
    P((u) => G(u, "pointerdown")),
    z(r),
    y(([, u]) => u),
    ge(void 0),
    Z({ refCount: !0, bufferSize: 1 })
  );
  return w(r, d).pipe(T(n.$.destroy$)).subscribe(), {
    ...n,
    selection: {
      selection$: r,
      selectionStart$: s,
      selectionEnd$: a,
      selectionOver$: c,
      lastSelectionOnPointerdown$: d,
      getSelection: () => i,
      createOrderedRangeFromSelection: Zo
    }
  };
}, Pt = [
  {
    name: "bright",
    backgroundColor: "white"
  },
  {
    name: "sepia",
    backgroundColor: "#eaddc7",
    foregroundColor: "black"
  },
  {
    name: "night",
    backgroundColor: "#191717",
    foregroundColor: "#f1ebeb"
  }
], er = (t) => (e) => {
  const n = t(e), i = new Y(
    e.theme ?? "bright"
  ), o = () => {
    const a = Pt.find(
      (c) => c.name === i.value
    );
    return `
      body {
        ${a !== void 0 ? `background-color: ${a.backgroundColor} !important;` : ""}
      }
      ${a?.foregroundColor ? `
          body * {
            
            color: ${a.foregroundColor};
          }
        ` : ""}
    `;
  }, r = ({
    container: a
  }) => {
    const c = Pt.find(
      (d) => d.name === i.value
    );
    return c && a.style.setProperty(
      "background-color",
      c.backgroundColor
    ), () => {
    };
  }, s = () => {
    n.spineItemsManager.items.forEach((a) => {
      const c = a.renderer.getDocumentFrame();
      c && re(c, "prose-reader-theme", o()), r({ container: a.element });
    });
  };
  return n.hookManager.register("item.onDocumentLoad", ({ itemId: a }) => {
    const c = n.spineItemsManager.get(a);
    if (c?.renditionLayout !== "pre-paginated") {
      const d = c?.renderer.getDocumentFrame();
      d && re(d, "prose-reader-theme", o());
    }
  }), n.spineItemsManager.items$.pipe(
    k(
      (a) => a.map(
        ({ element: c }) => r({ container: c })
      )
    ),
    Q(n.$.destroy$)
  ).subscribe(), i.pipe(
    k(() => {
      s();
    }),
    Q(n.$.destroy$)
  ).subscribe(), {
    ...n,
    theme: {
      set: (a) => {
        a !== i.value && i.next(a);
      },
      get: () => i.value,
      $: {
        theme$: i.asObservable()
      }
    }
  };
}, tr = (t) => (e) => {
  const n = t(e);
  return {
    ...n,
    utils: {
      isOrIsWithinValidLink: (r) => !!(zt(r) && (r.nodeName === "a" ? r : r.closest("a"))?.getAttribute("href")),
      injectScopedCSS: (r, s, a) => {
        const c = a.replace("${id}", n.id);
        return it(r, `${s}-${n.id}`, c);
      }
    }
  };
};
navigator.userAgent.indexOf("") > -1 && navigator.userAgent.indexOf("Chrome") <= -1;
const nr = (t) => (e) => t(e), xt = (t, e, n) => {
  const { clientWidth: i, clientHeight: o } = n.value.element, r = {
    maxX: 0,
    minX: i * (1 - e),
    maxY: 0,
    minY: o * (1 - e)
  };
  return {
    x: Math.min(Math.max(t.x, r.minX), r.maxX),
    y: Math.min(Math.max(t.y, r.minY), r.maxY)
  };
}, ir = (t, e, n) => {
  n.style.transformOrigin = "0 0";
  const i = `translate3d(${e.x}px, ${e.y}px, 0px)`, o = `scale(${t})`;
  n.style.transform = `${i} ${o}`;
}, or = (t, e, n, i) => {
  const o = e / t, r = n.clientWidth, s = n.clientHeight, a = r / 2 - i.x, c = s / 2 - i.y;
  return {
    x: i.x + a * (1 - o),
    y: i.y + c * (1 - o)
  };
};
class Qt {
  constructor(e) {
    this.x = e.x, this.y = e.y;
  }
}
class $t extends Qt {
}
class rr extends Qt {
  constructor() {
    super(...arguments), this.__symbol = /* @__PURE__ */ Symbol("UnboundScrollPosition");
  }
}
class sr extends ee {
  constructor(e, n, i, o, r) {
    super({
      element: void 0
    }), this.viewport = e, this.settings = n, this.hookManager = i, this.spine = o, this.context = r, this.navigateSubject = new R(), this.scrollingSubject = new Y(!1), this.isScrolling$ = this.scrollingSubject.asObservable(), this.setViewportPosition = ({
      position: l
    }) => {
      const g = this.value.element;
      this.scrollingSubject.next(!0);
      const p = this.fromSpinePosition(l);
      g?.scrollTo({
        left: p.x,
        top: p.y,
        behavior: "instant"
      }), Me(1).pipe(
        x(() => {
          this.scrollingSubject.next(!1);
        }),
        T(w(this.scrollingSubject.pipe(Rt(1)), this.destroy$))
      ).subscribe(), this.hookManager.execute("onViewportOffsetAdjust", void 0, {});
    };
    const s = this.context.pipe(
      st(["rootElement"]),
      x(({ rootElement: l }) => {
        if (!l) return;
        const g = document.createElement("div");
        g.setAttribute(`data-${ji}`, ""), g.appendChild(this.viewport.value.element), l.appendChild(g), this.update({ element: g });
      })
    ), a = O([
      n.watch(["computedPageTurnMode"]),
      this.watch("element")
    ]).pipe(
      x(([{ computedPageTurnMode: l }, g]) => {
        g && (l === "scrollable" ? g.style.display = "block" : g.style.display = "contents");
      })
    ), c = this.navigateSubject.pipe(x(this.setViewportPosition));
    this.isNavigating$ = this.navigateSubject.pipe(
      ge(!1),
      P(() => w(v(!0), v(!1))),
      Z(1)
    );
    const d = w(
      o.element$.pipe(
        M(K),
        P((l) => at(l))
      ),
      o.element$.pipe(
        M(K),
        P((l) => G(l, "scroll"))
      ),
      o.spineItemsObserver.itemResize$
    ).pipe(
      P(
        () => Me(10).pipe(
          y(() => !1),
          ge(!0)
        )
      ),
      D(),
      ge(!1)
    ), u = O([
      d,
      this.isScrolling$
    ]).pipe(
      y(
        ([l, g]) => l || g
      ),
      Z(1)
    );
    this.userScroll$ = this.watch("element").pipe(
      M(K),
      P(
        (l) => n.watch(["computedPageTurnMode"]).pipe(
          P(
            ({ computedPageTurnMode: g }) => g === "controlled" ? Se : G(l, "scroll").pipe(
              z(u),
              M(
                ([, p]) => !p
              ),
              y(([p]) => p)
            )
          )
        )
      ),
      A()
    ), w(s, a, c).pipe(T(this.destroy$)).subscribe();
  }
  update(e) {
    this.mergeCompare(e);
  }
  navigate(e) {
    this.navigateSubject.next(e);
  }
  fromScrollPosition(e) {
    const n = this.viewport.scaleFactor;
    return new ce({
      x: e.x / n,
      y: e.y / n
    });
  }
  fromSpinePosition(e) {
    const n = this.viewport.scaleFactor;
    return new $t({
      x: e.x * n,
      y: e.y * n
    });
  }
  get scrollPosition() {
    const e = this.value.element;
    return new $t({
      x: e?.scrollLeft ?? 0,
      y: e?.scrollTop ?? 0
    });
  }
}
const ar = (t, e, n, i, o) => {
  const r = t.clientWidth, s = t.clientHeight, a = t.scrollLeft, c = t.scrollTop, d = a + r / 2 - i, u = c + s / 2 - o, l = n / e, g = d * l, p = u * l, h = g - r / 2 + i, m = p - s / 2 + o;
  return new rr({
    x: h,
    y: m
  });
}, cr = (t, e) => {
  const n = e.viewport, i = e.navigation.scrollNavigationController, o = n.value.element, r = i.value.element, s = i.value.element, a = Math.round(n.scaleFactor * 100) / 100, c = o.offsetLeft, d = o.offsetTop, u = ar(
    r ?? _n(),
    a,
    t,
    c,
    d
  ), l = t < 1 ? "down" : "up";
  s?.setAttribute(
    `data-${U}-zooming-direction`,
    l
  ), t < 1 ? o.style.transformOrigin = "top" : t > 1 && (o.style.transformOrigin = `${c}px ${d}px`), o.style.transform = `scale(${t})`;
  const g = i.fromScrollPosition(
    u
  );
  e.navigation.navigate({
    position: g
  });
}, $e = 200;
class ur extends ee {
  constructor(e) {
    super({
      isZooming: !1,
      currentScale: 1,
      currentPosition: { x: 0, y: 0 }
    }), this.reader = e, this.enterSubject = new R(), this.exitSubject = new R();
    const n = this.enterSubject.pipe(
      P((o) => {
        const { scale: r = 1, animate: s = !1 } = o ?? {}, a = this.reader.viewport.value.element;
        return this.viewport.element.setAttribute(
          `data-${U}-zooming`,
          "true"
        ), this.scrollNavigationController.value.element?.setAttribute(
          `data-${U}-zooming`,
          this.isControlled ? "false" : "true"
        ), s && this.isControlled && (a.style.transition = `transform ${$e}ms`), this.mergeCompare({
          isZooming: !0,
          currentScale: 1,
          currentPosition: { x: 0, y: 0 }
        }), this.updateZoom(r), r !== 1 && this.isControlled ? Me(o?.animate ? $e : 0).pipe(
          x(() => {
            this.reader.layout();
          })
        ) : X;
      })
    ), i = this.exitSubject.pipe(
      P((o) => {
        const r = this.reader.viewport.value.element;
        return this.viewport.element.setAttribute(
          `data-${U}-zooming`,
          "false"
        ), this.scrollNavigationController.value.element?.removeAttribute(
          `data-${U}-zooming-direction`
        ), this.scrollNavigationController.value.element?.setAttribute(
          `data-${U}-zooming`,
          "false"
        ), o?.animate && this.isControlled && (r.style.transition = `transform ${$e}ms`), this.updateZoom(1, { x: 0, y: 0 }), r.style.transform = "", this.mergeCompare({
          isZooming: !1
        }), Me(o?.animate ? $e : 0).pipe(
          x(() => {
            const s = this.reader.viewport.value.element;
            s.style.transformOrigin = "", s.style.transition = "", this.isControlled && this.reader.layout();
          }),
          T(this.enterSubject)
        );
      })
    );
    w(n, i).pipe(T(this.destroy$)).subscribe();
  }
  enter(e) {
    this.enterSubject.next(e);
  }
  exit(e) {
    this.exitSubject.next(e);
  }
  move(e, n) {
    this.isControlled && (this.viewport.element.style.transition = "", this.updateZoom(this.value.currentScale, e, {
      constraints: n?.constrain === "within-viewport" ? (...i) => xt(...i, this.reader.viewport) : void 0
    }));
  }
  scaleAt(e, n) {
    this.viewport.element.style.transition = "";
    const o = Math.ceil(e * 100) / 100;
    this.updateZoom(o, void 0, {
      constraints: n?.constrain === "within-viewport" ? (...r) => xt(...r, this.reader.viewport) : void 0
    });
  }
  updateZoom(e, n, i) {
    if (this.isControlled) {
      const o = n || or(
        this.value.currentScale,
        e,
        this.viewport.element,
        this.value.currentPosition
      ), r = i?.constraints?.(o, e) ?? o;
      return this.applyZoom(e, r), this.mergeCompare({
        currentScale: e,
        currentPosition: r
      });
    }
    this.applyZoom(e, n ?? this.value.currentPosition), this.mergeCompare({
      currentScale: e
    });
  }
  applyZoom(e, n) {
    this.isControlled ? ir(
      e,
      n,
      this.viewport.element
    ) : cr(e, this.reader);
  }
  get isControlled() {
    return this.reader.settings.values.computedPageTurnMode === "controlled";
  }
  get scrollNavigationController() {
    return this.reader.navigation.scrollNavigationController;
  }
  get viewport() {
    return this.reader.viewport.value;
  }
}
const dr = (t) => (e) => {
  const n = t(e), i = new ur(n), o = () => {
    i.destroy(), n.destroy();
  }, r = i;
  return {
    ...n,
    destroy: o,
    zoom: {
      enter: i.enter.bind(i),
      scaleAt: i.scaleAt.bind(i),
      move: i.move.bind(i),
      exit: i.exit.bind(i),
      state$: r,
      get state() {
        return i.value;
      }
    }
  };
}, en = (t) => t?.renditionLayout === "pre-paginated" || t?.spineItems.every((e) => e.renditionLayout === "pre-paginated");
class lr {
  constructor() {
    this.navigationSubject = new lt(1), this.viewportStateSubject = new Y("free"), this.paginationSubject = new lt(), this.navigationIsLockedSubject = new Y(!1), this.pagination$ = this.paginationSubject.asObservable(), this.navigationUnlocked$ = this.navigationIsLockedSubject.pipe(
      D(),
      M((e) => !e)
    ), this.viewportState$ = this.viewportStateSubject.asObservable(), this.viewportFree$ = this.viewportState$.pipe(
      M((e) => e === "free")
    ), this.viewportBusy$ = this.viewportState$.pipe(
      M((e) => e === "busy")
    ), this.navigation$ = this.navigationSubject.asObservable();
  }
}
class pr extends ee {
  constructor() {
    super({
      assumedRenditionLayout: "reflowable"
    }), this.bridgeEvent = new lr(), this.manifest$ = this.pipe(
      oe((e) => e.manifest),
      me(K),
      Pe()
    ), this.isRTL = () => this.value.manifest?.readingDirection === "rtl";
  }
  update(e) {
    const n = this.value, i = e.manifest ?? n.manifest, o = {
      ...n,
      ...e,
      ...e.manifest && {
        isFullyPrePaginated: en(i),
        assumedRenditionLayout: i?.renditionLayout ?? "reflowable"
      }
    };
    this.mergeCompare(o);
  }
  get manifest() {
    return this.value.manifest;
  }
  get readingDirection() {
    return this.manifest?.readingDirection;
  }
}
class gr extends ee {
  constructor(e, n) {
    super({
      supportedPageTurnAnimation: ["fade", "none", "slide"],
      supportedPageTurnMode: ["controlled", "scrollable"],
      supportedPageTurnDirection: ["horizontal", "vertical"],
      supportedComputedPageTurnDirection: ["horizontal", "vertical"]
    }), O([
      e.watch(["manifest", "hasVerticalWriting"]),
      n.watch(["computedPageTurnMode"])
    ]).pipe(
      y(([{ manifest: i, hasVerticalWriting: o }, { computedPageTurnMode: r }]) => ({
        hasVerticalWriting: o,
        renditionFlow: i?.renditionFlow,
        renditionLayout: i?.renditionLayout,
        computedPageTurnMode: r
      })),
      D(V),
      y(
        ({
          hasVerticalWriting: i,
          renditionFlow: o,
          renditionLayout: r,
          computedPageTurnMode: s
        }) => ({
          ...this.value,
          supportedPageTurnMode: o === "scrolled-continuous" ? ["scrollable"] : ["controlled", "scrollable"],
          supportedPageTurnAnimation: o === "scrolled-continuous" || s === "scrollable" ? ["none"] : i ? ["fade", "none"] : ["fade", "none", "slide"],
          supportedPageTurnDirection: s === "scrollable" ? ["vertical"] : r === "reflowable" ? ["horizontal"] : ["horizontal", "vertical"]
        })
      ),
      T(this.destroy$)
    ).subscribe(this.next.bind(this));
  }
}
class hr {
  constructor() {
    this._hooks = [], this._hookExecutions = [];
  }
  /**
   * Will:
   * - call destroy function for every execution of this specific hook
   * - remove the hook for further calls
   */
  deregister(e) {
    return this._hooks = this._hooks.filter((n) => n !== e), this.destroy(e.name, void 0, e);
  }
  /**
   * Ideal when your logic only needs to apply something to the item when it's loaded.
   * You can manipulate your item later if you need to update it and trigger a layout.
   * This logic will not run every time there is a layout.
   */
  register(e, n) {
    const i = {
      name: e,
      runFn: n
    };
    return this._hooks.push(i), () => {
      this.deregister(i);
    };
  }
  execute(e, n, i) {
    return this._hooks.filter(
      (s) => e === s.name
    ).map((s) => {
      let a = () => v(void 0);
      const c = new R(), d = (g) => {
        a = g;
      }, u = () => (c.next(), c.complete(), a() ?? v(void 0)), l = s.runFn({
        // biome-ignore lint/suspicious/noExplicitAny: TODO
        ...i,
        destroy$: c.asObservable(),
        destroy: d
      });
      return this._hookExecutions.push({
        name: e,
        id: n,
        destroyFn: u,
        ref: s
      }), l;
    });
  }
  destroy(e, n, i) {
    const o = this._hookExecutions.filter(
      (s) => (
        // by ref is higher priority
        i && s.ref === i || // otherwise we refine by name and eventually by id
        e === s.name && (!n || n && n === s.id)
      )
    );
    this._hookExecutions = this._hookExecutions.filter(
      (s) => !o.includes(s)
    );
    const r = o.map(({ destroyFn: s }) => s());
    return O(r);
  }
}
const mr = "[data-prose-reader-spine-item][data-is-ready=false]{pointer-events:none;user-select:none;-webkit-user-select:none}[data-prose-reader-scroll-navigator][data-prose-reader-zooming=true]{overflow:scroll}[data-prose-reader-scroll-navigator][data-prose-reader-zooming=true][data-prose-reader-zooming-direction=down]{overflow-x:auto}[data-prose-reader-scroll-navigator]{height:100%;width:100%;position:relative;overflow-y:scroll;overflow-x:hidden}[data-prose-reader-viewport]{background-color:#fff;position:relative;height:100%;width:100%}", fr = (t) => ({
  x: -t.x,
  y: -t.y
}), Sr = (t) => t instanceof DOMMatrix ? new L({
  x: -t.e,
  y: -t.f
}) : new L({
  x: -t.x,
  y: -t.y
}), yr = "navigation/ViewportNavigator", Tt = E.namespace(yr);
class Ir extends q {
  constructor(e, n, i, o, r) {
    super(), this.settings = e, this.hookManager = n, this.context = i, this.spine = o, this.viewport = r, this.navigateSubject = new R(), this.element$ = new Y(
      document.createElement("div")
    );
    const s = this.spine.element$.pipe(
      M(K),
      z(this.element$),
      x(([u, l]) => {
        l.style.cssText = `
          height: 100%;
          width: 100%;
          position: relative;
        `, l.className = `${U}-controlled-navigator`, l.innerHTML = "", l.appendChild(u), this.viewport.value.element.appendChild(l), this.element$.next(l);
      })
    ), a = e.watch([
      "computedPageTurnDirection",
      "computedPageTurnMode",
      "numberOfAdjacentSpineItemToPreLoad"
    ]), c = O([
      a,
      this.element$
    ]).pipe(
      x(([, u]) => {
        e.values.computedPageTurnMode === "scrollable" ? u.style.display = "contents" : u.style.display = "block";
      })
    );
    this.layout$ = c.pipe(
      x(() => {
        Tt.info("layout", e.values);
      }),
      A()
    );
    const d = this.navigateSubject.pipe(
      x((u) => {
        Tt.info("Navigation requested", u);
      })
    );
    this.isNavigating$ = d.pipe(
      y(({ animation: u, position: l }) => ({
        type: "manualAdjust",
        shouldAnimate: !(!u || u === "turn" && e.values.computedPageTurnAnimation === "none"),
        animation: u,
        position: l
      })),
      P((u) => {
        const l = this.element$.getValue();
        return l.style.setProperty("transition", "none"), l.style.setProperty("opacity", "1"), w(
          v(!0),
          v(null).pipe(
            fe(() => {
              if (u?.type !== "manualAdjust") return v(!1);
              const g = u.animation === "snap" ? e.values.snapAnimationDuration : e.values.computedPageTurnAnimationDuration, p = u.animation === "snap" ? "slide" : e.values.computedPageTurnAnimation;
              return v(u).pipe(
                /**
                 * @important
                 * Optimization:
                 * When the adjustment does not need animation it means we want to be there as fast as possible
                 * One example is when we adjust position after layout. In this case we don't want to have flicker or see
                 * anything for x ms while we effectively adjust. We want it to be immediate.
                 * However when user is repeatedly turning page, we can improve smoothness by delaying a bit the adjustment
                 */
                u.shouldAnimate ? Fe(1, be) : Ce,
                x((h) => {
                  const m = this.element$.getValue();
                  h.shouldAnimate ? p === "fade" ? (m.style.setProperty(
                    "transition",
                    `opacity ${g / 2}ms`
                  ), m.style.setProperty("opacity", "0")) : (u.animation === "snap" || p === "slide") && (m.style.setProperty(
                    "transition",
                    `transform ${g}ms`
                  ), m.style.setProperty("opacity", "1")) : (m.style.setProperty("transition", "none"), m.style.setProperty("opacity", "1"));
                }),
                /**
                 * @important
                 * We always need to adjust the reading offset. Even if the current viewport value
                 * is the same as the payload position. This is because an already running animation could
                 * be active, meaning the viewport is still adjusting itself (after animation duration). So we
                 * need to adjust to anchor to the payload position. This is because we use viewport computed position,
                 * not the value set by `setProperty`
                 */
                x((h) => {
                  p !== "fade" && this.setViewportPosition(h.position);
                }),
                u.shouldAnimate ? Fe(g / 2, be) : Ce,
                x((h) => {
                  const m = this.element$.getValue();
                  p === "fade" && (this.setViewportPosition(h.position), m.style.setProperty("opacity", "1"));
                }),
                u.shouldAnimate ? Fe(g / 2, be) : Ce,
                x((h) => {
                  p === "fade" && this.setViewportPosition(h.position);
                })
              );
            }),
            y(() => !1)
          )
        );
      }),
      ge(!1),
      Z(1)
    ), w(s, this.isNavigating$, this.layout$).pipe(T(this.destroy$)).subscribe();
  }
  /**
   * Programmatically set the viewport position.
   *
   * Usually occurs due to navigation.
   *
   * @see https://stackoverflow.com/questions/22111256/translate3d-vs-translate-performance
   * for remark about flicker / fonts smoothing
   */
  setViewportPosition(e) {
    const n = this.element$.getValue(), i = fr(e);
    n.style.transform = `translate(${i.x}px, ${i.y}px)`, this.hookManager.execute("onViewportOffsetAdjust", void 0, {});
  }
  navigate(e) {
    this.navigateSubject.next(e);
  }
  /**
   * @important The reason we use computed transform and not bounding client is to avoid
   * transformation inconsistency between the viewport and the spine.
   */
  get viewportPosition() {
    const e = this.element$.getValue(), n = window.getComputedStyle(e), i = n.transform || n.webkitTransform;
    if (!i || i === "none")
      return new L({ x: 0, y: 0 });
    const o = new DOMMatrix(i);
    return Sr(o);
  }
}
const br = () => (t) => t.pipe(
  y(({ navigation: e, pagination: n, ...i }) => ({
    navigation: {
      ...e,
      paginationBeginCfi: n.beginCfi
    },
    ...i
  }))
), vr = (t, e, n) => t.bridgeEvent.pagination$.pipe(
  z(e),
  M(
    ([i, o]) => i.navigationId === o.id
  ),
  /**
   * We only register the pagination cfi IF the spine item is ready.
   * Otherwise we might save something incomplete and thus restore
   * the user to an invalid location.
   */
  P(([i, o]) => (n.spineItemsManager.get(o.spineItem)?.isReady$.pipe(j()) ?? v(!1)).pipe(
    M((s) => s),
    y(() => ({
      pagination: i,
      navigation: o
    }))
  )),
  br(),
  D(
    (i, o) => i.navigation.paginationBeginCfi === o.navigation.paginationBeginCfi
  ),
  y(
    ({ navigation: i }) => ({
      ...i,
      meta: {
        triggeredBy: "pagination"
      }
    })
  )
), wr = ({ navigationResolver: t }) => (e) => e.pipe(
  y(([n, i]) => {
    const o = {
      type: "api",
      meta: {
        triggeredBy: "user"
      },
      id: /* @__PURE__ */ Symbol(),
      animation: "turn",
      ...n,
      /**
       * @important
       * For now we do not allow out of bounds positions. (this happens when scroll mode with a zoom out).
       * Although it's technically possible to be out of bounds, in terms of navigation and the rest of the app
       * it's hard to predict what will happens with negative x/y. spine item will not be retrieved, the spine position might
       * not be within anything (although the viewport slice is).
       *
       * Spine position should ideally be within the spine (even when viewport is scaled down/up). That's why we have viewport
       * on top of the spine.
       *
       * For now having things centered (negative x) on zoom out in scroll mode can be achieved with transform origin for example.
       * This "limitation" is here at the moment to avoid unexpected behaviors.
       *
       * @note
       * Has a bug where scaling from < 1 to 1 was creating positive x offset. This is "expected" since on scroll mode the viewport
       * is at the offset 0 at scale 0.2 for eg: then when calculating new scroll delta, we get positive x offset. Anyway, to prevent
       * out of bounds position this should make sure we always stay within an item.
       */
      position: n.position ? t.fromOutOfBoundsSpinePosition(
        n.position
      ) : void 0
    };
    return {
      previousNavigation: i,
      navigation: o
    };
  })
), Pr = ({ navigationResolver: t }) => (e) => e.pipe(
  y((n) => {
    if (n.navigation.cfi) {
      const i = t.getNavigationForCfi(
        n.navigation.cfi
      );
      if (i)
        return {
          ...n,
          navigation: {
            ...n.navigation,
            position: i
          }
        };
    }
    return n;
  })
), xr = ({
  navigation: t,
  previousNavigation: e,
  settings: n
}) => t.directionFromLastNavigation ? t.directionFromLastNavigation : t.url !== void 0 || t.cfi !== void 0 ? "anchor" : e.spineItem === void 0 || t.spineItem || !t.position ? "forward" : n.values.computedPageTurnDirection === "vertical" ? t.position.y > e.position.y || t.position.y === e.position.y && e.directionFromLastNavigation !== "backward" ? "forward" : "backward" : Math.abs(t.position.x) > Math.abs(e.position.x) || t.position.x === e.position.x && e.directionFromLastNavigation !== "backward" ? "forward" : "backward", $r = ({
  context: t,
  settings: e
}) => (n) => n.pipe(
  y(({ navigation: i, previousNavigation: o }) => {
    const r = xr({
      navigation: i,
      previousNavigation: o,
      settings: e
    }), s = {
      ...i,
      directionFromLastNavigation: r
    };
    return {
      previousNavigation: o,
      navigation: s,
      direction: r
    };
  })
), Tr = ({
  spineItemsManager: t,
  navigationResolver: e,
  settings: n
}) => (i) => i.pipe(
  y(({ navigation: o, ...r }) => {
    const s = t.get(o.spineItem);
    if (o.position)
      return n.values.computedPageTurnMode === "scrollable" ? {
        navigation: {
          ...o,
          position: o.position
        },
        ...r
      } : {
        navigation: {
          ...o,
          position: e.fromUnboundSpinePosition(
            o.position
          )
        },
        ...r
      };
    if (!s)
      return {
        navigation: {
          ...o,
          position: r.previousNavigation.position
        },
        ...r
      };
    const a = e.getNavigationForSpineIndexOrId(s), c = n.values.computedPageTurnMode === "scrollable" ? new ce({
      x: a.x + r.previousNavigation.position.x,
      y: a.y
    }) : e.fromUnboundSpinePosition(a);
    return {
      navigation: {
        ...o,
        position: c
      },
      ...r
    };
  })
), Ft = ({
  settings: t,
  spineItemsManager: e,
  navigationResolver: n,
  spineLocator: i
}) => (o) => {
  const r = (s) => {
    const {
      position: a,
      spineItem: c,
      cfi: d,
      directionFromLastNavigation: u
    } = s, { navigationSnapThreshold: l, computedPageTurnMode: g } = t.values;
    if (c !== void 0) {
      const p = e.get(c);
      if (p) return p;
    }
    if (typeof c == "number")
      return c > e.items.length - 1 ? e.get(e.items.length - 1) : e.get(0);
    if (d) {
      const p = e.getSpineItemFromCfi(d);
      if (p) return p;
    }
    if (a && g === "controlled") {
      const { beginIndex: p, endIndex: h } = i.getVisibleSpineItemsFromPosition({
        position: a,
        threshold: l,
        restrictToScreen: !1
      }) ?? {}, m = (u === "forward" || u === "anchor" ? h : p) ?? p, f = e.get(m);
      if (!f) return;
      const { endPageIndex: S, beginPageIndex: I } = i.getVisiblePagesFromViewportPosition({
        position: a,
        spineItem: f,
        threshold: l,
        restrictToScreen: !1
      }) ?? {}, b = (u === "forward" || u === "anchor" ? S : I) ?? 0, F = n.getNavigationForSpineItemPage({
        pageIndex: b,
        spineItemId: f
      }), $ = i.getVisibleSpineItemsFromPosition({
        position: F,
        threshold: l,
        restrictToScreen: !1
      }), C = u === "forward" || u === "anchor" ? $?.beginIndex : $?.endIndex;
      return e.get(C);
    }
    return a && g === "scrollable" ? i.getSpineItemFromPosition(a) : e.get(0);
  };
  return o.pipe(
    y(({ navigation: s, ...a }) => {
      const c = r(s);
      return {
        navigation: {
          ...s,
          spineItem: e.getSpineItemIndex(c)
        },
        ...a
      };
    })
  );
}, Ct = ({ spine: t }) => (e) => e.pipe(
  P(({ navigation: n, ...i }) => {
    const o = t.getSpineItemSpineLayoutInfo(
      n.spineItem
    ), r = t.spineItemsManager.get(n.spineItem);
    return (r?.isReady$ ?? v(!1)).pipe(
      j(),
      y(
        (s) => ({
          navigation: {
            ...n,
            spineItemHeight: o?.height,
            spineItemWidth: o?.width,
            spineItemLeft: o.left,
            spineItemTop: o.top,
            spineItemIsUsingVerticalWriting: r?.isUsingVerticalWriting(),
            spineItemIsReady: s
          },
          ...i
        })
      )
    );
  })
), ze = ({
  settings: t,
  spineItemsManager: e,
  spineLocator: n,
  navigationResolver: i
}) => (o) => {
  const r = (s) => {
    const { navigationSnapThreshold: a, computedPageTurnMode: c } = t.values, d = e.get(s.spineItem);
    if (!(!d || !s.position)) {
      if (c === "controlled") {
        const { endPageIndex: u, beginPageIndex: l } = n.getVisiblePagesFromViewportPosition({
          position: s.position,
          spineItem: d,
          threshold: a,
          restrictToScreen: !1
        }) ?? {}, g = (s.directionFromLastNavigation === "forward" || s.directionFromLastNavigation === "anchor" ? u : l) ?? 0, p = i.getNavigationForSpineItemPage({
          pageIndex: g,
          spineItemId: d
        }), h = n.getVisiblePagesFromViewportPosition({
          position: p,
          spineItem: d,
          threshold: { type: "percentage", value: 0 },
          restrictToScreen: !0
        }), m = (s.directionFromLastNavigation === "forward" || s.directionFromLastNavigation === "anchor" ? h?.beginPageIndex : h?.endPageIndex) ?? 0;
        return n.spineItemLocator.getSpineItemPositionFromPageIndex({
          pageIndex: m,
          spineItem: d
        });
      }
      return n.getSpineItemPositionFromSpinePosition(
        s.position,
        d
      );
    }
  };
  return o.pipe(
    y(({ navigation: s, ...a }) => ({
      navigation: {
        ...s,
        positionInSpineItem: r(s)
      },
      ...a
    }))
  );
}, Fr = ({ navigationResolver: t }) => (e) => e.pipe(
  y((n) => {
    if (n.navigation.url) {
      const i = t.getNavigationForUrl(
        n.navigation.url
      );
      if (i)
        return {
          ...n,
          navigation: {
            ...n.navigation,
            position: i.position,
            spineItem: i.spineItemId
          }
        };
    }
    return n;
  })
);
class tn {
  constructor() {
    this.isLockedSubject = new Y(0), this.isLocked$ = this.isLockedSubject.pipe(
      y((e) => !!e),
      D()
    );
  }
  lock() {
    let e = !1;
    return this.isLockedSubject.next(this.isLockedSubject.getValue() + 1), () => {
      e || (e = !0, this.isLockedSubject.next(this.isLockedSubject.getValue() - 1));
    };
  }
}
const Cr = ({
  spineLocator: t,
  navigation: e,
  navigationResolver: n,
  spineItemsManager: i,
  spine: o
}) => {
  const r = i.get(e.spineItem);
  return r ? r.isReady$.pipe(
    j(),
    y((s) => {
      const a = o.getSpineItemSpineLayoutInfo(r), c = t.isPositionWithinSpineItem(
        e.position,
        r
      ), d = a.width - (e.spineItemWidth ?? 0), u = a.height - (e.spineItemHeight ?? 0), l = d !== 0 || u !== 0;
      if (e.url !== void 0 && (d || u || // when spine item is ready dimensions may have not changed but the position
      // of dom elements may have!
      s && !e.spineItemIsReady)) {
        const p = n.getNavigationForUrl(
          e.url
        );
        if (p)
          return p.position;
      }
      const g = e.cfi ?? e.paginationBeginCfi;
      if (g !== void 0 && !Ze(g) && (d || u || // when spine item is ready dimensions may have not changed but the position
      // of dom elements may have!
      s && !e.spineItemIsReady)) {
        const p = n.getNavigationForCfi(g);
        if (p)
          return p;
      }
      if (c && l && e.directionFromLastNavigation === "backward") {
        const p = new N({
          x: (e.positionInSpineItem?.x ?? 0) + d,
          y: (e.positionInSpineItem?.y ?? 0) + u
        });
        return n.getNavigationFromSpineItemPosition({
          spineItem: r,
          spineItemPosition: p
        });
      }
      if (e.positionInSpineItem && e.spineItemHeight && e.spineItemWidth) {
        const p = t.spineItemLocator.getSpineItemPageIndexFromPosition({
          itemWidth: e.spineItemWidth,
          itemHeight: e.spineItemHeight,
          isUsingVerticalWriting: !!e.spineItemIsUsingVerticalWriting,
          position: e.positionInSpineItem
        });
        return n.getNavigationForSpineItemPage({
          pageIndex: p,
          spineItemId: r
        });
      }
      return c ? n.getNavigationForPosition(e.position) : n.getNavigationForSpineIndexOrId(r);
    })
  ) : v(new L({ x: 0, y: 0 }));
}, Lr = ({
  navigation: t,
  spineLocator: e,
  spineItemsManager: n,
  settings: i,
  navigationResolver: o,
  spine: r
}) => {
  const { spineItem: s } = t, a = n.get(s);
  if (!a) return new L({ x: 0, y: 0 });
  const { height: c, top: d } = r.getSpineItemSpineLayoutInfo(a), u = e.isPositionWithinSpineItem(
    t.position,
    a
  ), l = t.positionInSpineItem ?? new N({
    y: 0,
    x: 0
  });
  if (i.values.computedPageTurnDirection === "vertical") {
    if (d === t.spineItemTop && c === t.spineItemHeight && u)
      return t.position;
    if (d === t.spineItemTop && c === t.spineItemHeight && !u)
      return o.getNavigationForSpineIndexOrId(a);
    if (d !== t.spineItemTop) {
      const g = e.getSafeSpineItemPositionFromUnsafeSpineItemPosition(
        t.positionInSpineItem ?? new N({
          x: 0,
          y: 0
        }),
        a
      );
      return e.getSpinePositionFromSpineItemPosition({
        spineItemPosition: g,
        spineItem: a
      });
    }
    if (d === t.spineItemTop && c !== t.spineItemHeight) {
      const g = (t.spineItemHeight ?? l.y) - l.y, p = new N({
        y: t.directionFromLastNavigation === "backward" ? c - g : l.y,
        x: t.position.x
      });
      if (u) {
        const h = e.getSafeSpineItemPositionFromUnsafeSpineItemPosition(
          p,
          a
        );
        return e.getSpinePositionFromSpineItemPosition({
          spineItemPosition: h,
          spineItem: a
        });
      }
      if (!u) {
        if (!(t.position.y < d)) {
          const m = new N({
            y: c - g,
            x: t.position.x
          });
          return e.getSpinePositionFromSpineItemPosition({
            spineItemPosition: e.getSafeSpineItemPositionFromUnsafeSpineItemPosition(
              m,
              a
            ),
            spineItem: a
          });
        }
        return o.getNavigationForSpineIndexOrId(a);
      }
    }
  }
  return t.position;
}, Mr = ({
  navigation: t,
  spineItemsManager: e,
  settings: n,
  spineLocator: i,
  navigationResolver: o,
  spine: r
}) => n.values.computedPageTurnMode === "scrollable" ? v(
  Lr({
    navigation: t,
    spineLocator: i,
    navigationResolver: o,
    settings: n,
    spineItemsManager: e,
    spine: r
  })
) : Cr({
  navigation: t,
  spineLocator: i,
  navigationResolver: o,
  spineItemsManager: e,
  spine: r
}), Lt = ({
  settings: t,
  navigationResolver: e,
  context: n,
  spine: i
}) => (o) => o.pipe(
  P((r) => Mr({
    spineLocator: i.locator,
    navigation: r.navigation,
    navigationResolver: e,
    settings: t,
    spineItemsManager: i.spineItemsManager,
    spineItemLocator: i.locator.spineItemLocator,
    spine: i
  }).pipe(
    y((s) => ({
      ...r,
      navigation: {
        ...r.navigation,
        position: s
      }
    }))
  ))
), Er = "navigation/InternalNavigator", Nr = E.namespace(Er);
class Ar extends q {
  constructor(e, n, i, o, r, s, a, c) {
    super(), this.settings = e, this.context = n, this.userNavigation$ = i, this.controlledNavigationController = o, this.scrollNavigationController = r, this.navigationResolver = s, this.spine = a, this.isRestorationLocked$ = c, this.navigationSubject = new Y({
      animation: !1,
      position: new L({ x: 0, y: 0 }),
      meta: {
        triggeredBy: "user"
      },
      spineItemIsReady: !1,
      type: "api",
      id: /* @__PURE__ */ Symbol()
    }), this.navigated$ = this.navigationSubject.pipe(Rt(1)), this.navigation$ = this.navigationSubject.pipe(
      y(({ position: S, id: I }) => ({
        position: S,
        id: I
      })),
      D(
        ({ position: S, ...I }, { position: b, ...F }) => V(I, F) && V(S, b)
      ),
      Z(1)
    ), this.locker = new tn();
    const d = i.pipe(
      z(this.navigationSubject),
      wr({
        navigationResolver: s
      }),
      /**
       * Url lookup is heavier so we start with it to fill
       * as much information as needed to reduce later lookup
       */
      Fr({
        navigationResolver: s
      }),
      /**
       * Cfi lookup is heavier so we start with it to fill
       * as much information as needed to reduce later lookup
       */
      Pr({
        navigationResolver: s
      }),
      $r({ context: n, settings: e }),
      Ft({
        navigationResolver: s,
        settings: e,
        spineItemsManager: a.spineItemsManager,
        spineLocator: a.locator
      }),
      ze({
        navigationResolver: s,
        settings: e,
        spineItemsManager: a.spineItemsManager,
        spineLocator: a.locator
      }),
      Ct({
        spine: a
      })
    ).pipe(
      Tr({
        navigationResolver: s,
        spineItemsManager: a.spineItemsManager,
        settings: e
      }),
      z(c),
      P(([S, I]) => {
        const b = S.navigation.cfi || S.navigation.url || e.values.computedPageTurnMode === "scrollable" || I;
        return v(S).pipe(
          b ? Ce : Lt({
            navigationResolver: s,
            settings: e,
            spine: a,
            context: n
          })
        );
      }),
      ze({
        spineItemsManager: a.spineItemsManager,
        spineLocator: a.locator,
        settings: e,
        navigationResolver: s
      }),
      y((S) => S.navigation),
      A()
    ), u = d.pipe(
      z(c),
      M(([, S]) => S),
      P(([S]) => {
        const I = this.locker.lock();
        return c.pipe(
          M((b) => !b),
          j(),
          y(
            () => ({
              ...S,
              animation: "snap"
            })
          ),
          ye(() => {
            I();
          }),
          T(d)
        );
      }),
      A()
    ), l = w(
      o.layout$,
      a.layout$
    ).pipe(
      P(() => v(null).pipe(
        P(
          () => c.pipe(
            M((S) => !S),
            j()
          )
        ),
        y(
          () => ({
            ...this.navigationSubject.getValue(),
            animation: !1
          })
        ),
        /**
         * We need to cancel the restoration as soon as there is
         * another navigation. Whether it's user or internal, it means
         * it has been controlled outside.
         */
        T(
          w(u, d)
        )
      ))
    ), g = w(
      l,
      u
    ).pipe(
      y((S) => ({ navigation: S })),
      Lt({
        navigationResolver: s,
        settings: e,
        context: n,
        spine: a
      }),
      y((S) => {
        const I = {
          ...S.navigation,
          meta: {
            triggeredBy: "restoration"
          }
        };
        return {
          ...S,
          navigation: I
        };
      }),
      /**
       * The spine item may be undefined after a restoration.
       * eg: after the reader load and the user has never navigated
       * yet.
       */
      Ft({
        navigationResolver: s,
        settings: e,
        spineItemsManager: a.spineItemsManager,
        spineLocator: a.locator
      }),
      Ct({
        spine: a
      }),
      ze({
        spineItemsManager: a.spineItemsManager,
        spineLocator: a.locator,
        settings: e,
        navigationResolver: s
      }),
      y(({ navigation: S }) => S),
      A()
    ), p = vr(
      n,
      this.navigationSubject,
      a
    ), h = w(
      g,
      d,
      p
    ), m = (S) => S.pipe(
      x(([I, b]) => {
        Nr.info(
          `navigation updated from ${I.meta.triggeredBy} of type ${I.type}`,
          {
            previousNavigation: b,
            currentNavigation: I
          }
        ), this.navigationSubject.next(I);
      })
    ), f = (S) => S.pipe(
      x(([I, b]) => {
        const F = I.type === "scroll", $ = I.meta.triggeredBy === "pagination", C = I.meta.triggeredBy === "restoration", W = V(
          b.position,
          I.position
        );
        if (F && !C || $ || W)
          return;
        const _ = {
          position: I.position,
          animation: I.animation
        };
        e.values.computedPageTurnMode === "scrollable" ? this.scrollNavigationController.navigate(_) : this.controlledNavigationController.navigate({
          ..._,
          /**
           * @important
           * Explicitly trust the unbound position to be valid at this point.
           * Thanks to consolidation and restoration.
           */
          position: L.from(_.position)
        });
      })
    );
    h.pipe(
      z(this.navigationSubject),
      /**
       * @important
       *
       * We need to start navigating viewport before notifying navigation
       * change, this is to keep viewportState sync and avoid a "free" ping
       * in between.
       */
      f,
      m,
      T(this.destroy$)
    ).subscribe();
  }
  get navigation() {
    return this.navigationSubject.getValue();
  }
}
const Ne = (t, e, n) => {
  const i = n - t, o = n * (e * t) / (n || 1);
  return Math.max(0, Math.min(i, o));
}, Je = (t, e) => (e || 0) === 0 || (t || 0) === 0 ? 1 : Math.floor(Math.max(1, t / e)), We = (t, e, n) => {
  const i = Je(n, e), o = [...Array(i)].map((r, s) => s * e);
  return t >= i * e ? o[o.length - 1] || 0 : o.find((r) => t < r + e) || 0;
}, Mt = (t, e, n) => {
  const i = [...Array(n)].map((o, r) => r * e);
  return t <= 0 || n === 0 ? 0 : t >= n * e ? n - 1 : i.findIndex((o) => t < o + e) ?? 0;
}, Dr = ({
  itemWidth: t,
  itemHeight: e,
  spineItemPosition: n
}) => new N({
  x: Math.min(t, Math.max(0, n.x)),
  y: Math.min(e, Math.max(0, n.y))
}), nn = ({
  itemHeight: t,
  itemWidth: e,
  isUsingVerticalWriting: n,
  pageWidth: i,
  pageHeight: o,
  pageTurnDirection: r,
  pageTurnMode: s
}) => r === "vertical" && s === "scrollable" ? 1 : n || r === "vertical" ? Je(t, o) : Je(e, i), Et = ({
  itemWidth: t,
  itemHeight: e,
  position: n,
  isUsingVerticalWriting: i,
  pageWidth: o,
  pageHeight: r,
  pageTurnDirection: s,
  pageTurnMode: a,
  isRTL: c
}) => {
  const u = Dr({
    spineItemPosition: n,
    itemHeight: e,
    itemWidth: t
  }).x, l = nn({
    isUsingVerticalWriting: i,
    itemHeight: e,
    itemWidth: t,
    pageWidth: o,
    pageHeight: r,
    pageTurnDirection: s,
    pageTurnMode: a
  });
  if (i)
    return Mt(n.y, r, l);
  const g = Mt(u, o, l);
  return c ? l - 1 - g : g;
}, Rr = ({
  pageIndex: t,
  itemLayout: e,
  context: n,
  isUsingVerticalWriting: i,
  viewport: o
}) => {
  if (i) {
    const s = Ne(
      o.pageSize.height,
      t,
      e.height
    );
    return new N({
      x: 0,
      y: s
    });
  }
  const r = Ne(
    o.pageSize.width,
    t,
    e.width
  );
  return n.isRTL() ? new N({
    x: e.width - r - o.pageSize.width,
    y: 0
  }) : new N({
    x: r,
    y: 0
  });
}, on = ({
  context: t,
  settings: e,
  viewport: n
}) => {
  const i = (a, c, d) => {
    let u;
    a?.nodeName === "img" || a?.textContent === "" && a.nodeType === Node.ELEMENT_NODE ? u = a.getBoundingClientRect().x : a && (u = (a ? Un(a, c) : void 0)?.getBoundingClientRect().x || u);
    const l = d.layoutInfo?.width || 0, g = n.pageSize.width;
    if (u !== void 0) {
      const p = We(
        u,
        g,
        l
      );
      return new N({ x: p, y: 0 });
    }
  };
  return {
    getSpineItemPositionFromNode: i,
    getSpineItemPositionFromPageIndex: ({
      pageIndex: a,
      spineItem: c
    }) => Rr({
      context: t,
      isUsingVerticalWriting: !!c.isUsingVerticalWriting(),
      itemLayout: c.layoutInfo,
      pageIndex: a,
      viewport: n
    }),
    getSpineItemPageIndexFromPosition: (a) => Et({
      ...a,
      isRTL: t.isRTL(),
      pageWidth: n.pageSize.width,
      pageHeight: n.pageSize.height,
      pageTurnDirection: e.values.computedPageTurnDirection,
      pageTurnMode: e.values.pageTurnMode
    }),
    getSpineItemPageIndexFromNode: (a, c, d) => {
      const u = i(a, c, d), { height: l, width: g } = d.layoutInfo;
      return u ? Et({
        isUsingVerticalWriting: !!d.isUsingVerticalWriting(),
        position: u,
        itemHeight: l,
        itemWidth: g,
        isRTL: t.isRTL(),
        pageWidth: n.pageSize.width,
        pageHeight: n.pageSize.height,
        pageTurnDirection: e.values.computedPageTurnDirection,
        pageTurnMode: e.values.pageTurnMode
      }) : void 0;
    },
    getSpineItemClosestPositionFromUnsafePosition: (a, c) => {
      const { width: d, height: u } = c.layoutInfo;
      return new N({
        x: We(
          a.x,
          n.pageSize.width,
          d
        ),
        y: We(
          a.y,
          n.pageSize.height,
          u
        )
      });
    },
    getSpineItemPagePositionFromSpineItemPosition: (a, c, d) => {
      const { width: u, height: l } = d.layoutInfo, g = n.pageSize.width, p = n.pageSize.height;
      if (!!d.isUsingVerticalWriting()) {
        const f = Ne(
          p,
          c,
          l
        );
        return new ke({
          x: a.x,
          y: a.y - f
        });
      }
      const m = Ne(g, c, u);
      if (t.isRTL()) {
        const f = u - (c + 1) * g;
        return new ke({
          x: a.x - Math.max(0, f),
          y: a.y
        });
      }
      return new ke({
        x: a.x - m,
        y: a.y
      });
    }
  };
}, Or = ({
  context: t,
  settings: e,
  viewport: n
}) => {
  const i = on({
    context: t,
    settings: e,
    viewport: n
  });
  return {
    getNavigationForLastPage: (a) => {
      const c = a.numberOfPages;
      return i.getSpineItemPositionFromPageIndex({
        pageIndex: c - 1,
        spineItem: a
      });
    },
    getNavigationForPosition: (a, c) => i.getSpineItemClosestPositionFromUnsafePosition(
      c,
      a
    ),
    getNavigationFromNode: (a, c, d) => i.getSpineItemPositionFromNode(
      c,
      d,
      a
    ) || new N({ x: 0, y: 0 })
  };
};
function rn({
  position: t,
  isRTL: e,
  spineItemsManager: n,
  spine: i,
  viewportWidth: o
}) {
  const r = n.get(
    n.items.length - 1
  ), s = i.getSpineItemSpineLayoutInfo(
    r || 0
  ), a = s.bottom - 1, c = Math.max(0, t.y), d = Math.min(c, a);
  if (e) {
    const p = Math.min(o, t.x), h = Math.max(p, s.left);
    return new L({
      x: h,
      y: d
    });
  }
  const u = s.right - 1, l = Math.max(0, t.x), g = Math.min(l, u);
  return new L({
    x: g,
    y: d
  });
}
const Nt = ({
  position: t,
  isRTL: e,
  pageSizeHeight: n,
  spineItemsManager: i,
  visibleAreaRectWidth: o,
  spine: r
}) => {
  const s = rn({
    position: t,
    isRTL: e,
    spineItemsManager: i,
    spine: r,
    viewportWidth: o
  }), a = i.get(
    i.items.length - 1
  ), c = r.getSpineItemSpineLayoutInfo(a), d = c.bottom - n, u = Math.min(s.y, d);
  if (e) {
    const g = Math.min(0, s.x);
    return new L({
      x: g,
      y: u
    });
  }
  const l = c.right - o;
  return new L({
    x: Math.min(s.x, l),
    y: u
  });
}, ie = ({
  position: { x: t, y: e },
  pageSizeWidth: n,
  visibleAreaRectWidth: i
}) => {
  const r = t % i !== 0 ? t - n : t;
  return new L({ x: r, y: e });
}, Ke = ({
  viewportPosition: t,
  spineLocator: e,
  spineItemNavigationResolver: n,
  viewport: i
}) => {
  const o = e.getSpineItemFromPosition(t);
  if (o) {
    const r = e.getSpineItemPositionFromSpinePosition(
      t,
      o
    ), s = n.getNavigationForPosition(
      o,
      r
    ), a = e.getSpinePositionFromSpineItemPosition({
      spineItemPosition: s,
      spineItem: o
    });
    return ie({
      position: a,
      pageSizeWidth: i.pageSize.width,
      visibleAreaRectWidth: i.absoluteViewport.width
    });
  }
  return new L({ x: 0, y: 0 });
}, Vr = ({
  pageIndex: t,
  spineItemsManager: e,
  spineItemId: n,
  spineLocator: i,
  spineItemNavigationResolver: o,
  viewport: r
}) => {
  const s = e.get(n);
  if (!s) {
    const d = t * r.pageSize.width;
    return Ke({
      viewportPosition: new L({ x: d, y: 0 }),
      spineItemNavigationResolver: o,
      spineLocator: i,
      viewport: r
    });
  }
  const a = i.spineItemLocator.getSpineItemPositionFromPageIndex({
    pageIndex: t,
    spineItem: s
  }), c = i.getSpinePositionFromSpineItemPosition({
    spineItemPosition: a,
    spineItem: s
  });
  return ie({
    position: c,
    pageSizeWidth: r.pageSize.width,
    visibleAreaRectWidth: r.absoluteViewport.width
  });
}, kr = (t, e) => {
  if (e && e instanceof HTMLIFrameElement) {
    const n = e.contentDocument, i = t.trim();
    if (!n || i.length === 0)
      return;
    if (i.startsWith("#")) {
      const o = i.slice(1);
      return o.length === 0 ? void 0 : n.getElementById(o);
    }
    try {
      return n.querySelector(i);
    } catch {
      return;
    }
  }
}, jr = ({
  anchor: t,
  spineItem: e,
  spine: n
}) => {
  const i = kr(
    t,
    e.renderer.getDocumentFrame()
  );
  return i ? n.spineItemLocator.getSpineItemPositionFromNode(i, 0, e)?.x ?? 0 : 0;
}, Ur = ({
  anchor: t,
  spineItem: e,
  spineLocator: n,
  spine: i
}) => {
  const o = jr({
    anchor: t,
    spineItem: e,
    spine: i
  });
  return n.getSpinePositionFromSpineItemPosition({
    spineItemPosition: new N({ x: o, y: 0 }),
    spineItem: e
  });
}, zr = ({
  anchor: t,
  spineItem: e,
  spineLocator: n,
  spine: i,
  pageSizeWidth: o,
  visibleAreaRectWidth: r
}) => {
  const s = Ur({
    anchor: t,
    spineItem: e,
    spineLocator: n,
    spine: i
  });
  return ie({
    position: s,
    pageSizeWidth: o,
    visibleAreaRectWidth: r
  });
}, Wr = ({
  spine: t,
  spineItemsManager: e,
  spineLocator: n,
  url: i,
  context: o,
  pageSizeWidth: r,
  visibleAreaRectWidth: s
}) => {
  try {
    const a = i instanceof URL ? i : new URL(i), c = `${a.origin}${a.pathname}`, d = o.manifest?.spineItems.find(
      (u) => u.href === c
    );
    if (d) {
      const u = e.get(d.id);
      if (u) {
        const l = zr({
          anchor: a.hash,
          spineItem: u,
          spine: t,
          spineLocator: n,
          pageSizeWidth: r,
          visibleAreaRectWidth: s
        });
        return {
          position: ie({
            position: l,
            pageSizeWidth: r,
            visibleAreaRectWidth: s
          }),
          spineItemId: d.id
        };
      }
    }
    return;
  } catch (a) {
    console.error(a);
    return;
  }
}, _r = ({
  spineItem: t,
  spineItemPosition: e,
  spineLocator: n,
  spineItemLocator: i,
  viewport: o
}) => {
  const r = i.getSpineItemClosestPositionFromUnsafePosition(
    e,
    t
  ), s = n.getSpinePositionFromSpineItemPosition({
    spineItemPosition: r,
    spineItem: t
  });
  return ie({
    position: s,
    pageSizeWidth: o.pageSize.width,
    visibleAreaRectWidth: o.absoluteViewport.width
  });
}, Hr = "spineNavigator", Br = ({
  context: t,
  spineItemsManager: e,
  locator: n,
  settings: i,
  spine: o,
  viewport: r
}) => {
  const s = Or({
    context: t,
    settings: i,
    viewport: r
  });
  return {
    getNavigationForUrl: (p) => Wr({
      context: t,
      spineItemsManager: e,
      spineLocator: n,
      url: p,
      pageSizeWidth: r.pageSize.width,
      visibleAreaRectWidth: r.absoluteViewport.width,
      spine: o
    }),
    getNavigationForSpineItemPage: (p) => Vr({
      ...p,
      spineItemsManager: e,
      spineItemNavigationResolver: s,
      spineLocator: n,
      viewport: r
    }),
    getNavigationFromSpineItemPosition: (p) => _r({
      ...p,
      spineItemLocator: n.spineItemLocator,
      spineLocator: n,
      viewport: r
    }),
    getNavigationForCfi: (p) => {
      const h = e.getSpineItemFromCfi(p), { node: m, offset: f = 0 } = Zt({
        cfi: p,
        spineItemsManager: e
      });
      if (!h) {
        E.warn(Hr, `unable to detect item id from cfi ${p}`);
        return;
      }
      const S = m ? s.getNavigationFromNode(h, m, f) : new N({ x: 0, y: 0 }), I = n.getSpinePositionFromSpineItemPosition({
        spineItemPosition: S,
        spineItem: h
      });
      return ie({
        position: I,
        pageSizeWidth: r.pageSize.width,
        visibleAreaRectWidth: r.absoluteViewport.width
      });
    },
    getNavigationForLastPage: (p) => {
      const h = s.getNavigationForLastPage(p), m = n.getSpinePositionFromSpineItemPosition({
        spineItemPosition: h,
        spineItem: p
      });
      return ie({
        position: m,
        pageSizeWidth: r.pageSize.width,
        visibleAreaRectWidth: r.absoluteViewport.width
      });
    },
    getNavigationForSpineIndexOrId: (p) => {
      const h = e.get(p);
      if (h) {
        const m = n.getSpinePositionFromSpineItem(h);
        return ie({
          position: m,
          pageSizeWidth: r.pageSize.width,
          visibleAreaRectWidth: r.absoluteViewport.width
        });
      }
      return new L({ x: 0, y: 0 });
    },
    getNavigationForPosition: (p) => Ke({
      viewportPosition: p,
      spineItemNavigationResolver: s,
      spineLocator: n,
      viewport: r
    }),
    getMostPredominantNavigationForPosition: (p) => {
      const h = i.values.computedPageTurnDirection, m = 0.5, f = h === "horizontal" ? p.x + r.absoluteViewport.width * m : 0, S = h === "horizontal" ? 0 : p.y + r.absoluteViewport.height * m, I = Nt({
        position: new L({
          x: f,
          y: S
        }),
        isRTL: t.isRTL(),
        pageSizeHeight: r.pageSize.height,
        visibleAreaRectWidth: r.absoluteViewport.width,
        spineItemsManager: e,
        spine: o
      });
      return Ke({
        spineItemNavigationResolver: s,
        spineLocator: n,
        viewportPosition: I,
        viewport: r
      });
    },
    fromUnboundSpinePosition: (p) => Nt({
      position: p,
      isRTL: t.isRTL(),
      pageSizeHeight: r.pageSize.height,
      visibleAreaRectWidth: r.absoluteViewport.width,
      spineItemsManager: e,
      spine: o
    }),
    fromOutOfBoundsSpinePosition: (p) => rn({
      position: p,
      isRTL: t.isRTL(),
      spineItemsManager: e,
      spine: o,
      viewportWidth: r.absoluteViewport.width
    }),
    isNavigationGoingForwardFrom: (p, h) => i.values.computedPageTurnDirection === "vertical" ? p.y > h.y : p.x > h.x,
    arePositionsDifferent: (p, h) => p.x !== h.x || p.y !== h.y,
    getAdjustedPositionForSpread: (p) => ie({
      position: p,
      pageSizeWidth: r.pageSize.width,
      visibleAreaRectWidth: r.absoluteViewport.width
    }),
    spineItemNavigator: s
  };
}, Xr = ({
  spineItemsManager: t,
  context: e,
  hookManager: n,
  spine: i,
  settings: o,
  viewport: r
}) => {
  const s = new R(), a = s.asObservable(), c = new tn(), d = Br({
    context: e,
    settings: o,
    spineItemsManager: t,
    locator: i.locator,
    spine: i,
    viewport: r
  }), u = new Ir(
    o,
    n,
    e,
    i,
    r
  ), l = new sr(
    r,
    o,
    n,
    i,
    e
  ), g = new Ar(
    o,
    e,
    a,
    u,
    l,
    d,
    i,
    c.isLocked$
  ), p = O([
    u.isNavigating$,
    l.isNavigating$,
    c.isLocked$,
    g.locker.isLocked$
  ]).pipe(
    oe((f) => f.some((S) => S) ? "busy" : "free"),
    Pe(),
    tt(1)
  );
  return {
    destroy: () => {
      u.destroy(), g.destroy();
    },
    getNavigation: () => g.navigation,
    internalNavigator: g,
    scrollNavigationController: l,
    controlledNavigationController: u,
    locker: c,
    navigationState$: p,
    navigate: (f) => {
      E.info("User navigation", f), s.next(f);
    },
    /**
     * Prevent further navigation until the lock is released.
     * Useful if you want to start navigation by panning for example.
     */
    lock() {
      return c.lock();
    },
    navigationResolver: d,
    navigation$: g.navigation$
  };
};
class Yr extends ee {
  constructor(e, n) {
    super({
      beginPageIndexInSpineItem: void 0,
      beginNumberOfPagesInSpineItem: 0,
      beginCfi: void 0,
      beginSpineItemIndex: void 0,
      endPageIndexInSpineItem: void 0,
      endNumberOfPagesInSpineItem: 0,
      endCfi: void 0,
      endSpineItemIndex: void 0,
      navigationId: void 0
    }), this.context = e, this.spineItemsManager = n;
  }
  update(e) {
    this.mergeCompare(e);
  }
}
class qr extends q {
  constructor(e, n, i, o, r) {
    super(), this.context = e, this.pagination = n, this.spineItemsManager = i, this.spine = o, this.spineItemLocator = r;
    const s = w(
      this.context.bridgeEvent.navigation$,
      o.layout$
    ).pipe(
      P(() => {
        const c = ({
          spineItem: d,
          position: u
        }) => this.spine.locator.getVisiblePagesFromViewportPosition({
          spineItem: d,
          position: u,
          threshold: { type: "percentage", value: 0.5 }
        });
        return this.context.bridgeEvent.navigationUnlocked$.pipe(
          Dt(1),
          z(this.context.bridgeEvent.navigation$),
          x(([, d]) => {
            const { position: u } = d, l = this.pagination.value, {
              beginIndex: g,
              endIndex: p
            } = this.spine.locator.getVisibleSpineItemsFromPosition({
              position: u,
              threshold: { type: "percentage", value: 0.5 }
            }) ?? {}, h = this.spineItemsManager.get(g), m = this.spineItemsManager.get(p);
            if (!h || !m) return;
            const f = l.beginCfi, S = l.endCfi, { beginPageIndex: I = 0 } = c({
              spineItem: h,
              position: u
            }) ?? {}, { endPageIndex: b = 0 } = c({
              spineItem: m,
              position: u
            }) ?? {}, F = f === void 0 || Ze(f) || l.beginSpineItemIndex !== g, $ = l.endSpineItemIndex !== p || S === void 0 || Ze(S), C = F ? le(h.item) : f, W = $ ? le(m.item) : S, _ = h.numberOfPages, ae = m.numberOfPages;
            this.pagination.update({
              beginCfi: C,
              beginNumberOfPagesInSpineItem: _,
              beginPageIndexInSpineItem: I,
              beginSpineItemIndex: g,
              endCfi: W,
              endNumberOfPagesInSpineItem: ae,
              endPageIndexInSpineItem: b,
              endSpineItemIndex: p,
              navigationId: d.id
            });
          })
        );
      })
    ), a = s.pipe(
      Re(this.context.bridgeEvent.viewportFree$),
      x(() => {
        const {
          beginSpineItemIndex: c,
          endSpineItemIndex: d,
          beginPageIndexInSpineItem: u,
          endPageIndexInSpineItem: l
        } = this.pagination.value;
        if (u === void 0 || l === void 0 || c === void 0 || d === void 0)
          return;
        const g = this.spineItemsManager.get(c), p = this.spineItemsManager.get(d);
        if (g === void 0 || p === void 0) return;
        const h = this.spine.pages.fromSpineItemPageIndex(
          g,
          u
        ), m = this.spine.pages.fromSpineItemPageIndex(
          p,
          l
        );
        this.pagination.update({
          beginCfi: h?.firstVisibleNode ? Ge({
            spineItem: g.item,
            pageNode: h?.firstVisibleNode
          }) : le(g.item),
          endCfi: m?.firstVisibleNode ? Ge({
            spineItem: p.item,
            pageNode: m?.firstVisibleNode
          }) : le(p.item)
        });
      })
    );
    w(s, a).pipe(T(this.destroy$)).subscribe();
  }
}
const Gr = ({
  manifest: t,
  spreadMode: e
}) => t?.renditionFlow === "scrolled-continuous" ? !1 : e;
class Zr extends q {
  constructor(e) {
    super();
    const n = {
      ...this.getDefaultSettings(),
      ...e
    };
    this.inputSettings = n, this.outputSettingsUpdateSubject = new R(), this._settings$ = this.outputSettingsUpdateSubject.asObservable().pipe(Z(1)), this._settings$.subscribe();
  }
  _prepareUpdate(e) {
    const n = Le(this.inputSettings, e), i = this.getOutputSettings(n), o = !V(this.outputSettings, i);
    return {
      hasChanged: o,
      state: i,
      commit: () => (this.inputSettings = n, this.outputSettings = i, o && this.outputSettingsUpdateSubject.next(i), i)
    };
  }
  update(e) {
    const { commit: n } = this._prepareUpdate(e);
    n();
  }
  get values() {
    if (!this.outputSettings) {
      const { commit: e } = this._prepareUpdate(this.inputSettings);
      return e();
    }
    return this.outputSettings;
  }
  get values$() {
    if (!this.outputSettings) {
      const { commit: e } = this._prepareUpdate(this.inputSettings);
      e();
    }
    return this._settings$;
  }
  watch(e) {
    return this.values$.pipe(
      Wt(e),
      D(V)
    );
  }
  destroy() {
    super.destroy(), this.outputSettingsUpdateSubject.complete();
  }
}
class Jr extends Zr {
  constructor(e, n) {
    super(e), this.context = n, n.watch(["manifest", "hasVerticalWriting"]).pipe(
      x(() => this.update(this.values)),
      T(this.destroy$)
    ).subscribe();
  }
  getComputedSettings(e) {
    const n = this.context.manifest, i = this.context.value.hasVerticalWriting ?? !1, o = {
      computedPageTurnDirection: e.pageTurnDirection,
      computedPageTurnAnimation: e.pageTurnAnimation,
      computedPageTurnMode: e.pageTurnMode,
      computedPageTurnAnimationDuration: 0,
      computedSpreadMode: Gr({
        spreadMode: e.spreadMode,
        manifest: n
      })
    };
    return n?.renditionFlow === "scrolled-continuous" && (o.computedPageTurnMode = "scrollable"), o.computedPageTurnMode === "scrollable" && (o.computedPageTurnDirection = "vertical"), i && o.computedPageTurnAnimation === "slide" && (E.warn(
      `pageTurnAnimation ${o.computedPageTurnAnimation} incompatible with current book, switching back to default`
    ), o.computedPageTurnAnimation = "none"), o.computedPageTurnMode === "scrollable" ? (o.computedPageTurnAnimationDuration = 0, o.computedPageTurnAnimation = "none") : o.computedPageTurnAnimationDuration = e.pageTurnAnimationDuration !== void 0 ? e.pageTurnAnimationDuration : 300, o;
  }
  getOutputSettings(e) {
    const n = this.getComputedSettings(e);
    return { ...this.outputSettings, ...e, ...n };
  }
  getDefaultSettings() {
    return {
      spreadMode: !1,
      pageTurnAnimation: "slide",
      pageTurnDirection: "horizontal",
      pageTurnAnimationDuration: void 0,
      pageTurnMode: "controlled",
      snapAnimationDuration: 300,
      navigationSnapThreshold: { type: "pixels", value: 40 },
      numberOfAdjacentSpineItemToPreLoad: 3
    };
  }
}
class Kr extends Ee {
  onUnload() {
    return X;
  }
  onCreateDocument() {
    return v(document.createElement("div"));
  }
  onLoadDocument() {
    return X;
  }
  onLayout() {
    return v(void 0);
  }
  onRenderHeadless() {
    return X;
  }
  getDocumentFrame() {
  }
}
class Qr extends q {
  constructor(e, n, i, o, r, s, a) {
    super(), this.item = e, this.containerElement = n, this.context = i, this.hookManager = o, this.renderer = r, this.settings = s, this.viewport = a, this.layoutTriggerSubject = new R(), this.lastLayout = null, this.computeLayoutInformation = ({
      isLastItem: d,
      horizontalOffset: u
    }) => {
      let l = this.viewport.value.pageSize.width, g = "none";
      const p = u % this.viewport.absoluteViewport.width === 0, h = this.context.manifest, m = en(h) ?? !1;
      if (this.settings.values.computedSpreadMode) {
        !m && this.renderer.renditionLayout === "reflowable" && !d && (l = this.viewport.value.pageSize.width * 2), !m && this.renderer.renditionLayout === "reflowable" && d && p && (l = this.viewport.value.pageSize.width * 2);
        const f = p && d && m;
        this.item.pageSpreadRight && p && !this.context.isRTL() ? (g = "before", l = this.viewport.value.pageSize.width * 2) : this.item.pageSpreadLeft && p && this.context.isRTL() ? (g = "before", l = this.viewport.value.pageSize.width * 2) : f && (this.context.isRTL() ? g = "before" : g = "after", l = this.viewport.value.pageSize.width * 2);
      }
      return {
        minimumWidth: l,
        blankPagePosition: g
      };
    }, this.adjustPositionOfElement = ({
      right: d,
      left: u,
      top: l
    }) => {
      d !== void 0 ? this.containerElement.style.right = `${d}px` : this.containerElement.style.removeProperty("right"), u !== void 0 ? this.containerElement.style.left = `${u}px` : this.containerElement.style.removeProperty("left"), l !== void 0 ? this.containerElement.style.top = `${l}px` : this.containerElement.style.removeProperty("top");
    }, this.updateContainerLayout = ({
      minimumWidth: d,
      edgeX: u,
      edgeY: l
    }) => (g) => g.pipe(
      y((p) => {
        const h = V(
          this.lastLayout?.pageSize,
          this.viewport.pageSize
        ) ? this.lastLayout : void 0, { width: m, height: f } = h ?? {}, { width: S = m, height: I = f } = p ?? {}, { width: b, height: F } = this.viewport.pageSize, $ = this.validateDimension(
          S ?? b,
          b,
          d
        ), C = this.settings.values.computedPageTurnMode === "scrollable" ? I ?? F : this.validateDimension(
          I ?? F,
          F,
          F
        );
        return this.lastLayout = {
          width: $,
          height: C,
          pageSize: this.viewport.pageSize
        }, this.containerElement.style.width = `${$}px`, this.containerElement.style.height = `${C}px`, this.settings.values.computedPageTurnDirection === "vertical" ? (this.adjustPositionOfElement({
          top: l,
          left: u
        }), { width: $, height: C }) : (this.adjustPositionOfElement(
          this.context.isRTL() ? { right: u, top: 0 } : { left: u, top: 0 }
        ), { width: $, height: C });
      })
    ), this.layout = (d) => {
      const u = ii(this.didLayout$.pipe(j()));
      return this.layoutTriggerSubject.next(d), u();
    };
    const c = this.layoutTriggerSubject.pipe(
      P(
        ({ spreadPosition: d, horizontalOffset: u, isLastItem: l, edgeX: g, edgeY: p }) => {
          const { blankPagePosition: h, minimumWidth: m } = this.computeLayoutInformation({ horizontalOffset: u, isLastItem: l });
          this.hookManager.execute("item.onBeforeLayout", void 0, {
            blankPagePosition: h,
            item: this.item,
            minimumWidth: m
          });
          const f = this.renderer.layout({
            blankPagePosition: h,
            minPageSpread: m / this.viewport.pageSize.width,
            minimumWidth: m,
            spreadPosition: d
          });
          return w(
            v({ type: "start" }),
            f.pipe(
              /**
               * Now that inner layout is done from the renderer, we adjust the dimensions
               * of the outer container element
               */
              this.updateContainerLayout({
                minimumWidth: m,
                edgeX: g,
                edgeY: p
              }),
              y((S) => (this.hookManager.execute("item.onAfterLayout", void 0, {
                blankPagePosition: h,
                item: this.item,
                minimumWidth: m
              }), {
                type: "end",
                data: S
              }))
            )
          );
        }
      ),
      A()
    );
    this.didLayout$ = c.pipe(
      M((d) => d.type === "end"),
      y((d) => d.data),
      A()
    );
  }
  validateDimension(e, n, i) {
    if (e <= 0) return i;
    const o = Math.max(e, i), s = Math.ceil(o / n) * n;
    return Math.max(s, n);
  }
  get layoutInfo() {
    const e = this.containerElement.style, n = e.width ? parseFloat(e.width) : 0, i = e.height ? parseFloat(e.height) : 0;
    return {
      width: n,
      height: i
    };
  }
}
class sn extends ee {
  constructor(e, n, i, o, r, s, a) {
    super({
      isLoaded: !1,
      isReady: !1,
      isDirty: !1,
      isError: !1,
      error: void 0
    }), this.item = e, this.parentElement = n, this.context = i, this.settings = o, this.hookManager = r, this.index = s, this.viewport = a, this.load = () => {
      this.renderer.load();
    }, this.unload = () => {
      this.renderer.unload();
    }, this.markDirty = () => {
      this.mergeCompare({
        isDirty: !0
      });
    }, this.destroy = () => {
      super.destroy(), this.containerElement.remove(), this.renderer.destroy();
    }, this.isUsingVerticalWriting = () => !!this.renderer.writingMode?.startsWith("vertical"), this.layout = (l) => this._layout.layout(l), this.containerElement = es(
      n,
      e,
      r
    ), n.appendChild(this.containerElement);
    const c = this.settings.values.getRenderer?.(e);
    this.resourcesHandler = new ct(e, this.settings);
    const d = {
      context: i,
      settings: o,
      hookManager: r,
      item: e,
      containerElement: this.containerElement,
      resourcesHandler: this.resourcesHandler,
      viewport: this.viewport
    };
    this.renderer = c ? c(d) : new Kr(d), this._layout = new Qr(
      e,
      this.containerElement,
      i,
      r,
      this.renderer,
      this.settings,
      this.viewport
    );
    const u = this.renderer.state$.pipe(
      k(({ state: l, error: g }) => {
        this.mergeCompare({
          isLoaded: l === "loaded",
          isError: l === "error",
          error: l === "error" ? g : void 0
        });
      })
    );
    this.didLayout$ = this._layout.didLayout$.pipe(
      k(() => {
        this.mergeCompare({
          isDirty: !1,
          isReady: this.renderer.state$.value.state === "loaded"
        });
      }),
      En()
    ), w(
      /**
       * @important
       * The order is important here. We want to ensure the state value
       * is set before dispatching the layout event. Elements reacting
       * to layout changes may rely on the state value to be updated.
       */
      u,
      this.didLayout$
    ).pipe(Q(this.destroy$)).subscribe();
  }
  /**
   * Renderer loaded + spine item layout done
   * Will switch back and forth every new layout.
   */
  get isReady$() {
    return this.watch("isReady");
  }
  get element() {
    return this.containerElement;
  }
  /**
   * @important
   * Do not use this value for layout and navigation. It will be in possible conflict
   * with the global reading direction. A book should not mix them anyway. A page can have
   * a different reading direction for style reason but it should be in theory pre-paginated.
   * For example an english page in a japanese manga. That's expected and will
   * be confined to a single page.
   */
  get readingDirection() {
    return this.renderer.readingDirection;
  }
  /**
   * Note that this is dispatched AFTER the state has been updated.
   */
  get loaded$() {
    return this.renderer.loaded$;
  }
  /**
   * Note that this is dispatched AFTER the state has been updated.
   */
  get unloaded$() {
    return this.renderer.unloaded$;
  }
  get renditionLayout() {
    return this.renderer.renditionLayout;
  }
  get layoutInfo() {
    return this._layout.layoutInfo;
  }
  get numberOfPages() {
    return nn({
      isUsingVerticalWriting: !!this.isUsingVerticalWriting(),
      itemHeight: this._layout.layoutInfo.height,
      itemWidth: this._layout.layoutInfo.width,
      pageWidth: this.viewport.pageSize.width,
      pageHeight: this.viewport.pageSize.height,
      pageTurnDirection: this.settings.values.computedPageTurnDirection,
      pageTurnMode: this.settings.values.pageTurnMode
    });
  }
}
const es = (t, e, n) => {
  const i = t.ownerDocument.createElement("div");
  return i.setAttribute(`data-${Ui}`, ""), i.classList.add("spineItem"), i.classList.add(`spineItem-${e.renditionLayout ?? "reflowable"}`), i.style.cssText = `
    position: absolute;
    overflow: hidden;
  `, i.dataset.isReady = "false", n.execute("item.onBeforeContainerCreated", void 0, { element: i }), i;
};
class ts extends q {
  constructor(e, n, i, o, r) {
    super(), this.context = e, this.spineItemsManager = n, this.spineLocator = i, this.settings = o, this.spineLayout = r, this.forcedOpenSubject = new Y(
      /* @__PURE__ */ new Map()
    );
    const s = this.forcedOpenSubject.pipe(
      y((d) => Array.from(d.keys()).sort((u, l) => u - l)),
      D(xn),
      Z({ bufferSize: 1, refCount: !0 })
    );
    w(
      this.context.bridgeEvent.navigation$,
      this.spineLayout.layout$,
      s,
      o.watch(["numberOfAdjacentSpineItemToPreLoad"])
    ).pipe(
      // this can be changed by whatever we want and SHOULD not break navigation.
      // Ideally loading faster is better but loading too close to user navigating can
      // be dangerous.
      De(100),
      Re(this.context.bridgeEvent.viewportFree$),
      z(this.context.bridgeEvent.navigation$, s),
      y(([, d, u]) => {
        const { numberOfAdjacentSpineItemToPreLoad: l } = o.values, { beginIndex: g = 0, endIndex: p = 0 } = i.getVisibleSpineItemsFromPosition({
          position: d.position,
          threshold: { type: "percentage", value: 0 },
          useAbsoluteViewport: !1
        }) || {}, h = l === 1 / 0 ? 0 : g - l, m = l === 1 / 0 ? n.items.length - 1 : p + l, f = Array.from(
          { length: m - h + 1 },
          (I, b) => h + b
        ), S = [...u, ...f];
        n.items.forEach((I, b) => {
          S.includes(b) ? I.load() : I.unload();
        });
      })
    ).pipe(T(this.destroy$)).subscribe();
  }
  forceOpen(e) {
    const n = e.map(
      (o) => typeof o == "number" ? o : o.index
    ), i = (o) => {
      const r = this.forcedOpenSubject.value;
      n.forEach((s) => {
        const a = r.get(s) || 0, c = o ? a + 1 : a - 1;
        c <= 0 ? r.delete(s) : r.set(s, c);
      }), this.forcedOpenSubject.next(r);
    };
    return i(!0), () => {
      this.isDestroyed || i(!1);
    };
  }
  destroy() {
    super.destroy(), this.forcedOpenSubject.complete();
  }
}
const an = (t, e, n) => {
  const i = (n.width - e.width) / 2, o = (n.height - e.height) / 2;
  return new ce({
    x: t.x - i,
    y: t.y - o
  });
};
class we extends DOMRect {
  constructor() {
    super(...arguments), this.__symbol = /* @__PURE__ */ Symbol("ViewportPosition");
  }
  static from(e, n) {
    if (n) {
      const o = e;
      return new we(
        o.x,
        o.y,
        n.width,
        n.height
      );
    }
    const i = e;
    return new we(i.x, i.y, i.width, i.height);
  }
}
class ns {
  constructor({ width: e, height: n }) {
    this.__symbol = /* @__PURE__ */ Symbol("AbsoluteViewport"), this.width = e, this.height = n;
  }
}
class is {
  constructor({ width: e, height: n }) {
    this.__symbol = /* @__PURE__ */ Symbol("RelativeViewport"), this.width = e, this.height = n;
  }
}
const os = ({
  pageIndex: t,
  spineItemOrId: e,
  spineItemsManager: n
}) => {
  const i = n.items, o = n.get(e);
  if (!o) return;
  const { currentAbsolutePage: r } = i.reduce(
    (s, a) => {
      if (s.found) return s;
      const c = a.numberOfPages;
      return o === a && t <= c - 1 ? {
        currentAbsolutePage: s.currentAbsolutePage + t,
        found: !0
      } : {
        ...s,
        currentAbsolutePage: s.currentAbsolutePage + c
      };
    },
    { currentAbsolutePage: 0, found: !1 }
  );
  return r;
}, rs = ({
  itemHeight: t,
  itemWidth: e,
  visibleWidthOfItem: n,
  visibleHeightOfItem: i,
  threshold: o
}) => {
  const r = n / e, s = i / t;
  return o.type === "percentage" ? r >= o.value && s >= o.value : n >= o.value && i >= o.value;
}, ss = ({
  visibleWidthOfItem: t,
  visibleHeightOfItem: e,
  threshold: n,
  viewportPosition: i
}) => {
  const o = t / i.width, r = e / i.height;
  return n.type === "percentage" ? r >= n.value && o >= n.value : e >= n.value && t >= n.value;
}, cn = ({
  itemPosition: {
    bottom: t,
    left: e,
    right: n,
    top: i,
    width: o,
    height: r
  },
  threshold: s,
  viewportPosition: a,
  restrictToScreen: c
}) => {
  const d = a.x, u = a.x + (a.width - 1), l = a.y, g = Math.max(
    a.y + (a.height - 1),
    0
  ), p = Math.max(
    0,
    Math.min(n, u) - Math.max(e, d)
  ), h = Math.max(
    0,
    Math.min(t, g) - Math.max(i, l)
  );
  if (p <= 0 || h <= 0) return { visible: !1 };
  const f = ss({
    threshold: s,
    visibleHeightOfItem: h,
    visibleWidthOfItem: p,
    viewportPosition: a
  });
  return c ? { visible: f } : {
    visible: rs({
      itemHeight: r,
      itemWidth: o,
      threshold: s,
      visibleHeightOfItem: h,
      visibleWidthOfItem: p
    }) || f
  };
}, Qe = ({
  position: t,
  spineItemsManager: e,
  spineLayout: n
}) => {
  const i = e.items.find((o) => {
    const { left: r, right: s, bottom: a, top: c } = n.getSpineItemSpineLayoutInfo(o), d = t.x >= r && t.x < s, u = t.y >= c && t.y < a;
    return d && u;
  });
  return t.x === 0 && !i ? e.items[0] : i;
}, _e = ({
  spineItemPosition: t,
  itemLayout: { left: e, top: n }
}) => new L({
  x: e + t.x,
  y: n + t.y
}), as = ({
  position: t,
  threshold: e,
  restrictToScreen: n,
  spineItemsManager: i,
  spineLayout: o,
  useAbsoluteViewport: r = !0,
  viewport: s
}) => {
  const a = Qe({
    position: t,
    spineItemsManager: i,
    spineLayout: o
  }) || i.get(0), c = i.items.reduce(
    (p, h) => {
      const m = o.getSpineItemSpineLayoutInfo(h), f = r ? s.absoluteViewport : s.relativeViewport, S = an(
        t,
        s.absoluteViewport,
        f
      ), I = we.from(
        S,
        f
      ), { visible: b } = cn({
        itemPosition: m,
        threshold: e,
        viewportPosition: I,
        restrictToScreen: n
      });
      return b ? [...p, h] : p;
    },
    []
  ), d = c[0] ?? a, u = c[c.length - 1] ?? d;
  if (!d || !u) return;
  const l = i.getSpineItemIndex(d), g = i.getSpineItemIndex(u);
  return {
    beginIndex: l ?? 0,
    endIndex: g ?? 0
  };
}, cs = ({
  spineItemsManager: t,
  context: e,
  spineItemLocator: n,
  settings: i,
  spineLayout: o,
  viewport: r
}) => {
  const s = (h, m) => {
    const { left: f, top: S } = o.getSpineItemSpineLayoutInfo(m);
    return new N({
      /**
       * when using spread the item could be on the right and therefore will be negative
       * @example
       * 400 (position = viewport), page of 200
       * 400 - 600 = -200.
       * However we can assume we are at 0, because we in fact can see the beginning of the item
       */
      x: Math.max(h.x - f, 0),
      y: Math.max(h.y - S, 0)
    });
  }, a = (h) => _e({
    spineItemPosition: new N({ x: 0, y: 0 }),
    itemLayout: o.getSpineItemSpineLayoutInfo(h)
  }), c = (h) => t.items.find((m) => m.renderer.getDocumentFrame() === h), d = (h, m, f) => {
    if (typeof f == "number") {
      const S = t.get(f);
      return S ? n.getSpineItemPageIndexFromNode(
        h,
        m || 0,
        S
      ) : void 0;
    }
    return n.getSpineItemPageIndexFromNode(
      h,
      m || 0,
      f
    );
  }, u = ({
    position: h,
    threshold: m,
    spineItem: f,
    restrictToScreen: S,
    useAbsoluteViewport: I = !0,
    viewport: b
  }) => {
    const F = f.numberOfPages, C = Array.from(Array(F)).map((ae, ue) => {
      const Oe = n.getSpineItemPositionFromPageIndex({
        pageIndex: ue,
        spineItem: f
      }), de = _e({
        spineItemPosition: Oe,
        itemLayout: o.getSpineItemSpineLayoutInfo(f)
      });
      return {
        index: ue,
        absolutePosition: {
          width: b.pageSize.width,
          height: b.pageSize.height,
          left: de.x,
          top: de.y,
          bottom: de.y + b.pageSize.height,
          right: de.x + b.pageSize.width
        }
      };
    }).reduce(
      (ae, { absolutePosition: ue, index: Oe }) => {
        const de = I ? b.absoluteViewport : b.relativeViewport, un = an(
          h,
          b.absoluteViewport,
          de
        ), dn = we.from(
          un,
          de
        ), { visible: ln } = cn({
          viewportPosition: dn,
          restrictToScreen: S,
          threshold: m,
          itemPosition: ue
        });
        return ln ? [...ae, Oe] : ae;
      },
      []
    ), W = C[0], _ = C[C.length - 1] ?? W;
    if (!(W === void 0 || _ === void 0))
      return {
        beginPageIndex: W,
        endPageIndex: _
      };
  };
  return {
    getSpinePositionFromSpineItemPosition: ({
      spineItem: h,
      spineItemPosition: m
    }) => {
      const f = o.getSpineItemSpineLayoutInfo(h);
      return _e({
        itemLayout: f,
        spineItemPosition: m
      });
    },
    /**
     * @deprecated use Pages
     */
    _getAbsolutePageIndexFromPageIndex: (h) => os({
      ...h,
      spineItemsManager: t
    }),
    getSpineItemPagePositionFromSpinePosition: (h) => {
      const m = Qe({
        position: h,
        spineItemsManager: t,
        spineLayout: o
      });
      if (!m)
        return;
      const f = s(
        h,
        m
      ), S = n.getSpineItemPageIndexFromPosition({
        itemWidth: m.layoutInfo.width,
        itemHeight: m.layoutInfo.height,
        position: f,
        isUsingVerticalWriting: !!m.isUsingVerticalWriting()
      }), I = n.getSpineItemPagePositionFromSpineItemPosition(
        f,
        S,
        m
      );
      return {
        spineItem: m,
        spineItemPageIndex: S,
        spineItemPagePosition: I,
        pageSize: r.value.pageSize
      };
    },
    getSpinePositionFromSpineItem: a,
    getSpineItemPositionFromSpinePosition: s,
    getSpineItemFromPosition: (h) => Qe({
      position: h,
      spineItemsManager: t,
      spineLayout: o
    }),
    getSpineItemFromIframe: c,
    getSpineItemPageIndexFromNode: d,
    getVisibleSpineItemsFromPosition: (h) => as({
      spineItemsManager: t,
      spineLayout: o,
      viewport: r,
      ...h
    }),
    getVisiblePagesFromViewportPosition: (h) => u({
      ...h,
      viewport: r
    }),
    isPositionWithinSpineItem: (h, m) => {
      const { bottom: f, left: S, right: I, top: b } = o.getSpineItemSpineLayoutInfo(m);
      return h.x >= S && h.x <= I && h.y <= f && h.y >= b;
    },
    spineItemLocator: n,
    getSafeSpineItemPositionFromUnsafeSpineItemPosition: (h, m) => {
      const { height: f, width: S } = o.getSpineItemSpineLayoutInfo(m);
      return new N({
        x: Math.min(Math.max(0, h.x), S),
        y: Math.min(Math.max(0, h.y), f)
      });
    }
  };
};
class us extends q {
  constructor(e) {
    super(), this.spineItemsManager = e, this.states$ = this.spineItemsManager.items$.pipe(
      P((n) => w(
        ...n.map(
          (i) => i.pipe(
            y((o) => ({ item: i, ...o })),
            D(V)
          )
        )
      )),
      A()
    ), this.itemResize$ = this.spineItemsManager.items$.pipe(
      P((n) => {
        const i = n.map(
          (o) => at(o.element).pipe(
            y((r) => ({ entries: r, item: o }))
          )
        );
        return w(...i);
      }),
      A()
    ), this.itemLoad$ = this.spineItemsManager.items$.pipe(
      P((n) => w(...n.map((i) => i.loaded$.pipe(y(() => i))))),
      A()
    ), this.itemUnload$ = this.spineItemsManager.items$.pipe(
      P((n) => w(
        ...n.map((i) => i.unloaded$.pipe(y(() => i)))
      )),
      A()
    );
  }
}
class ds extends q {
  constructor(e, n, i, o, r) {
    super(), this.spineItemsManager = e, this.spineItemsObserver = n, this.context = i, this.settings = o, this.viewport = r, this.externalLayoutTrigger = new R(), this.spineItemsRelativeLayouts = [];
    const s = w(
      n.itemLoad$,
      n.itemUnload$
    ), a = w(
      this.externalLayoutTrigger,
      s
    );
    this.layout$ = a.pipe(
      x(() => {
        this.spineItemsManager.items.forEach((c) => {
          c.markDirty();
        });
      }),
      De(50),
      P(
        () => this.spineItemsManager.items.reduce(
          (c, d, u) => c.pipe(
            bn(({ horizontalOffset: l, verticalOffset: g }) => {
              const p = l % r.absoluteViewport.width === 0, h = u === e.items.length - 1, m = o.values.computedPageTurnDirection === "vertical", f = i.isRTL(), S = this.getSpreadPosition(
                p,
                f
              ), { edgeX: I, edgeY: b } = this.getStartEdges(
                m,
                p,
                l,
                g,
                r.absoluteViewport.height
              );
              return d.layout({
                spreadPosition: S,
                horizontalOffset: l,
                isLastItem: h,
                edgeX: I,
                edgeY: b
              }).pipe(
                y(({ width: F, height: $ }) => {
                  const C = this.createSpineItemLayout(
                    m,
                    f,
                    I,
                    b,
                    F,
                    $,
                    r.absoluteViewport.width
                  );
                  return this.spineItemsRelativeLayouts[u] = C, {
                    horizontalOffset: I + F,
                    verticalOffset: m ? b + $ : 0
                  };
                })
              );
            })
          ),
          v({ horizontalOffset: 0, verticalOffset: 0 })
        )
      ),
      T(this.destroy$),
      A()
    ), this.layout$.subscribe(), this.watchForVerticalWritingUpdate();
  }
  watchForVerticalWritingUpdate() {
    this.spineItemsObserver.itemLoad$.pipe(
      x((e) => {
        this.context.update({
          hasVerticalWriting: e.isUsingVerticalWriting()
        });
      }),
      T(this.destroy$)
    ).subscribe();
  }
  layout() {
    this.externalLayoutTrigger.next(void 0);
  }
  getSpineItemSpineLayoutInfo(e) {
    const n = this.spineItemsManager.getSpineItemIndex(e) ?? 0;
    return this.spineItemsRelativeLayouts[n] || new je({
      left: 0,
      right: 0,
      top: 0,
      bottom: 0,
      width: 0,
      height: 0,
      x: 0,
      y: 0
    });
  }
  get numberOfPages() {
    return this.spineItemsManager.items.reduce((e, n) => e + n.numberOfPages, 0);
  }
  destroy() {
    super.destroy(), this.externalLayoutTrigger.complete();
  }
  getSpreadPosition(e, n) {
    return this.settings.values.computedSpreadMode ? e ? n ? "right" : "left" : n ? "left" : "right" : "none";
  }
  getStartEdges(e, n, i, o, r) {
    return e ? {
      edgeX: n ? 0 : i,
      edgeY: n ? o : o - r
    } : {
      edgeX: i,
      edgeY: 0
    };
  }
  createSpineItemLayout(e, n, i, o, r, s, a) {
    if (e) {
      const d = r + i, u = s + o;
      return new je({
        left: i,
        right: d,
        top: o,
        bottom: u,
        height: s,
        width: r,
        x: i,
        y: o
      });
    }
    const c = n ? a - i - r : i;
    return new je({
      right: n ? a - i : i + r,
      left: c,
      x: c,
      top: o,
      bottom: s,
      height: s,
      width: r,
      y: o
    });
  }
}
class ls extends q {
  constructor(e, n, i, o, r, s, a) {
    super(), this.context = e, this.pagination = n, this.spineItemsManager = i, this.spineItemLocator = o, this.settings = r, this.hookManager = s, this.viewport = a, this.elementSubject = new Y(
      void 0
    ), this.element$ = this.elementSubject.asObservable(), this.spineItemsObserver = new us(i), this.spineLayout = new ds(
      i,
      this.spineItemsObserver,
      e,
      r,
      a
    ), this.locator = cs({
      context: e,
      spineItemsManager: i,
      spineItemLocator: o,
      settings: r,
      spineLayout: this.spineLayout,
      viewport: a
    }), this.spineItemsLoader = new ts(
      this.context,
      i,
      this.locator,
      r,
      this.spineLayout
    ), this.pages = new pe(
      this.spineLayout,
      this.spineItemsManager,
      this.spineItemLocator,
      this.context,
      this.locator,
      this.viewport
    );
    const c = e.watch("rootElement").pipe(
      me(K),
      k((u) => {
        const l = u.ownerDocument.createElement("div");
        l.style.cssText = `
          height: 100%;
          position: relative;
        `, l.className = `${U}-spine`, this.elementSubject.next(l);
      })
    ), d = O([
      this.context.manifest$,
      this.element$
    ]).pipe(
      k(([u, l]) => {
        if (!l) return;
        this.spineItemsManager.destroyItems();
        const g = u.spineItems.map(
          (p, h) => new sn(
            p,
            l,
            this.context,
            this.settings,
            this.hookManager,
            h,
            this.viewport
          )
        );
        this.spineItemsManager.addMany(g);
      })
    );
    w(d, c).pipe(Q(this.destroy$)).subscribe();
  }
  get element() {
    return this.elementSubject.getValue();
  }
  layout() {
    this.spineLayout.layout();
  }
  getSpineItemSpineLayoutInfo(e) {
    return this.spineLayout.getSpineItemSpineLayoutInfo(e);
  }
  get layout$() {
    return this.pages.layout$;
  }
  destroy() {
    super.destroy(), this.pages.destroy(), this.spineItemsLoader.destroy(), this.elementSubject.getValue()?.remove(), this.elementSubject.complete();
  }
}
class ps extends q {
  constructor(e, n) {
    super(), this.context = e, this.settings = n, this.orderedSpineItemsSubject = new Y([]), this.items$ = this.orderedSpineItemsSubject.asObservable();
  }
  get(e) {
    return typeof e == "number" ? this.orderedSpineItemsSubject.value[e] : typeof e == "string" ? this.orderedSpineItemsSubject.value.find(
      ({ item: n }) => n.id === e
    ) : e;
  }
  comparePositionOf(e, n) {
    const i = this.getSpineItemIndex(e) ?? 0, o = this.getSpineItemIndex(n) ?? 0;
    return i > o ? "after" : i === o ? "same" : "before";
  }
  getSpineItemIndex(e) {
    const n = e instanceof sn ? e : this.get(e);
    if (!n) return;
    const i = this.orderedSpineItemsSubject.value.indexOf(n);
    return i < 0 ? void 0 : i;
  }
  addMany(e) {
    this.orderedSpineItemsSubject.next([
      ...this.orderedSpineItemsSubject.getValue(),
      ...e
    ]);
  }
  // @todo move
  getSpineItemFromCfi(e) {
    const { itemIndex: n } = dt(e);
    if (n !== void 0)
      return this.get(n);
  }
  get items() {
    return this.orderedSpineItemsSubject.value;
  }
  /**
   * @todo handle reload, remove subscription to each items etc. See add()
   */
  destroyItems() {
    this.orderedSpineItemsSubject.value.forEach((e) => {
      e.destroy();
    });
  }
}
class gs extends ee {
  constructor(e, n) {
    const i = document.createElement("div");
    i.setAttribute(`data-${ki}`, ""), super({
      element: i,
      pageSize: {
        width: 1,
        height: 1
      },
      width: 1,
      height: 1
    }), this.context = e, this.settingsManager = n;
    const o = this.settingsManager.watch(["computedSpreadMode"]).pipe(
      x(() => {
        this.mergeCompare({
          pageSize: this.calculatePageSize(this.value)
        });
      })
    ), r = this.context.watch("rootElement").pipe(
      x(() => {
        this.layout();
      })
    );
    w(o, r).pipe(T(this.destroy$)).subscribe();
  }
  calculatePageSize(e) {
    const { computedSpreadMode: n } = this.settingsManager.values;
    return {
      width: n ? e.width / 2 : e.width,
      height: e.height
    };
  }
  layout() {
    const e = {
      width: this.value.element.clientWidth,
      height: this.value.element.clientHeight
    };
    this.mergeCompare({
      pageSize: this.calculatePageSize(e),
      ...e
    });
  }
  get absoluteViewport() {
    return new ns({
      width: this.value.width,
      height: this.value.height
    });
  }
  get pageSize() {
    return this.value.pageSize;
  }
  get scaleFactor() {
    const e = this.absoluteViewport;
    return (this.value.element.getBoundingClientRect()?.width ?? e.width) / e.width;
  }
  /**
   * Returns the relative viewport after eventual transforms.
   * For example if the viewport was zoomed out, we start seeing more left and right
   * items. Therefore we can virtually expand the viewport.
   * Inversely if the viewport is zoomed in, we see less left and right items.
   *
   * This is mostly useful for detecting what should be visible, navigable, etc.
   *
   * @important
   * Contains long floating values.
   *
   * @todo take position of translate into consideration in something
   * like relativeViewportPosition or even better a ViewportSlicePosition
   */
  get relativeViewport() {
    const e = this.absoluteViewport, n = this.scaleFactor;
    return new is({
      width: e.width / n,
      height: e.height / n
    });
  }
}
const At = `${Oi}-core`, hs = (t) => {
  const e = crypto.randomUUID(), n = new R(), i = new R(), o = new hr(), r = new pr(), s = new Jr(t, r), a = new gr(r, s), c = new ps(r, s), d = new gs(r, s), u = on({
    context: r,
    settings: s,
    viewport: d
  }), l = new Yr(r, c), g = new ls(
    r,
    l,
    c,
    u,
    s,
    o,
    d
  ), p = Xr({
    context: r,
    spineItemsManager: c,
    hookManager: o,
    spine: g,
    settings: s,
    viewport: d
  }), h = new qr(
    r,
    l,
    c,
    g,
    u
  );
  p.navigationState$.subscribe(r.bridgeEvent.viewportStateSubject), p.navigation$.subscribe(r.bridgeEvent.navigationSubject), p.locker.isLocked$.subscribe(
    r.bridgeEvent.navigationIsLockedSubject
  ), l.subscribe(r.bridgeEvent.paginationSubject);
  const m = () => {
    n.next();
  }, f = ($) => {
    const { containerElement: C, manifest: W } = $;
    if (r.manifest) {
      E.warn("loading a new book is not supported yet");
      return;
    }
    E.log("load", { options: $ });
    const _ = ms(C, e);
    r.update({
      manifest: W,
      rootElement: _
    }), m();
  }, S = s.watch(["computedSpreadMode"]).pipe(Ot(1), k(m)), I = n.pipe(
    k(() => {
      const $ = r.value.rootElement;
      $ && ($.style.setProperty("overflow", "hidden"), d.layout(), g.layout());
    }),
    Q(i)
  ), b = w(I, S).subscribe();
  it(document, At, mr);
  const F = () => {
    ot(document, At), b.unsubscribe(), c.destroy(), h.destroy(), s.destroy(), l.destroy(), r.destroy(), p.destroy(), g.destroy(), a.destroy(), i.next(), i.complete(), d.destroy();
  };
  return {
    id: e,
    context: r,
    spine: g,
    hookManager: o,
    cfi: {
      generateCfiFromRange: uo,
      parseCfi: dt,
      generateCfiForSpineItemPage: Ge,
      resolveCfi: ($) => Zt({ ...$, spineItemsManager: c })
    },
    navigation: p,
    spineItemsObserver: g.spineItemsObserver,
    spineItemsManager: c,
    layout: m,
    load: f,
    destroy: F,
    pagination: {
      get state() {
        return l.value;
      },
      get state$() {
        return l;
      }
    },
    settings: s,
    renderHeadless: ($) => c.get($)?.renderer.renderHeadless() ?? v(void 0),
    viewport: d,
    viewportState$: r.bridgeEvent.viewportState$,
    viewportFree$: r.bridgeEvent.viewportFree$,
    /**
     * Dispatched when the reader has loaded a book and is rendering a book.
     * Using navigation API and getting information about current content will
     * have an effect.
     * It can typically be used to hide a loading indicator.
     */
    state$: r.manifest$.pipe(
      oe(($) => $ ? "ready" : "idle")
    ),
    features: a,
    $: {
      destroy$: i
    }
  };
}, ms = (t, e) => (t.style.cssText = `
    ${t.style.cssText}
    background-color: white;
    position: relative;
  `, t.classList.add(`${U}-reader`), t.setAttribute(Vi, e), t.setAttribute("data-prose-reader-container", e), t), Ls = (
  //__
  Qo(
    si(
      nr(
        ri(
          Zn(
            qo(
              tr(
                dr(
                  co(
                    Li(
                      Zi(
                        Jn(
                          ti(
                            Bo(
                              Gi(
                                er(
                                  // __
                                  hs
                                )
                              )
                            )
                          )
                        )
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    )
  )
), Ms = (t) => (e) => t(e);
export {
  Yt as AbstractSpinePosition,
  pr as Context,
  Ir as ControlledNavigationController,
  q as DestroyableClass,
  Ee as DocumentRenderer,
  gr as Features,
  Vi as HTML_ATTRIBUTE_DATA_READER_ID,
  U as HTML_PREFIX,
  ji as HTML_PREFIX_SCROLL_NAVIGATOR,
  Ui as HTML_PREFIX_SPINE_ITEM,
  ki as HTML_PREFIX_VIEWPORT,
  Oi as HTML_STYLE_PREFIX,
  hr as HookManager,
  Ar as InternalNavigator,
  ut as LayoutEntry,
  tn as Locker,
  Ts as PAGINATION_UPDATE_AFTER_VIEWPORT_ADJUSTMENT_DEBOUNCE,
  xs as PROSE_READER_NAMESPACE,
  pe as Pages,
  Yr as Pagination,
  qr as PaginationController,
  ee as ReactiveEntity,
  ct as ResourceHandler,
  ho as ResourcesLocator,
  sr as ScrollNavigationController,
  $t as ScrollPosition,
  Zr as SettingsManager,
  ls as Spine,
  sn as SpineItem,
  Mi as SpineItemPageLayout,
  Ei as SpineItemPageSpineLayout,
  N as SpineItemPosition,
  je as SpineItemSpineLayout,
  ps as SpineItemsManager,
  us as SpineItemsObserver,
  L as SpinePosition,
  rr as UnboundScrollPosition,
  ke as UnboundSpineItemPagePosition,
  ce as UnboundSpinePosition,
  $s as VIEWPORT_ADJUSTMENT_THROTTLE,
  gs as Viewport,
  go as consolidate,
  Ls as createReader,
  ht as deferIdle,
  ii as deferNextResult,
  Ge as generateCfiForSpineItemPage,
  uo as generateCfiFromRange,
  le as generateRootCfi,
  gt as getAttributeValueFromString,
  vs as getBase64FromBlob,
  rt as getFrameViewportInfo,
  Cs as getItemAnchor,
  Li as htmlEnhancer,
  _t as idle,
  Yn as injectCSSToFrame,
  K as isDefined,
  zt as isHtmlElement,
  Bn as isHtmlTagElement,
  Ze as isRootCfi,
  As as isShallowEqual,
  Gi as layoutEnhancer,
  Wt as mapKeysTo,
  ws as observeIntersection,
  mt as observeMutation,
  at as observeResize,
  Bo as paginationEnhancer,
  dt as parseCfi,
  ni as pick,
  bs as removeCSS,
  Zt as resolveCfi,
  Ms as rootEnhancer,
  fo as spinePositionToSpineItemSpineLayout,
  re as upsertCSSToFrame,
  qn as waitForFrameLoad,
  Gn as waitForFrameReady,
  Re as waitForSwitch,
  st as watchKeys
};
