// @prose-reader/streamer/dist/index.js
import { Report as se, detectMimeTypeFromName as j, isXmlBasedMimeType as ie, parseContentType as ae, urlJoin as I } from "@prose-reader/shared";
import { XmlDocument as y, XmlTextNode as ce, XmlElement as de } from "xmldoc";
import { Subject as C, mergeMap as O, EMPTY as B, pairwise as le, filter as R, startWith as pe, from as T, tap as P, catchError as W, switchMap as A, merge as X, first as V, takeUntil as me, map as v, ignoreElements as fe, BehaviorSubject as ue, distinctUntilChanged as he, shareReplay as ge, NEVER as H, timer as ye, finalize as _, of as G, lastValueFrom as q } from "rxjs";
const D = (n, e) => {
  const s = n.split(/(\d+)/), r = e.split(/(\d+)/);
  for (let t = 0, o = s.length; t < o; t++)
    if (s[t] !== r[t])
      return s[t]?.match(/\d/) ? +(s[t] || "") - +(r[t] || "") : (s[t] || "").localeCompare(r[t] || "");
  return 1;
}, $ = (n) => n.substring(n.lastIndexOf("/") + 1) || n, J = (n) => n.endsWith("/") ? n.slice(0, -1) : n, be = (n) => {
  const e = n.lastIndexOf("/");
  return e >= 0 ? n.substring(0, e) : "";
}, it = async (n, { orderByAlpha: e, name: s } = {}) => {
  let r = n;
  return e && (r = r.slice().sort((t, o) => D(t.name, o.name))), {
    filename: s || "",
    records: r.map((t) => ({
      dir: t.isDir,
      basename: $(t.name),
      uri: t.name,
      blob: async () => new Blob([await t.data()]),
      string: async () => {
        const o = await t.data();
        return String.fromCharCode.apply(
          null,
          Array.from(new Uint16Array(o))
        );
      },
      size: t.size
    })),
    close: () => Promise.resolve()
  };
}, we = "@prose-reader/streamer", m = se.namespace(we, !1, {
  color: "#ffae42"
}), ve = (n) => {
  const e = {};
  for (const r of n) {
    const t = r.split("/");
    let o = e;
    for (let i = 0; i < t.length; i++) {
      const a = t[i];
      a !== void 0 && (o[a] || (o[a] = {}), o = o[a]);
    }
  }
  const s = (r, t = "") => Object.keys(r).sort().map((o, i, a) => {
    const l = i === a.length - 1, c = t + (l ? "└── " : "├── "), d = t + (l ? "    " : "│   "), p = r[o];
    return p && Object.keys(p).length > 0 ? `${c}${o}/
${s(p, d)}` : `${c}${o}`;
  }).join(`
`);
  return s(e);
}, at = async (n, { orderByAlpha: e, name: s } = {}) => {
  let r = Object.values(n.files);
  e && (r = r.slice().sort((o, i) => D(o.name, i.name)));
  const t = {
    filename: s || "",
    records: r.map((o) => ({
      dir: o.dir,
      basename: $(o.name),
      uri: o.name,
      blob: () => o.async("blob"),
      string: () => o.async("string"),
      ...o.internalStream && {
        stream: o.internalStream
      },
      // this is private API
      // @ts-expect-error
      size: o._data.uncompressedSize
    })),
    close: () => Promise.resolve()
  };
  if (m.log("Generated archive", t), process.env.NODE_ENV === "development" && m.isEnabled()) {
    const o = ve(r.map((i) => i.name));
    m.groupCollapsed(...m.getGroupArgs("Archive folder structure")), m.log(`
${o}`), m.groupEnd();
  }
  return t;
}, ct = async (n, { name: e } = {}) => {
  const s = await n.getFilesArray(), r = {
    close: () => n.close(),
    filename: e ?? "",
    records: s.map((t) => ({
      dir: !1,
      basename: t.file.name,
      size: t.file.size,
      uri: `${t.path}${t.file.name}`,
      blob: async () => await t.file.extract(),
      string: async () => (await t.file.extract()).text()
    }))
  };
  return m.log("Generated archive", r), r;
}, dt = async (n, {
  mimeType: e,
  direction: s
} = { mimeType: "text/plain" }) => {
  const r = `
      <?xml version="1.0" encoding="UTF-8"?>
      <package xmlns="http://www.idpf.org/2007/opf" version="3.0" xml:lang="ja" prefix="rendition: http://www.idpf.org/vocab/rendition/#"
        unique-identifier="ootuya-id">
        <metadata xmlns:opf="http://www.idpf.org/2007/opf" xmlns:dc="http://purl.org/dc/elements/1.1/"
              xmlns:dcterms="http://purl.org/dc/terms/">
              <meta property="rendition:layout">reflowable</meta>
        </metadata>
        <manifest>
            <item id="p01" href="p01.txt" media-type="text/plain"/>
        </manifest>
        <spine page-progression-direction="${s ?? "ltr"}">
          <itemref idref="p01" />
        </spine>
      </package>
    `;
  return {
    filename: "content.txt",
    records: [
      {
        dir: !1,
        basename: $("generated.opf"),
        uri: "generated.opf",
        blob: async () => new Blob([r]),
        string: async () => r,
        size: 0
      },
      {
        dir: !1,
        basename: $("p01.txt"),
        uri: "p01.txt",
        blob: async () => typeof n == "string" ? new Blob([n]) : n,
        string: async () => typeof n == "string" ? n : n.text(),
        size: typeof n == "string" ? n.length : n.size,
        encodingFormat: e
      }
    ],
    close: () => Promise.resolve()
  };
}, lt = async (n, e) => {
  const s = `
    <?xml version="1.0" encoding="UTF-8"?><package xmlns="http://www.idpf.org/2007/opf" version="2.0" unique-identifier="bookid">
      <metadata>
        <meta property="rendition:layout">${e?.useRenditionFlow ? "reflowable" : "pre-paginated"}</meta>
        ${e?.useRenditionFlow ? '<meta property="rendition:flow">scrolled-continuous</meta>' : ""}
      </metadata>
      <manifest>
        ${n.map(
    (o) => `<item id="${$(o)}" href="${o}" media-type="${j(o)}"/>`
  ).join(`
`)}
      </manifest>
      <spine>
        ${n.map((o) => `<itemref idref="${$(o)}" />`).join(`
`)}
      </spine>
    </package>
  `, r = n.map((o) => ({
    dir: !1,
    basename: $(o),
    encodingFormat: j(o),
    uri: o,
    size: 100 / n.length,
    blob: async () => (await fetch(o)).blob(),
    string: async () => ""
  }));
  return {
    filename: "",
    records: [{
      dir: !1,
      basename: "content.opf",
      uri: "content.opf",
      size: 0,
      blob: async () => new Blob(),
      string: async () => s
    }, ...r],
    close: () => Promise.resolve()
  };
}, pt = ({
  enableReport: n
} = {}) => {
  m.enable(!!n);
}, F = (n) => {
  const s = Object.values(n.records).filter(
    (r) => !r.dir
  ).find((r) => r.uri.endsWith(".opf"));
  return {
    data: s,
    basePath: s?.uri.substring(0, s.uri.lastIndexOf("/")) || ""
  };
}, $e = ({ archive: n }) => async (e) => {
  const s = n.records.find(
    (t) => t.basename.toLowerCase() === "com.apple.ibooks.display-options.xml"
  );
  if (!s || s.dir)
    return e;
  const r = await (await s.blob()).text();
  try {
    const i = new y(r).childNamed("platform")?.childrenNamed("option")?.find((a) => a.attr.name === "fixed-layout")?.val || "false";
    return {
      ...e,
      renditionLayout: i === "true" ? "pre-paginated" : e.renditionLayout
    };
  } catch (t) {
    return console.error(
      `Unable to parse com.apple.ibooks.display-options.xml for content
`,
      r
    ), console.error(t), e;
  }
}, xe = async (n) => {
  const e = {
    renditionLayout: void 0
  };
  return await Promise.all(
    n.records.map(async (s) => {
      if (s.uri.endsWith("com.kobobooks.display-options.xml") && !s.dir) {
        const t = new y(await s.string()).childNamed("platform")?.childNamed("option");
        t?.attr?.name === "fixed-layout" && t.val === "true" && (e.renditionLayout = "pre-paginated");
      }
    })
  ), e;
}, Fe = ({ archive: n }) => async (e) => {
  const s = await xe(n);
  return {
    ...e,
    renditionLayout: e.renditionLayout ?? s.renditionLayout
  };
}, Ee = ({ archive: n, baseUrl: e }) => async () => {
  const s = Object.values(n.records).filter((r) => !r.dir);
  return {
    filename: n.filename,
    title: n.records.find(({ dir: r }) => r)?.basename.replace(/\/$/, "") || n.filename,
    renditionLayout: void 0,
    renditionSpread: "auto",
    readingDirection: "ltr",
    spineItems: s.filter((r) => !r.basename.endsWith(".db")).map((r, t) => {
      const o = e || (/^https?:\/\//.test(r.uri) ? "" : "file://");
      return {
        // some books such as cbz can have same basename inside different sub folder
        // we need to make sure to have unique index
        // /chap01/01.png, /chap02/01.png, etc
        id: `${t}.${r.basename}`,
        index: t,
        href: encodeURI(`${o}${r.uri}`),
        renditionLayout: void 0,
        progressionWeight: 1 / s.length,
        pageSpreadLeft: void 0,
        pageSpreadRight: void 0,
        mediaType: r.encodingFormat
      };
    }),
    items: s.map((r, t) => ({
      id: `${t}.${r.basename}`,
      href: encodeURI(`${e}${r.uri}`)
    }))
  };
}, Y = async ({
  archive: n
}) => {
  const { data: e, basePath: s } = F(n) || {}, r = await e?.string();
  if (!r) return [];
  const t = new y(r), o = t.childNamed("manifest"), a = t.childNamed("spine")?.childrenNamed("itemref").map((d) => d.attr.idref), l = o?.childrenNamed("item").filter((d) => a.includes(d.attr.id || "")) || [];
  return n.records.filter((d) => l.find((p) => s ? `${s}/${p.attr.href}` === d.uri : `${p.attr.href}` === d.uri));
}, Ne = (n) => {
  const e = n.attr.properties?.split(" ") || [];
  let s;
  return e.find((r) => r === "rendition:layout-reflowable") && (s = "reflowable"), e.find((r) => r === "rendition:layout-pre-paginated") && (s = "pre-paginated"), {
    renditionLayout: s,
    pageSpreadLeft: e.some((r) => r === "page-spread-left") || void 0,
    pageSpreadRight: e.some((r) => r === "page-spread-right") || void 0
  };
}, Ae = (n, e, s) => {
  const r = n.attr.href || "", t = s?.(n);
  return {
    href: e ? `${t}${e}/${r}` : `${t}${r}`,
    id: n.attr.id || "",
    mediaType: n.attr["media-type"]
  };
}, Q = (n, e, s) => {
  const r = n.childNamed("manifest"), { basePath: t } = F(e) || {};
  return r?.childrenNamed("item")?.map((o) => Ae(o, t, s)) || [];
}, Ie = ({ archive: n, baseUrl: e }) => async (s) => {
  const { data: r, basePath: t } = F(n) || {};
  if (!r)
    return s;
  const o = await r.string();
  m.groupCollapsed(...m.getGroupArgs("OPF data")), m.log("data", o), m.groupEnd();
  const i = new y(o), a = i.childNamed("metadata"), l = i.childNamed("manifest"), c = i.childNamed("spine"), d = i.childNamed("guide"), p = a?.childNamed("dc:title"), g = a?.childrenNamed("meta") || [], b = g.find(
    (f) => f.attr.property === "rendition:layout"
  ), E = g.find(
    (f) => f.attr.property === "rendition:flow"
  ), u = g.find(
    (f) => f.attr.property === "rendition:spread"
  ), L = b?.val, h = E?.val, x = u?.val, S = p?.val || n.records.find(({ dir: f }) => f)?.basename || "", ne = c?.attr["page-progression-direction"], re = (await Y({ archive: n })).reduce(
    (f, N) => N.size + f,
    0
  );
  return {
    filename: n.filename,
    renditionLayout: L,
    renditionFlow: h || "auto",
    renditionSpread: x,
    title: S,
    readingDirection: ne || "ltr",
    /**
     * @see https://www.w3.org/TR/epub/#sec-itemref-elem
     */
    spineItems: c?.childrenNamed("itemref").map((f, N) => {
      const w = l?.childrenNamed("item").find((k) => k.attr.id === f?.attr.idref), M = w?.attr.href || "", oe = n.records.find((k) => k.uri.endsWith(M))?.size || 0, U = e || (/^https?:\/\//.test(M) ? "" : "file://"), z = Ne(f);
      return {
        ...z,
        id: w?.attr.id || "",
        index: N,
        href: w?.attr.href?.startsWith("https://") ? w?.attr.href : t ? `${U}${t}/${w?.attr.href}` : `${U}${w?.attr.href}`,
        renditionLayout: z.renditionLayout ?? L,
        progressionWeight: oe / re,
        // size: itemSize
        mediaType: w?.attr["media-type"]
      };
    }) || [],
    items: Q(i, n, (f) => {
      const N = f.attr.href || "";
      return /^https?:\/\//.test(N) ? "" : e || "file://";
    }),
    guide: d?.childrenNamed("reference").map((f) => ({
      href: f.attr.href || "",
      title: f.attr.title || "",
      type: f.attr.type
    }))
  };
}, Te = (n) => {
  const e = n.descendantWithPath("head")?.childrenNamed("meta").find((s) => s.attr.name === "viewport");
  return !!(e && e.attr.name === "viewport");
}, We = (n) => n.reduce(async (e, s) => {
  if (!await e || !ie({
    mimeType: s.encodingFormat,
    uri: s.uri
  }))
    return !1;
  const t = s.dir ? null : await s.string();
  return t ? Te(new y(t)) : !1;
}, Promise.resolve(!0)), Le = ({ archive: n }) => async (e) => {
  if (e.renditionLayout === "reflowable" && e.spineItems.every((r) => r.renditionLayout === "reflowable")) {
    const r = await Y({ archive: n });
    if (await We(r))
      return {
        ...e,
        spineItems: e.spineItems.map((o) => ({
          ...o,
          renditionLayout: "pre-paginated"
        })),
        renditionLayout: "pre-paginated"
      };
  }
  return e;
}, Se = ({ archive: n }) => async (e) => {
  const s = n.records.find(
    (o) => o.basename.toLowerCase() === "comicinfo.xml" && !o.dir
  );
  if (!s || s.dir)
    return e;
  const r = {
    ...e,
    spineItems: e.spineItems.filter((o) => !o.id.toLowerCase().endsWith("comicinfo.xml")).map((o, i, a) => ({
      ...o,
      progressionWeight: 1 / a.length
    }))
  }, t = await s.string();
  try {
    const i = new y(t).childNamed("Manga")?.val || "unknown";
    return {
      ...r,
      readingDirection: i === "YesAndRightToLeft" ? "rtl" : "ltr"
    };
  } catch (o) {
    return console.error(`Unable to parse comicinfo.xml for content
`, t), console.error(o), r;
  }
}, ke = (n) => n.records.some((e) => e.basename.endsWith(".opf")), Ce = ({ archive: n }) => async (e) => ke(n) ? e : {
  ...e,
  spineItems: e.spineItems.map((r) => {
    const t = n.records.find(
      (i) => decodeURI(r.href).endsWith(i.uri)
    ), o = ae(t?.encodingFormat ?? "") ?? j(t?.basename ?? "");
    return {
      ...r,
      renditionLayout: o?.startsWith("image/") ? "pre-paginated" : r.renditionLayout
    };
  })
}, Z = (n) => n ? n.children.map((e) => e instanceof ce ? e.text : e instanceof de ? Z(e) : "").join("").trim() : "", ee = (n, { basePath: e, baseUrl: s }) => {
  const r = {
    contents: [],
    path: "",
    href: "",
    title: ""
  };
  let t = n.childNamed("span") || n.childNamed("a");
  r.title = (t?.attr.title || t?.val.trim() || Z(t)) ?? "";
  let o = t?.name;
  o !== "a" && (t = n.descendantWithPath(`${o}.a`), t && (o = t.name.toLowerCase())), o === "a" && t?.attr.href && (r.path = I(e, t.attr.href), r.href = I(s, e, t.attr.href));
  const i = n.childNamed("ol");
  if (i) {
    const a = i.childrenNamed("li");
    a && a.length > 0 && (r.contents = a.map(
      (l) => ee(l, { basePath: e, baseUrl: s })
    ));
  }
  return r;
}, Re = (n, { basePath: e, baseUrl: s }) => {
  const r = [];
  let t;
  return n.descendantWithPath("body.nav.ol") ? t = n.descendantWithPath("body.nav.ol")?.children : n.descendantWithPath("body.section.nav.ol") && (t = n.descendantWithPath("body.section.nav.ol")?.children), t && t.length > 0 && t.filter((o) => o.name === "li").forEach((o) => {
    r.push(ee(o, { basePath: e, baseUrl: s }));
  }), r;
}, Pe = async (n, e, { baseUrl: s }) => {
  const r = n.childNamed("manifest")?.childrenNamed("item").find((t) => t.attr.properties === "nav");
  if (r) {
    const t = Object.values(e.records).find(
      (o) => o.uri.endsWith(r.attr.href || "")
    );
    if (t && !t.dir) {
      const o = new y(await t.string()), i = be(t.uri);
      return Re(o, { basePath: i, baseUrl: s });
    }
  }
}, te = (n, {
  opfBasePath: e,
  baseUrl: s,
  prefix: r
}) => {
  const t = n?.childNamed(`${r}content`)?.attr.src || "", o = {
    title: n?.descendantWithPath(`${r}navLabel.${r}text`)?.val || "",
    path: I(e, t),
    href: I(s, e, t),
    contents: []
  }, i = n.childrenNamed(`${r}navPoint`);
  return i && i.length > 0 && (o.contents = i.map(
    (a) => te(a, { opfBasePath: e, baseUrl: s, prefix: r })
  )), o;
}, je = (n, { opfBasePath: e, baseUrl: s }) => {
  const r = [], t = n.name;
  let o = "";
  return t.indexOf(":") !== -1 && (o = `${t.split(":")[0]}:`), n.childNamed(`${o}navMap`)?.childrenNamed(`${o}navPoint`).forEach((i) => {
    r.push(te(i, { opfBasePath: e, baseUrl: s, prefix: o }));
  }), r;
}, Oe = async ({
  opfData: n,
  opfBasePath: e,
  baseUrl: s,
  archive: r
}) => {
  const o = n.childNamed("spine")?.attr.toc;
  if (o) {
    const i = n.childNamed("manifest")?.childrenNamed("item").find((a) => a.attr.id === o);
    if (i) {
      const a = `${e}${e === "" ? "" : "/"}${i.attr.href}`, l = Object.values(r.records).find(
        (c) => c.uri.endsWith(a)
      );
      if (l && !l.dir) {
        const c = new y(await l.string());
        return je(c, { opfBasePath: e, baseUrl: s });
      }
    }
  }
}, De = async (n, e, { baseUrl: s }) => {
  const { basePath: r } = F(e) || {}, t = await Pe(n, e, {
    baseUrl: s
  });
  if (t)
    return t;
  const o = await Oe({
    opfData: n,
    opfBasePath: r,
    archive: e,
    baseUrl: s
  });
  if (o)
    return o;
}, Me = (n, { baseUrl: e }) => {
  const s = [...n.records].sort(
    (t, o) => D(t.uri, o.uri)
  ), r = (t, o, i, a, l) => {
    const c = t.find((g) => g.title === o), [d, ...p] = i;
    return c ? d ? [
      ...t.filter((b) => b !== c),
      {
        ...c,
        contents: [
          ...c.contents,
          ...r(
            c.contents,
            d,
            p,
            a,
            l
          )
        ]
      }
    ] : c.path.split("/").length > l.split("/").length ? [
      ...t.filter((b) => b !== c),
      {
        ...c,
        path: l,
        href: a
      }
    ] : t : d ? [
      ...t,
      {
        contents: r(
          [],
          d,
          p,
          a,
          l
        ),
        href: a,
        path: l,
        title: o
      }
    ] : [
      ...t,
      {
        contents: [],
        href: a,
        path: l,
        title: o
      }
    ];
  };
  return s.reduce((t, o) => {
    if (o.dir) return t;
    const i = o.uri.split("/").slice(0, -1), [a, ...l] = i;
    if (!a) return t;
    const c = I(e, encodeURI(o.uri)).replace(/\/$/, ""), d = o.uri.replace(/\/$/, "");
    return r(t, a, l, c, d);
  }, []);
}, Ue = async (n, { baseUrl: e }) => {
  const { data: s } = F(n) || {};
  if (s && !s.dir) {
    const t = new y(await s.string());
    return await De(t, n, { baseUrl: e }) || [];
  }
  const r = Me(n, { baseUrl: e });
  if (r.length !== 0)
    return r;
}, ze = ({ archive: n, baseUrl: e }) => async (s) => {
  if (s.nav) return s;
  const r = await Ue(n, { baseUrl: e });
  return r ? {
    ...s,
    nav: {
      toc: r
    }
  } : s;
}, Be = async (n, { baseUrl: e = "" } = {}) => {
  const s = [
    Ie({ archive: n, baseUrl: e }),
    Se({ archive: n }),
    $e({ archive: n }),
    Ce({ archive: n }),
    Le({ archive: n }),
    Fe({ archive: n }),
    ze({ archive: n, baseUrl: e })
  ];
  try {
    const r = Ee({ archive: n, baseUrl: e })(), t = await s.reduce(async (o, i) => await i(await o), r);
    if (m.log("Generated manifest", t), process.env.NODE_ENV === "development" && m.isEnabled()) {
      const o = JSON.stringify(t, null, 2);
      m.groupCollapsed(...m.getGroupArgs("Generated manifest")), m.log(`
${o}`), m.groupEnd();
    }
    return t;
  } catch (r) {
    throw m.error(r), r;
  }
}, Xe = (n) => {
  const e = n.descendantWithPath("head")?.childrenNamed("meta").find((s) => s.attr.name === "calibre:cover");
  return !!(e && e.attr.name === "calibre:cover");
}, Ve = (n) => n.descendantWithPath("body")?.descendantWithPath("div")?.childrenNamed("svg")?.find(
  (e) => e.attr.width === "100%" && e.attr.preserveAspectRatio === "none"
), He = ({ archive: n, resourcePath: e }) => async (s) => {
  const r = Object.values(n.records).find(
    (t) => t.uri === e && !t.dir
  );
  if (r && !r.dir && r.basename.endsWith(".xhtml")) {
    const t = s.body ?? await r.string(), o = new y(t);
    if (Xe(o)) {
      const i = Ve(o);
      return i && delete i.attr.preserveAspectRatio, {
        ...s,
        body: o?.toString()
      };
    }
  }
  return s;
}, _e = ({ archive: n, resourcePath: e }) => async (s) => He({ archive: n, resourcePath: e })(s), Ge = ({ archive: n, resourcePath: e }) => async (s) => {
  const r = Object.values(n.records).find(
    (t) => t.uri === e && !t.dir
  );
  if (r && !r.dir && r.basename.endsWith(".css")) {
    const o = (s.body ?? await r.string()).replaceAll(
      "-webkit-writing-mode",
      "writing-mode"
    );
    return {
      ...s,
      body: o
    };
  }
  return s;
}, qe = async (n, e) => {
  const r = await F(n).data?.string();
  if (r) {
    const t = new y(r), o = Q(t, n, () => "");
    if (o.find(
      (a) => e.endsWith(a.href)
    )?.mediaType)
      return {
        mediaType: o.find((a) => e.endsWith(a.href))?.mediaType
      };
  }
  return {
    mediaType: Je(e)
  };
}, Je = (n) => {
  if (n.endsWith(".css"))
    return "text/css; charset=UTF-8";
  if (n.endsWith(".jpg"))
    return "image/jpg";
  if (n.endsWith(".xhtml"))
    return "application/xhtml+xml";
  if (n.endsWith(".mp4"))
    return "video/mp4";
  if (n.endsWith(".svg"))
    return "image/svg+xml";
}, Ke = ({ archive: n, resourcePath: e }) => async (s) => {
  const r = Object.values(n.records).find(
    (o) => o.uri === e && !o.dir
  );
  if (!r || r.dir) return s;
  const t = await qe(n, e);
  return {
    ...s,
    params: {
      ...s.params,
      ...r?.encodingFormat && {
        contentType: r.encodingFormat
      },
      ...t.mediaType && {
        contentType: t.mediaType
      }
    }
  };
}, K = [
  "div",
  "span",
  "p",
  "a",
  "li",
  "ul",
  "ol",
  "h1",
  "h2",
  "h3",
  "h4",
  "h5",
  "h6",
  "table",
  "tr",
  "td",
  "th",
  "thead",
  "tbody",
  "tfoot",
  "section",
  "article",
  "header",
  "footer",
  "nav",
  "aside",
  "main",
  "figure",
  "figcaption",
  "blockquote",
  "pre",
  "code",
  "form",
  "textarea",
  "select",
  "option",
  "button",
  "label",
  "fieldset",
  "legend",
  "caption",
  "dl",
  "dt",
  "dd",
  "iframe",
  "video",
  "audio",
  "canvas",
  "script",
  "style"
], Ye = ({ archive: n, resourcePath: e }) => async (s) => {
  const r = Object.values(n.records).find(
    (t) => t.uri === e && !t.dir
  );
  if (r && !r.dir && r.basename.endsWith(".xhtml")) {
    const t = s.body ?? await r.string();
    if (!new RegExp(
      `<(${K.join("|")})[\\s/>]`,
      "i"
    ).test(t))
      return s;
    const i = new RegExp(
      `<(${K.join("|")})(\\s[^>]*)?\\s*/>`,
      "gi"
    ), a = t.replace(
      i,
      (l, c, d = "") => `<${c} ${d.trim()}></${c}>`
    );
    return {
      ...s,
      body: a
    };
  }
  return s;
}, Qe = async (n, e) => {
  const s = Object.values(n.records).find(
    (o) => o.uri === e && !o.dir
  );
  if (!s || s.dir)
    throw new Error(`no file found for resourcePath:${e}`);
  const r = {
    params: {}
  }, t = [
    Ke({ archive: n, resourcePath: e }),
    Ye({ archive: n, resourcePath: e }),
    Ge({ archive: n, resourcePath: e }),
    _e({ archive: n, resourcePath: e })
  ];
  try {
    const o = await t.reduce(async (i, a) => await a(await i), Promise.resolve(r));
    return m.log("Generated resource", e, o), {
      ...o,
      body: o.body ?? await s.blob()
    };
  } catch (o) {
    throw m.error(o), o;
  }
};
class Ze {
  constructor(e) {
    this.cleanArchiveAfter = e, this.state$ = new ue({
      status: "idle",
      locks: 0
    });
  }
  update(e) {
    this.state$.next({ ...this.state$.getValue(), ...e });
  }
  get locks$() {
    return this.state$.pipe(v(({ locks: e }) => e));
  }
  get state() {
    return this.state$.getValue();
  }
  get isUnlocked$() {
    return this.locks$.pipe(
      v((e) => e <= 0),
      he(),
      ge()
    );
  }
  get overTTL$() {
    return this.isUnlocked$.pipe(
      A(
        (e) => e ? this.cleanArchiveAfter === 1 / 0 ? H : ye(this.cleanArchiveAfter) : H
      )
    );
  }
}
const et = ({
  getArchive: n,
  cleanArchiveAfter: e = 300 * 1e3
  // 5mn
}) => {
  const s = new C(), r = new C(), t = new C(), o = {}, i = s.pipe(
    O((c) => {
      const d = o[c];
      if (!d || d.state.status !== "idle") return B;
      let p = !1;
      const g = (h) => {
        m.debug(`Cleaning up archive with key: ${h}`);
        const x = o[h];
        delete o[h], p || (x?.state.archive?.close(), p = !0);
      };
      d.update({
        status: "loading"
      });
      const b = d.locks$, E = d.isUnlocked$, u = b.pipe(
        le(),
        R(([h, x]) => x > h),
        pe(!0)
      );
      return T(n(c)).pipe(
        P((h) => {
          d.update({
            archive: h,
            status: "success"
          });
        }),
        W((h) => (g(c), d.update({
          status: "error",
          error: h
        }), B)),
        A(() => {
          const h = u.pipe(
            A(() => t),
            A(() => E),
            R((S) => S)
          );
          return X(h, d.overTTL$).pipe(
            V(),
            P(() => {
              g(c);
            })
          );
        })
      );
    }),
    me(r)
  ), a = (c) => {
    let d = !1;
    const p = o[c] ?? new Ze(e);
    o[c] = p, p.update({
      locks: p.state.locks + 1
    });
    const g = () => {
      d || (d = !0, p.update({
        locks: p.state.locks - 1
      }));
    };
    s.next(c);
    const b = p.state$.pipe(
      v(({ archive: u }) => u),
      R((u) => !!u)
    ), E = p.state$.pipe(
      P(({ error: u }) => {
        if (u)
          throw u;
      }),
      fe()
    );
    return X(b, E).pipe(
      V(),
      v((u) => ({ archive: u, release: g })),
      W((u) => {
        throw g(), u;
      })
    );
  }, l = () => {
    t.next();
  };
  return i.subscribe(), {
    access: a,
    purge: l,
    _archives: o
  };
};
class tt {
  constructor({
    onError: e,
    onManifestSuccess: s,
    ...r
  }) {
    this.onError = (t) => (console.error(t), new Response(String(t), { status: 500 })), this.epubLoader = et(r), this.onManifestSuccess = s ?? (({ manifest: t }) => Promise.resolve(t)), this.onError = e ?? this.onError;
  }
  prune() {
    this.epubLoader.purge();
  }
  accessArchive(e) {
    return this.lastAccessedKey !== void 0 && this.lastAccessedKey !== e && this.epubLoader.purge(), this.lastAccessedKey = e, this.epubLoader.access(e);
  }
  accessArchiveWithoutLock(e) {
    return this.accessArchive(e).pipe(
      v(({ archive: s, release: r }) => (r(), s))
    );
  }
  fetchManifest({ key: e, baseUrl: s }) {
    const r = this.accessArchive(e).pipe(
      O(({ archive: t, release: o }) => T(
        Be(t, { baseUrl: s })
      ).pipe(
        A(
          (a) => T(this.onManifestSuccess({ manifest: a, archive: t }))
        ),
        v(
          (a) => new Response(JSON.stringify(a), {
            status: 200
          })
        ),
        _(() => {
          o();
        })
      )),
      W((t) => G(this.onError(t)))
    );
    return q(r);
  }
  fetchResource({
    key: e,
    resourcePath: s
  }) {
    const r = this.accessArchive(e).pipe(
      O(({ archive: t, release: o }) => {
        const i = s.replaceAll("file://", "");
        return T(
          Qe(t, i)
        ).pipe(
          v(
            (l) => new Response(l.body, {
              status: 200,
              headers: {
                ...l.params.contentType && {
                  "Content-Type": l.params.contentType
                }
              }
            })
          ),
          _(() => {
            o();
          })
        );
      }),
      W((t) => G(this.onError(t)))
    );
    return q(r);
  }
}
class mt extends tt {
  constructor({
    getUriInfo: e,
    ...s
  }) {
    super(s), this.getUriInfo = e, this.fetchEventListener = this.fetchEventListener.bind(this);
  }
  fetchEventListener(e) {
    try {
      const s = this.getUriInfo(e);
      if (!s) return;
      const r = J(s.baseUrl), t = e.request.url.substring(
        r.length + 1
      ), [o = ""] = t.split("/"), i = decodeURIComponent(
        J(t.substring(o.length + 1))
      );
      t.endsWith("/manifest") ? e.respondWith(
        this.fetchManifest({ key: o, baseUrl: `${r}/${o}/` })
      ) : e.respondWith(this.fetchResource({ key: o, resourcePath: i }));
    } catch (s) {
      e.respondWith(new Response(String(s), { status: 500 }));
    }
  }
}
export {
  mt as ServiceWorkerStreamer,
  tt as Streamer,
  pt as configure,
  it as createArchiveFromArrayBufferList,
  at as createArchiveFromJszip,
  ct as createArchiveFromLibArchive,
  dt as createArchiveFromText,
  lt as createArchiveFromUrls,
  Be as generateManifestFromArchive,
  Qe as generateResourceFromArchive,
  F as getArchiveOpfInfo,
  be as getUriBasePath,
  $ as getUriBasename,
  J as removeTrailingSlash,
  D as sortByTitleComparator
};
